{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d5fd69ad-af31-45bd-ae34-6519aaa9b527",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from pyteomics import mgf\n",
    "from collections import OrderedDict\n",
    "\n",
    "match_mass_tol = 0.1  # in Da\n",
    "prefix_mass_tol = 0.5  # in Da\n",
    "\n",
    "tools_list = [\n",
    "    \"AdaNovo\",\n",
    "    \"CasanovoV1\",\n",
    "    \"CasanovoV2\",\n",
    "    \"ContraNovo\",\n",
    "    \"DeepNovo\",\n",
    "    \"InstaNovo\",\n",
    "    \"PepNet\",\n",
    "    \"PGPointNovo\",\n",
    "    \"pi-HelixNovo\"\n",
    "    \"pi-PrimeNovo\",\n",
    "    \"pNovo3\",\n",
    "    \"PointNovo\",\n",
    "    \"SMSNet\",   \n",
    "]\n",
    "\n",
    "_PAD = \"_PAD\"\n",
    "_GO = \"_GO\"\n",
    "_EOS = \"_EOS\"\n",
    "_START_VOCAB = [_PAD, _GO, _EOS]\n",
    "\n",
    "PAD_ID = 0\n",
    "GO_ID = 1\n",
    "EOS_ID = 2\n",
    "vocab_reverse = [\n",
    "    \"A\",\n",
    "    \"R\",\n",
    "    \"N\",\n",
    "    \"n\",\n",
    "    \"D\",\n",
    "    \"C\",\n",
    "    \"E\",\n",
    "    \"Q\",\n",
    "    \"q\",\n",
    "    \"G\",\n",
    "    \"H\",\n",
    "    \"I\",\n",
    "    \"L\",\n",
    "    \"K\",\n",
    "    \"M\",\n",
    "    \"m\",\n",
    "    \"F\",\n",
    "    \"P\",\n",
    "    \"S\",\n",
    "    \"T\",\n",
    "    \"W\",\n",
    "    \"Y\",\n",
    "    \"V\",\n",
    "    \"d\",\n",
    "    \"e\",\n",
    "    \"f\",\n",
    "    \"g\",\n",
    "    \"p\",\n",
    "]\n",
    "\n",
    "vocab_reverse = _START_VOCAB + vocab_reverse\n",
    "vocab = dict([(x, y) for (y, x) in enumerate(vocab_reverse)])\n",
    "vocab_size = len(vocab_reverse)\n",
    "\n",
    "# mass value\n",
    "mass_H = 1.0078\n",
    "mass_H2O = 18.0106\n",
    "mass_NH3 = 17.0265\n",
    "mass_N_terminus = 1.0078\n",
    "mass_C_terminus = 17.0027\n",
    "mass_CO = 27.9949\n",
    "mass_Phosphorylation = 79.96633\n",
    "\n",
    "mass_AA = {\n",
    "    \"_PAD\": 0.0,\n",
    "    \"_GO\": mass_N_terminus - mass_H,\n",
    "    \"_EOS\": mass_C_terminus + mass_H,\n",
    "    \"A\": 71.03711,  # 0\n",
    "    \"R\": 156.10111,  # 1\n",
    "    \"N\": 114.04293,  # 2\n",
    "    \"n\": 115.02695,\n",
    "    \"D\": 115.02694,  # 3\n",
    "    \"C\": 160.03065,  # 103.00919,  # 4\n",
    "    \"E\": 129.04259,  # 5\n",
    "    \"Q\": 128.05858,  # 6\n",
    "    \"q\": 129.0426,\n",
    "    \"G\": 57.02146,  # 7\n",
    "    \"H\": 137.05891,  # 8\n",
    "    \"I\": 113.08406,  # 9\n",
    "    \"L\": 113.08406,  # 10\n",
    "    \"K\": 128.09496,  # 11\n",
    "    \"M\": 131.04049,  # 12\n",
    "    \"m\": 147.0354,\n",
    "    \"F\": 147.06841,  # 13\n",
    "    \"P\": 97.05276,  # 14\n",
    "    \"S\": 87.03203,  # 15\n",
    "    \"T\": 101.04768,  # 16\n",
    "    \"W\": 186.07931,  # 17\n",
    "    \"Y\": 163.06333,  # 18\n",
    "    \"V\": 99.06841,  # 19\n",
    "    \"d\": 25.980265, #Carbamylation and NH3 loss\n",
    "    \"e\": -17.026549, # NH3 loss\n",
    "    \"f\": 43.005814, # Carbamylation\n",
    "    \"g\": 42.010565, # Acetylation\n",
    "    \"p\": 111.032, # \"Q\"-17.026549\n",
    "}\n",
    "\n",
    "mass_ID = [mass_AA[vocab_reverse[x]] for x in range(vocab_size)]\n",
    "mass_ID_np = np.array(mass_ID, dtype=np.float32)\n",
    "mass_AA_min = mass_AA[\"G\"]  # 57.02146"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8039d546-e35d-493a-8c8f-395a3aa4d96b",
   "metadata": {},
   "outputs": [],
   "source": [
    "##antibody data process\n",
    "import os\n",
    "def mAbdata_process(path, antibody):\n",
    "    out_path=path+'/process1'\n",
    "    if not os.path.exists(out_path):\n",
    "        os.makedirs(out_path)\n",
    "        os.makedirs(out_path+'/HCD')\n",
    "        os.makedirs(out_path+'/EThcD')\n",
    "    else:\n",
    "        print(\"folder existed:\", out_path)\n",
    "    \n",
    "    in_path = path+'/process/HCD/'\n",
    "    out_path = in_path.replace('process','process1')\n",
    "    files = os.listdir(in_path)\n",
    "    idx = 1\n",
    "    for file in files:\n",
    "        with open(in_path+file,'r') as fr:\n",
    "            lines = fr.readlines()\n",
    "            with open(out_path+file,'w') as fw:\n",
    "                for line in lines:\n",
    "                    if 'TITLE=' in line:\n",
    "                        scan = str(idx)\n",
    "                        idx += 1\n",
    "                        tmp = line.strip().split(', ')\n",
    "                        title = 'TITLE='+antibody+':'+tmp[0].split(': ')[-1]+':'+tmp[-1].split(': ')[-1] #antibody+sample+scan\n",
    "                    elif 'RTINSECONDS=' in line:\n",
    "                        rt = line\n",
    "                    elif 'PEPMASS=' in line:\n",
    "                        mass = line.strip().split(' ')[0]\n",
    "                    elif 'CHARGE=' in line:\n",
    "                        charge = line\n",
    "                        fw.write(title+'\\n')\n",
    "                        fw.write(mass+'\\n')\n",
    "                        fw.write(charge)\n",
    "                        fw.write('SCANS='+tmp[-1].split(': ')[-1]+'\\n')\n",
    "                        fw.write(rt)\n",
    "                        fw.write('SEQ=AAAAAAA'+'\\n')\n",
    "                    elif line.strip():\n",
    "                        fw.write(line)\n",
    "\n",
    "    #Merge data \n",
    "    in_path = path+'/process1/HCD/'\n",
    "    out_file = path+'/process1/spectrum_'+antibody+'_HCD.mgf'\n",
    "    files = os.listdir(in_path)\n",
    "    with open(out_file,'w') as fw:\n",
    "        idx = 1\n",
    "        for file in files:\n",
    "            with open(in_path+file,'r') as fr:\n",
    "                lines = fr.readlines()\n",
    "                for line in lines:\n",
    "                    if 'SCANS=' in line:\n",
    "                        fw.write('SCANS='+str(idx)+'\\n')\n",
    "                    else:\n",
    "                        fw.write(line)  \n",
    "\n",
    "\n",
    "    in_path = path+'/process/EThcD/'\n",
    "    out_path = in_path.replace('process','process1')\n",
    "    files = os.listdir(in_path)\n",
    "    idx = 1\n",
    "    for file in files:\n",
    "        with open(in_path+file,'r') as fr:\n",
    "            lines = fr.readlines()\n",
    "            with open(out_path+file,'w') as fw:\n",
    "                for line in lines:\n",
    "                    if 'TITLE=' in line:\n",
    "                        scan = str(idx)\n",
    "                        idx += 1\n",
    "                        tmp = line.strip().split(', ')\n",
    "                        title = 'TITLE='+antibody+':'+tmp[0].split(': ')[-1]+':'+tmp[-1].split(': ')[-1] #antibody+sample+scan\n",
    "                    elif 'RTINSECONDS=' in line:\n",
    "                        rt = line\n",
    "                    elif 'PEPMASS=' in line:\n",
    "                        mass = line.strip().split(' ')[0]\n",
    "                    elif 'CHARGE=' in line:\n",
    "                        charge = line\n",
    "                        fw.write(title+'\\n')\n",
    "                        fw.write(mass+'\\n')\n",
    "                        fw.write(charge)\n",
    "                        fw.write('SCANS='+tmp[-1].split(': ')[-1]+'\\n')\n",
    "                        fw.write(rt)\n",
    "                        fw.write('SEQ=AAAAAAA'+'\\n')\n",
    "                    elif line.strip():\n",
    "                        fw.write(line)\n",
    "    #Merge data\n",
    "    in_path = path+'/process1/EThcD/'\n",
    "    out_file = path+'/process1/spectrum_'+antibody+'_EThcD.mgf'\n",
    "    files = os.listdir(in_path)\n",
    "    with open(out_file,'w') as fw:\n",
    "        idx = 1\n",
    "        for file in files:\n",
    "            with open(in_path+file,'r') as fr:\n",
    "                lines = fr.readlines()\n",
    "                for line in lines:\n",
    "                    if 'SCANS=' in line:\n",
    "                        fw.write('SCANS='+str(idx)+'\\n')\n",
    "                    else:\n",
    "                        fw.write(line)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccdab1ec-0dbd-4cf4-bd6d-42e546662e40",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df=pd.read_csv('/mnt/data_nas/jwb/antibody-gw-finalData/131_Ab_EThcD_filtered_DB_search_psm_only3mod_Trypsin.csv')\n",
    "df=df[['Peptide','-10lgP','Mass','Length','ppm','m/z','RT','Area','Scan','Accession','PTM','AScore','Z','Protease','Antibody']]\n",
    "df['Spectrum Name']=df['Antibody']+':'+df['Scan']\n",
    "df[\"Modified Sequence\"]=[str(i).replace('M(+15.99)', 'm').replace('Q(+.98)', 'q').replace('N(+.98)', 'n').replace(' ','').replace('C(+57.02)', 'C').replace('I','L') for i in df['Peptide']]\n",
    "df_new=df[['Spectrum Name','Modified Sequence','Length','m/z','Z','Protease']]\n",
    "df_new.to_csv('/mnt/data_nas/jwb/antibody-gw-finalData/131_Ab_EThcD_filtered_DB_search_psm_only3mod_Trypsin_final.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d13e2522-e332-49bc-b12b-3a914425bdc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "def process_casanovoV1(casanovo_path,mgf_file):\n",
    "    \"\"\"Parse de novo results from Casanovo to dataframe.\n",
    "            :param\n",
    "                casanovo_path: path to the mztab file of Casanovo\n",
    "            :return\n",
    "                pnovo_df: dataframe with Score, Peptide, AAScore of Casanovo\n",
    "    \"\"\"\n",
    "    titles=[]\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'TITLE=' in line:\n",
    "                titles.append(line.strip().split('TITLE=')[-1])\n",
    "\n",
    "    casanovo_out = casanovo_path.replace('casanovo_denovo.mztab','CasanovoV1_result.csv')\n",
    "    \n",
    "    with open(casanovo_path) as f:           \n",
    "        # Identify header\n",
    "        for i, line in enumerate(f):\n",
    "            if line.startswith(\"PSH\"):\n",
    "                casanovo_header = i \n",
    "                break\n",
    "            \n",
    "    # Load the data into a DataFrame\n",
    "    df_deno = pd.read_csv(casanovo_path, sep=\"\\t\", header=casanovo_header)\n",
    "    df_denovo = df_deno[df_deno['sequence'].notna()].copy()\n",
    "    df_denovo['PSM_ID'] = df_denovo['PSM_ID'].astype(int)\n",
    "    df_denovo['PSM_ID'] = df_denovo['PSM_ID'].apply(lambda x: titles[x] if x < len(titles) else None)\n",
    "\n",
    "    sequence = df_denovo['sequence'].tolist()\n",
    "    PSM_ID = df_denovo['PSM_ID'].tolist()\n",
    "    Score = [int((i * 100 + 100)/2) for i in df_denovo['search_engine_score[1]'].tolist()]\n",
    "    aascore = df_denovo['opt_ms_run[1]_aa_scores'].tolist()\n",
    "    for i in range(len(aascore)):\n",
    "        aascore[i]=aascore[i].replace(',', ' ')\n",
    "    aascore= [i.replace(',', ' ') for i in aascore]\n",
    "    aascore = [i.split() for i in aascore]\n",
    "    aascore = [[str(int(float(j) * 100)) for j in i] for i in aascore]\n",
    "    df_denovo['aaScore'] = [\" \".join(i) for i in aascore]\n",
    "    aascore = df_denovo['aaScore'].tolist()\n",
    "    scan = df_denovo['PSM_ID'].tolist()\n",
    "    seq = [str(i).replace('M(+15.99)', 'm').replace('Q(+.98)', 'q').replace('N(+.98)', 'n').replace(' ','').replace('C(+57.02)', 'C').replace('I','L') for i in sequence]\n",
    "    DF = np.empty((len(PSM_ID), 4),dtype='object')\n",
    "    DF[:,0] = PSM_ID\n",
    "    DF[:,1] = seq\n",
    "    DF[:,2] = Score\n",
    "    DF[:,3] = aascore \n",
    "\n",
    "    name = ['Spectrum Name','CasanovoV1 Peptide','CasanovoV1 Score','CasanovoV1 aaScore']\n",
    "    df = pd.DataFrame(DF, columns=name)\n",
    "    df.to_csv(casanovo_out,index=False)\n",
    "process_casanovoV1('/mnt/data_nas/jwb/AbNovoBench/denovo/casanovo3/131ab_EThcd/casanovo_denovo.mztab','/mnt/data_nas/jwb/antibody-gw-finalData/131Ab_20250213_DBsearch_reformed_EThcD_only3mod_Trypsin_DENOVO.mgf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1f02187-01b8-4be5-b6b0-d19c416209c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "def process_casanovoV2(casanovo_path,mgf_file):\n",
    "    \"\"\"Parse de novo results from Casanovo to dataframe.\n",
    "            :param\n",
    "                casanovo_path: path to the mztab file of Casanovo\n",
    "            :return\n",
    "                pnovo_df: dataframe with Score, Peptide, AAScore of Casanovo\n",
    "    \"\"\"\n",
    "    titles=[]\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'TITLE=' in line:\n",
    "                titles.append(line.strip().split('TITLE=')[-1])\n",
    "    \n",
    "    casanovo_out = casanovo_path.replace('denovo.mztab','CasanovoV2_result.csv') #be carefull\n",
    "    with open(casanovo_path) as f:           \n",
    "        # Identify header\n",
    "        for i, line in enumerate(f):\n",
    "            if line.startswith(\"PSH\"):\n",
    "                casanovo_header = i \n",
    "                break\n",
    "            \n",
    "    # Load the data into a DataFrame\n",
    "    df_deno = pd.read_csv(casanovo_path, sep=\"\\t\", header=casanovo_header)\n",
    "    df_denovo = df_deno[df_deno['sequence'].notna()].copy()\n",
    "               \n",
    "    df_denovo['PSM_ID'] = [int(item.replace('ms_run[1]:index=', '')) for item in df_denovo['spectra_ref']]\n",
    "    df_denovo['PSM_ID'] = df_denovo['PSM_ID'].apply(lambda x: titles[x] if x < len(titles) else None)\n",
    "\n",
    "    sequence = df_denovo['sequence'].tolist()\n",
    "    PSM_ID = df_denovo['PSM_ID'].tolist()\n",
    "    Score = [int((i * 100 + 100)/2) for i in df_denovo['search_engine_score[1]'].tolist()]\n",
    "    aascore = df_denovo['opt_ms_run[1]_aa_scores'].tolist()\n",
    "    for i in range(len(aascore)):\n",
    "        aascore[i]=aascore[i].replace(',', ' ')\n",
    "    aascore= [i.replace(',', ' ') for i in aascore]\n",
    "    aascore = [i.split() for i in aascore]\n",
    "    aascore = [[str(int(float(j) * 100)) for j in i] for i in aascore]\n",
    "    df_denovo['aaScore'] = [\" \".join(i) for i in aascore]\n",
    "    aascore = df_denovo['aaScore'].tolist()\n",
    "    scan = df_denovo['PSM_ID'].tolist()\n",
    "    seq = [str(i).replace('M(+15.99)', 'm').replace('Q(+.98)', 'q').replace('N(+.98)', 'n').replace(' ','').replace('C(+57.02)', 'C').replace('I','L') for i in sequence]\n",
    "    DF = np.empty((len(PSM_ID), 4),dtype='object')\n",
    "    DF[:,0] = PSM_ID\n",
    "    DF[:,1] = seq \n",
    "    DF[:,2] = Score\n",
    "    DF[:,3] = aascore \n",
    "\n",
    "    name = ['Spectrum Name','CasanovoV2 Peptide','CasanovoV2 Score','CasanovoV2 aaScore']\n",
    "    df = pd.DataFrame(DF, columns=name)\n",
    "    df.to_csv(casanovo_out,index=False)\n",
    "\n",
    "process_casanovoV2('/mnt/data_nas/jwb/AbNovoBench/denovo/casanovo4/131ab_EThcd/denovo.mztab','/mnt/data_nas/jwb/antibody-gw-finalData/131Ab_20250213_DBsearch_reformed_EThcD_only3mod_Trypsin_DENOVO.mgf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46fdbcb3-ede4-47dc-9c92-d58695daa42d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_pNovo3(pnovo_path):\n",
    "    pnovo_df = pd.read_csv(pnovo_path, sep=\"\t\", header=None)\n",
    "    pnovo_df = pnovo_df[[0, 1, 4, 5]]\n",
    "    pnovo_df.columns = ['pNovo3 Index', 'pNovo3 Peptide', 'pNovo3 Score', 'pNovo3 aaScore']\n",
    "    pnovo_title = pnovo_df['pNovo3 Index'].tolist()\n",
    "    pnovo_title = [i.split('TITLE=')[-1] for i in pnovo_title]\n",
    "    pnovo_df['Spectrum Name']=pnovo_title\n",
    "    pnovo_peptide = pnovo_df['pNovo3 Peptide'].tolist()\n",
    "    pnovo_peptide = [str(i).replace('a', 'n').replace('b', 'q').replace('B', 'q').replace('c', 'm').replace('I','L') for i in pnovo_peptide]\n",
    "    pnovo_df['pNovo3 Peptide']=pnovo_peptide\n",
    "    pnovo_score = [int(i) for i in pnovo_df['pNovo3 Score'].tolist()]\n",
    "    pnovo_df['pNovo3 Score']=pnovo_score\n",
    "    aascore = pnovo_df['pNovo3 aaScore'].tolist()\n",
    "    aascore= [i.replace(',', ' ') for i in aascore]\n",
    "    aascore = [i.split() for i in aascore]\n",
    "    aascore = [[str(int(float(j))) for j in i] for i in aascore]\n",
    "    pnovo_df['pNovo3 aaScore'] = [\" \".join(i) for i in aascore]\n",
    "    pnovo_df = pnovo_df[['Spectrum Name', 'pNovo3 Peptide', 'pNovo3 Score', 'pNovo3 aaScore']]\n",
    "    pnovo_df.to_csv(os.path.join(os.path.dirname(pnovo_path), 'pNovo3_result.csv'),index=False)\n",
    "process_pNovo3('/mnt/data/luoling/benchmark_docker/131Ab_hcd_denovo/pNovo', 'results.res')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6f8f6485-c996-4a5b-9c97-79d9f6d29635",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/pnovo/mnt/res/ddw'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def process_PepNet(pepnet_path):\n",
    "    #'/mnt/data_nas/jwb/AbNovoBench/denovo/pepnet/131ab_EThcd/denovo.tsv'\n",
    "    pepnet_df = pd.read_csv(pepnet_path, sep='\\t')\n",
    "    pepnet_title = pepnet_df['TITLE'].tolist()\n",
    "    pepnet_title = [i.split('TITLE=')[-1] for i in pepnet_title]\n",
    "    pepnet_df['Spectrum Name']=pepnet_title\n",
    "    pepnet_peptide = pepnet_df['DENOVO'].tolist()\n",
    "    pepnet_df['PepNet Peptide'] = [i.replace('I','L') for i in pepnet_peptide]\n",
    "    pepnet_score = [int(i*100) for i in pepnet_df['Score'].tolist()]\n",
    "    pepnet_df['PepNet Score']=pepnet_score\n",
    "    aascore = pepnet_df['Positional Score'].tolist()\n",
    "    aascore= [i.replace('[', '').replace(']', '').replace(',', ' ') for i in aascore]\n",
    "    aascore = [i.split() for i in aascore]\n",
    "    aascore = [[str(int(float(j)*100)) for j in i] for i in aascore]\n",
    "    pepnet_df['PepNet aaScore'] = [\" \".join(i) for i in aascore]\n",
    "    pepnet_df = pepnet_df[['Spectrum Name', 'PepNet Peptide', 'PepNet Score', 'PepNet aaScore']]\n",
    "    pepnet_df.to_csv(os.path.join(os.path.dirname(pepnet_path), 'PepNet_result.csv'),index=False)\n",
    "\n",
    "process_PepNet('/mnt/data_nas/jwb/AbNovoBench/denovo/pepnet/131ab_EThcd', 'denovo.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ac15df9-84eb-47bc-9e95-f8787189ee33",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_DeepNovo(deepnovo_path, mgf_file):\n",
    "    \"\"\"parse de novo results from DeepNoo to dataframe\n",
    "\n",
    "            :param\n",
    "                deepnovo_path: path to the result file of DeepNovo\n",
    "            :return\n",
    "                deepnovo_df: dataframe with Score, Peptide, AAScore of DeepNovo\n",
    "    \"\"\"\n",
    "    try:\n",
    "        titles = []\n",
    "        scans = []\n",
    "        with open(mgf_file,'r') as fr:\n",
    "            lines = fr.readlines()\n",
    "            for line in lines:\n",
    "                if 'TITLE=' in line:\n",
    "                    titles.append(line.strip().split('TITLE=')[-1])\n",
    "                elif 'SCANS=' in line:\n",
    "                    scans.append(int(line.strip().split('SCANS=')[-1]))\n",
    "        scan_title_dict = dict(zip(scans, titles))\n",
    "\n",
    "        \n",
    "        with open(deepnovo_path) as f:\n",
    "            deepnovo_df = pd.read_csv(deepnovo_path, sep=\"\t\", header=0)\n",
    "\n",
    "            deepnovo_df = deepnovo_df[['scan', 'predicted_score', 'predicted_sequence',  'predicted_position_score']]\n",
    "\n",
    "            # bug from deepnovo (v.PNAS) which causes result to be wrong aligned\n",
    "\n",
    "            '''\n",
    "            scan    output_score    output_seq  aa_scpre\n",
    "            1       NaN             NaN         NaN\n",
    "            NaN     24              S,E,L       12, 41, 12\n",
    "            2       NaN ...\n",
    "            '''\n",
    "            deepnovo_scan = deepnovo_df['scan'].tolist()\n",
    "            deepnovo_df['scan'] = [scan_title_dict[i] for i in deepnovo_scan]\n",
    "\n",
    "            # change peptide from P,E,P,T,I,D to PEPTID\n",
    "\n",
    "            deepnovo_peptide = deepnovo_df['predicted_sequence'].tolist()\n",
    "            for i in range(len(deepnovo_peptide)):\n",
    "                deepnovo_peptide[i] = str(deepnovo_peptide[i])\n",
    "                deepnovo_peptide[i] = deepnovo_peptide[i].replace(\",\", \"\").replace(\"I\", \"L\").replace(\"Cmod\",\n",
    "                    \"C\").replace(\"Mmod\", \"m\").replace(\"Nmod\", \"n\").replace(\"Qmod\", \"q\")\n",
    "            deepnovo_df['predicted_sequence'] = deepnovo_peptide\n",
    "            deepnovo_df.columns = ['Spectrum Name','DeepNovo Score', 'DeepNovo Peptide', 'DeepNovo aaScore']\n",
    "\n",
    "            # scale peptide score from 0 to 100\n",
    "\n",
    "            deepnovo_score = deepnovo_df['DeepNovo Score']\n",
    "            deepnovo_df['DeepNovo Score'] = [int(np.exp(i) * 100) if not pd.isna(i) else 0 for i in deepnovo_score]\n",
    "\n",
    "            # scale amino acid score from 0 to 100\n",
    "\n",
    "            deepnovo_aascore = deepnovo_df['DeepNovo aaScore'].tolist()\n",
    "            deepnovo_aascore = [str(i).replace(\",\", \" \") for i in deepnovo_aascore]\n",
    "            deepnovo_aascore = [i.split() for i in deepnovo_aascore]\n",
    "            deepnovo_aascore = [[str(int(np.exp(float(j)) * 100)) if j not in ['nan', 'NaN', '', None] else '' for j in i] for i in deepnovo_aascore]\n",
    "            deepnovo_df['DeepNovo aaScore'] = [\" \".join(i) for i in deepnovo_aascore]\n",
    "            deepnovo_df = deepnovo_df[['Spectrum Name','DeepNovo Peptide', 'DeepNovo Score', 'DeepNovo aaScore']]\n",
    "            deepnovo_df.to_csv(os.path.join(os.path.dirname(deepnovo_path),'DeepNovo_result.csv'), index=False)\n",
    "    except IOError:\n",
    "        logger.warning(f\"DeepNovo results not accessible. Make sure they are placed in {deepnovo_path}\")\n",
    "        return pd.DataFrame()\n",
    "process_DeepNovo('/mnt/data_nas/jwb/AbNovoBench/denovo/deepnovo2/131ab_EThcd/denovo.deepnovo_denovo','/mnt/data_nas/jwb/AbNovoBench/denovo/deepnovo2/131ab_EThcd/131ab_EThcd_forDeepNovo.mgf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cf3c4c7-8a7d-45c0-9c1c-5bfa12131b3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def process_pointnovo(pointnovo_path, mgf_file):\n",
    "    \"\"\"parse de novo results from PointNovo to dataframe\n",
    "            :param\n",
    "                pointnovo_path: path to the result file of PointNovo\n",
    "            :return\n",
    "                pointnovo_df: dataframe with ScanNum, Score, Peptide, AAScore of PointNovo\n",
    "    \"\"\"\n",
    "    try:\n",
    "        titles = []\n",
    "        scans = []\n",
    "        with open(mgf_file,'r') as fr:\n",
    "            lines = fr.readlines()\n",
    "            for line in lines:\n",
    "                if 'TITLE=' in line:\n",
    "                    titles.append(line.strip().split('TITLE=')[-1])\n",
    "                elif 'SCANS=' in line:\n",
    "                    scans.append(int(line.strip().split('SCANS=')[-1]))\n",
    "        scan_title_dict = dict(zip(scans, titles))\n",
    "\n",
    "\n",
    "        \n",
    "        with open(pointnovo_path) as f:\n",
    "            pointnovo_df = pd.read_csv(pointnovo_path, sep=\"\\t\", header=0)\n",
    "            pointnovo_df['PSM_ID'] = [scan_title_dict[i] for i in pointnovo_df['feature_id']]\n",
    "            pointnovo_df = pointnovo_df[['PSM_ID', 'predicted_sequence', 'predicted_score', 'predicted_position_score']]\n",
    "\n",
    "            # change peptide from P,E,P,T,I,D to PEPTID\n",
    "\n",
    "            pointnovo_peptide = pointnovo_df['predicted_sequence'].tolist()\n",
    "            for i in range(len(pointnovo_peptide)):\n",
    "                pointnovo_peptide[i] = str(pointnovo_peptide[i])\n",
    "                pointnovo_peptide[i] = pointnovo_peptide[i].replace(\",\", \"\").replace(\"I\", \"L\").replace(\"N(Deamidation)\",\n",
    "                \"n\").replace(\"Q(Deamidation)\", \"q\").replace(\"C(Carbamidomethylation)\", \"C\").replace(\"M(Oxidation)\", \"m\")\n",
    "            pointnovo_df['predicted_sequence'] = pointnovo_peptide\n",
    "            pointnovo_df.columns = ['Spectrum Name','PointNovo Peptide', 'PointNovo Score', 'PointNovo aaScore']\n",
    "\n",
    "            # scale peptide score from 0 to 100\n",
    "\n",
    "            pointnovo_score = pointnovo_df['PointNovo Score']\n",
    "            pointnovo_df['PointNovo Score'] = [int(np.exp(i) * 100) if not pd.isna(i) else 0 for i in pointnovo_score]\n",
    "\n",
    "            # scale AAscore from 0 to 100\n",
    "\n",
    "            pointnovo_aascore = pointnovo_df['PointNovo aaScore'].tolist()\n",
    "            pointnovo_aascore = [str(i).replace(\",\", \" \") for i in pointnovo_aascore]\n",
    "            pointnovo_aascore = [i.split() for i in pointnovo_aascore]\n",
    "            pointnovo_aascore = [[str(int(np.exp(float(j)) * 100)) if j not in ['nan', 'NaN', '', None] else '' for j in i] for i in pointnovo_aascore]\n",
    "            pointnovo_df['PointNovo aaScore'] = [\" \".join(i) for i in pointnovo_aascore]\n",
    "            pointnovo_df.to_csv(os.path.join(os.path.dirname(pointnovo_path),'PointNovo_result.csv'), index=False)\n",
    "    except IOError:\n",
    "        logger.warning(f\"PointNovo results not accessible. Make sure they are placed in {pointnovo_path}\")\n",
    "        return pd.DataFrame()\n",
    "\n",
    "process_PointNovo('/mnt/data_nas/jwb/AbNovoBench/denovo/pointnovo/131ab_EThcd/131ab_EThcd_forPointNovo.deepnovo_denovo','/mnt/data_nas/jwb/AbNovoBench/denovo/pointnovo/131ab_EThcd/131ab_EThcd_forPointNovo.mgf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1eacc643-66f0-40fd-bb99-ae76576150ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_HelixNovo(pinovo_path):\n",
    "    \"\"\"parse de novo results from Pi-HelixNovo to dataframe\n",
    "            :param\n",
    "                pinovo_path: path to the result file of Pi-HelixNovo\n",
    "            :return\n",
    "                pinovo_df: dataframe with Tiltle, Peptide, Score of Pi-HelixNovo\n",
    "    \"\"\"\n",
    "    try:\n",
    "        with open(pinovo_path) as f:\n",
    "            pinovo_df = pd.read_csv(pinovo_path, sep=\"\\t\", header=None)\n",
    "            pinovo_df.columns = [\"Title\", \"sequence\", \"Score\"]\n",
    "            pinovo_title = pinovo_df['Title']\n",
    "            pinovo_df['Title'] = [i.split('TITLE=')[-1] for i in pinovo_title]\n",
    "            pinovo_peptide = pinovo_df['sequence'].tolist()\n",
    "            pinovo_df['pi-HelixNovo Peptide'] = [str(i).replace('M(+15.99)', 'm').replace('Q(+.98)', 'q').replace('N(+.98)', 'n').replace(' ',\n",
    "                                        '').replace('C(+57.02)', 'C').replace('I','L') for i in pinovo_peptide]\n",
    "            \n",
    "            # scale peptide score from 0 to 100\n",
    "\n",
    "            pinovo_score = pinovo_df['Score']\n",
    "            pinovo_df['pi-HelixNovo Score'] = [int(i * 100) for i in pinovo_score]\n",
    "            pinovo_df['Spectrum Name'] = pinovo_df['Title']\n",
    "\n",
    "            pinovo_df = pinovo_df[['Spectrum Name', 'pi-HelixNovo Peptide', \"pi-HelixNovo Score\"]]\n",
    "            pinovo_df.to_csv(os.path.join(os.path.dirname(pinovo_path),'pi-HelixNovo_result.csv'), index=False)\n",
    "    except IOError:\n",
    "        logger.warning(f\"Pi-HelixNovo results not accessible. Make sure they are placed in {pinovo_path}\")\n",
    "        return pd.DataFrame()\n",
    "\n",
    "process_HelixNovo('/mnt/data_nas/jwb/AbNovoBench/denovo/pi-helixnovo/131ab_EThcd/denovo_denovo.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72daa8a5-7f6f-44cd-81eb-51f1822918ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "def process_AdaNovo(adanovo_path,mgf_file):\n",
    "    \"\"\"Parse de novo results from AdaNovo to dataframe.\n",
    "            :param\n",
    "                adanovo_path: path to the mztab file of AdaNovo\n",
    "            :return\n",
    "                pnovo_df: dataframe with Score, Peptide, AAScore of AdaNovo\n",
    "    \"\"\"\n",
    "    titles = []\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'TITLE=' in line:\n",
    "                titles.append(line.strip().split('TITLE=')[-1])\n",
    "    \n",
    "    adanovo_out = adanovo_path.replace('denovo.mztab','AdaNovo_result.csv')\n",
    "    with open(adanovo_path) as f:           \n",
    "        # Identify header\n",
    "        for i, line in enumerate(f):\n",
    "            if line.startswith(\"PSH\"):\n",
    "                adanovo_header = i \n",
    "                break\n",
    "            \n",
    "    # Load the data into a DataFrame\n",
    "    df_deno = pd.read_csv(adanovo_path, sep=\"\\t\", header=adanovo_header)\n",
    "    df_denovo = df_deno[df_deno['sequence'].notna()].copy()\n",
    "\n",
    "    df_denovo['PSM_ID'] = [int(item.replace('ms_run[1]:index=', '')) for item in df_denovo['spectra_ref']]\n",
    "    df_denovo['PSM_ID'] = df_denovo['PSM_ID'].apply(lambda x: titles[x] if x < len(titles) else None)\n",
    "\n",
    "    sequence = df_denovo['sequence'].tolist()\n",
    "    PSM_ID = df_denovo['PSM_ID'].tolist()\n",
    "    Score = [int((i * 100 + 100)/2) for i in df_denovo['search_engine_score[1]'].tolist()]\n",
    "    aascore = df_denovo['opt_ms_run[1]_aa_scores'].tolist()\n",
    "    for i in range(len(aascore)):\n",
    "        aascore[i]=aascore[i].replace(',', ' ')\n",
    "    aascore= [i.replace(',', ' ') for i in aascore]\n",
    "    aascore = [i.split() for i in aascore]\n",
    "    aascore = [[str(int(float(j) * 100)) for j in i] for i in aascore]\n",
    "    df_denovo['aaScore'] = [\" \".join(i) for i in aascore]\n",
    "    aascore = df_denovo['aaScore'].tolist()\n",
    "    scan = df_denovo['PSM_ID'].tolist()\n",
    "    seq = [str(i).replace('M(+15.99)', 'm').replace('Q(+.98)', 'q').replace('N(+.98)', 'n').replace(' ','').replace('C(+57.02)', 'C').replace('I','L') for i in sequence]\n",
    "    DF = np.empty((len(PSM_ID), 4),dtype='object')\n",
    "    DF[:,0] = PSM_ID\n",
    "    DF[:,1] = seq\n",
    "    DF[:,2] = Score\n",
    "    DF[:,3] = aascore \n",
    "\n",
    "    name = ['Spectrum Name','AdaNovo Peptide','AdaNovo Score','AdaNovo aaScore']\n",
    "    df = pd.DataFrame(DF, columns=name)\n",
    "    df.to_csv(adanovo_out,index=False)\n",
    "\n",
    "process_AdaNovo('/mnt/data_nas/jwb/AbNovoBench/denovo/adanovo/131ab_EThcd/denovo.mztab','/mnt/data_nas/jwb/antibody-gw-finalData/131Ab_20250213_DBsearch_reformed_EThcD_only3mod_Trypsin_DENOVO.mgf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25ce9e54-cd83-4285-a257-7a0203f42f71",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "def process_InstaNovo(instanovo_path,mgf_file):\n",
    "    \"\"\"Parse de novo results from InstaNovo to dataframe.\n",
    "            :param\n",
    "                instanovo_path: path to the mztab file of InstaNovo\n",
    "            :return\n",
    "                pnovo_df: dataframe with Score, Peptide, AAScore of InstaNovo\n",
    "    \"\"\"\n",
    "    titles = []\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'TITLE=' in line:\n",
    "                titles.append(line.strip().split('TITLE=')[-1])\n",
    "    \n",
    "    instanovo_out = instanovo_path.replace('predictions_before_refinement.csv','InstaNovo_result.csv')\n",
    "            \n",
    "    # Load the data into a DataFrame\n",
    "    df_deno = pd.read_csv(instanovo_path)\n",
    "    df_denovo = df_deno[df_deno['predictions'].notna()].copy()\n",
    "\n",
    "    df_denovo['PSM_ID'] = [int(item) for item in df_denovo['scan_number']]\n",
    "    df_denovo['PSM_ID'] = df_denovo['PSM_ID'].apply(lambda x: titles[x] if x < len(titles) else None)\n",
    "\n",
    "    sequence = df_denovo['predictions'].tolist()\n",
    "    PSM_ID = df_denovo['PSM_ID'].tolist()\n",
    "    Score = [int(np.exp(i) * 100) for i in df_denovo['log_probabilities'].tolist()]\n",
    "    aascore = df_denovo['token_log_probabilities'].tolist()\n",
    "    for i in range(len(aascore)):\n",
    "        aascore[i]=aascore[i].replace(',', ' ').replace('[','').replace(']','')\n",
    "    aascore= [i.replace(',', ' ') for i in aascore]\n",
    "    aascore = [i.split() for i in aascore]\n",
    "    aascore = [[str(int(np.exp(float(j)) * 100)) for j in i] for i in aascore]\n",
    "\n",
    "    df_denovo['aaScore'] = [\" \".join(i) for i in aascore]\n",
    "    aascore = df_denovo['aaScore'].tolist()\n",
    "    scan = df_denovo['PSM_ID'].tolist()\n",
    "    seq = [str(i).replace('M[UNIMOD:35]', 'm').replace('Q[UNIMOD:7]', 'q').replace('N[UNIMOD:7]', 'n').replace('C[UNIMOD:4]', 'C').replace('I','L') for i in sequence]\n",
    "    DF = np.empty((len(PSM_ID), 4),dtype='object')\n",
    "    DF[:,0] = PSM_ID\n",
    "    DF[:,1] = seq\n",
    "    DF[:,2] = Score\n",
    "    DF[:,3] = aascore \n",
    "\n",
    "    name = ['Spectrum Name','InstaNovo Peptide','InstaNovo Score','InstaNovo aaScore']\n",
    "    df = pd.DataFrame(DF, columns=name)\n",
    "    df.to_csv(instanovo_out,index=False)\n",
    "process_InstaNovo('/mnt/data_nas/jwb/AbNovoBench/denovo/instanovo_new/131ab_EThcd/predictions_before_refinement.csv','/mnt/data_nas/jwb/antibody-gw-finalData/131Ab_20250213_DBsearch_reformed_EThcD_only3mod_Trypsin_DENOVO.mgf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b808e87-3670-4891-9a94-f5951904e032",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "def process_ContraNovo(contranovo_path):\n",
    "    \"\"\"Parse de novo results from ContraNovo to dataframe.\n",
    "            :param\n",
    "                contranovo_path: path to the mztab file of ContraNovo\n",
    "            :return\n",
    "                contranovo_df: dataframe with Score, Peptide, AAScore of ContraNovo\n",
    "    \"\"\"\n",
    "    \n",
    "    contranovo_out = contranovo_path.replace('denovo.mztab','ContraNovo_result.csv')\n",
    "    with open(contranovo_path) as f:           \n",
    "        # Identify header\n",
    "        for i, line in enumerate(f):\n",
    "            if line.startswith(\"PSH\"):\n",
    "                contranovo_header = i \n",
    "                break\n",
    "            \n",
    "    # Load the data into a DataFrame\n",
    "    df_deno = pd.read_csv(contranovo_path, sep=\"\\t\", header=contranovo_header)\n",
    "    df_denovo = df_deno[df_deno['sequence'].notna()].copy()\n",
    "    \n",
    "    df_denovo['PSM_ID'] = [i.split('TITLE=')[-1] for i in df_denovo['spectra_ref']]\n",
    "    df_denovo['PSM_ID'] = [i.replace(']','') for i in df_denovo['PSM_ID']]\n",
    "\n",
    "    sequence = df_denovo['sequence'].tolist()\n",
    "    PSM_ID = df_denovo['PSM_ID'].tolist()\n",
    "    Score = [int((i * 100 + 100)/2) for i in df_denovo['search_engine_score[1]'].tolist()]\n",
    "    aascore = df_denovo['opt_ms_run[1]_aa_scores'].tolist()\n",
    "    for i in range(len(aascore)):\n",
    "        aascore[i]=aascore[i].replace(',', ' ')\n",
    "    aascore= [i.replace(',', ' ') for i in aascore]\n",
    "    aascore = [i.split() for i in aascore]\n",
    "    aascore = [[str(int(float(j) * 100)) for j in i] for i in aascore]\n",
    "    df_denovo['aaScore'] = [\" \".join(i) for i in aascore]\n",
    "    aascore = df_denovo['aaScore'].tolist()\n",
    "    scan = df_denovo['PSM_ID'].tolist()\n",
    "    seq = [str(i).replace('M+15.995', 'm').replace('Q+0.984', 'q').replace('N+0.984', 'n').replace(' ','').replace('C+57.021', 'C').replace('-17.027Q','p').replace('+43.006-17.027','d').replace('-17.027','e').replace('+43.006','f').replace('+42.011','g').replace('I','L') for i in sequence]     \n",
    "    \n",
    "    DF = np.empty((len(PSM_ID), 4),dtype='object')\n",
    "    DF[:,0] = PSM_ID\n",
    "    DF[:,1] = seq\n",
    "    DF[:,2] = Score\n",
    "    DF[:,3] = aascore \n",
    "\n",
    "    name = ['Spectrum Name','ContraNovo Peptide','ContraNovo Score','ContraNovo aaScore']\n",
    "    df = pd.DataFrame(DF, columns=name)\n",
    "    df.to_csv(contranovo_out,index=False)\n",
    "process_ContraNovo('/mnt/data_nas/jwb/AbNovoBench/denovo/contranovo/131ab_EThcd/denovo.mztab')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db3df8d2-f383-4d09-b2b4-6c9fbe6efd79",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_PrimeNovo(primenovo_path):\n",
    "    \"\"\"parse de novo results from pi-PrimeNovo to dataframe\n",
    "            :param\n",
    "                primenovo_path: path to the mztab file of pi-PrimeNovo\n",
    "            :return\n",
    "                pnovo_df: dataframe with Score, Peptide, AAScore of pi-PrimeNovo\n",
    "    \"\"\"\n",
    "    try:\n",
    "        with open(primenovo_path) as f:\n",
    "            \n",
    "            primenovo_df = pd.read_table(primenovo_path)\n",
    "            primenovo_peptide = primenovo_df['prediction'].tolist()\n",
    "            primenovo_df['pi-PrimeNovo Peptide'] = [str(i).replace('M[+15.995]', 'm').replace('Q[+0.984]', 'q').replace('N[+0.984]', 'n').replace(' ',\n",
    "                                        '').replace('C[+57.021]', 'C').replace('[-17.027]-Q', 'p').replace('-[17.027]-Q', 'p').replace('[+43.006-17.027]-','d').replace('+[43.006-17.027]','d').replace('[17.027]-','e').replace('[-17.027]-','e').replace('-[17.027]','e').replace('[+43.006]-','f').replace('+[43.006]','f').replace('[+42.011]-','g').replace('+[42.011]','g').replace('I','L') for i in primenovo_peptide]\n",
    "            \n",
    "            score = primenovo_df['score']\n",
    "            primenovo_df['pi-PrimeNovo Score'] = [int(i * 100) for i in score]\n",
    "            primenovo_df['Spectrum Name'] = [i.split('TITLE=')[-1] for i in primenovo_df['label']]\n",
    "            primenovo_df = primenovo_df[['Spectrum Name', 'pi-PrimeNovo Peptide', 'pi-PrimeNovo Score']]\n",
    "            primenovo_df.to_csv(os.path.join(os.path.dirname(primenovo_path),'pi-PrimeNovo_result.csv'), index=False)\n",
    "    except IOError:\n",
    "        logger.warning(f\"pi-PrimeNovo results not accessible. Make sure they are placed in {primenovo_path}\")\n",
    "        return pd.DataFrame()\n",
    "\n",
    "process_PrimeNovo('/mnt/data_nas/jwb/AbNovoBench/denovo/pi-primenovo/131ab_EThcd/denovo.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d75cd4fe-74b5-4832-bacd-cd8a709cc9d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def process_PGPointNovo(pgpointnovo_path, mgf_file):\n",
    "    \"\"\"parse de novo results from PGPointNovo to dataframe\n",
    "            :param\n",
    "                pgpointnovo_path: path to the result file of PGPointNovo\n",
    "            :return\n",
    "                pgpointnovo_df: dataframe with ScanNum, Score, Peptide, AAScore of PGPointNovo\n",
    "    \"\"\"\n",
    "    try:\n",
    "        titles = []\n",
    "        scans = []\n",
    "        with open(mgf_file,'r') as fr:\n",
    "            lines = fr.readlines()\n",
    "            for line in lines:\n",
    "                if 'TITLE=' in line:\n",
    "                    titles.append(line.strip().split('TITLE=')[-1])\n",
    "                elif 'SCANS=' in line:\n",
    "                    scans.append(int(line.strip().split('SCANS=')[-1]))\n",
    "        scan_title_dict = dict(zip(scans, titles))\n",
    "\n",
    "        with open(pgpointnovo_path) as f:\n",
    "            pgpointnovo_df = pd.read_csv(pgpointnovo_path, sep=\"\\t\", header=0)\n",
    "            pgpointnovo_df['Spectrum Name'] = [scan_title_dict[i] for i in pgpointnovo_df['feature_id']]\n",
    "            pgpointnovo_df = pgpointnovo_df[['Spectrum Name', 'predicted_sequence', 'predicted_score', 'predicted_position_score']]\n",
    "\n",
    "            # change peptide from P,E,P,T,I,D to PEPTID\n",
    "\n",
    "            pgpointnovo_peptide = pgpointnovo_df['predicted_sequence'].tolist()\n",
    "            for i in range(len(pgpointnovo_peptide)):\n",
    "                pgpointnovo_peptide[i] = str(pgpointnovo_peptide[i])\n",
    "                pgpointnovo_peptide[i] = pgpointnovo_peptide[i].replace(\",\", \"\").replace(\"I\", \"L\").replace(\"N(Deamidation)\",\n",
    "                \"n\").replace(\"Q(Deamidation)\", \"q\").replace(\"C(Carbamidomethylation)\", \"C\").replace(\"M(Oxidation)\", \"m\")\n",
    "            pgpointnovo_df['predicted_sequence'] = pgpointnovo_peptide\n",
    "            pgpointnovo_df.columns = ['Spectrum Name', 'PGPointNovo Peptide', 'PGPointNovo Score', 'PGPointNovo aaScore']\n",
    "\n",
    "            # scale peptide score from 0 to 100\n",
    "\n",
    "            pgpointnovo_score = pgpointnovo_df['PGPointNovo Score']\n",
    "            pgpointnovo_df['PGPointNovo Score'] = [int(np.exp(i) * 100) if not pd.isna(i) else 0 for i in pgpointnovo_score]\n",
    "\n",
    "            # scale AAscore from 0 to 100\n",
    "\n",
    "            pgpointnovo_aascore = pgpointnovo_df['PGPointNovo aaScore'].tolist()\n",
    "            pgpointnovo_aascore = [str(i).replace(\",\", \" \") for i in pgpointnovo_aascore]\n",
    "            pgpointnovo_aascore = [i.split() for i in pgpointnovo_aascore]\n",
    "            pgpointnovo_aascore = [[str(int(np.exp(float(j)) * 100)) if j not in ['nan', 'NaN', '', None] else '' for j in i] for i in pgpointnovo_aascore]\n",
    "\n",
    "            pgpointnovo_df['PGPointNovo aaScore'] = [\" \".join(i) for i in pgpointnovo_aascore]\n",
    "            pgpointnovo_df.to_csv(os.path.join(os.path.dirname(pgpointnovo_path),'PGPointNovo_result.csv'), index=False)\n",
    "    except IOError:\n",
    "        logger.warning(f\"PGPointNovo results not accessible. Make sure they are placed in {pgpointnovo_path}\")\n",
    "        return pd.DataFrame()\n",
    "\n",
    "process_PGPointNovo('/mnt/data_nas/jwb/AbNovoBench/denovo/pgpointnovo/131ab_EThcd/denovo.deepnovo_denovo','/mnt/data_nas/jwb/antibody-gw-finalData/131Ab_20250213_DBsearch_reformed_EThcD_only3mod_Trypsin_DENOVO.mgf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81ebb6bf-1cf8-4538-9cf4-d68e960849f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_SMSNet(smsnet_path, mgf_file):\n",
    "    \"\"\"parse de novo results from SMSNet to dataframe\n",
    "\n",
    "            :param\n",
    "                smsnet_path: path to the result file of SMSNet\n",
    "            :return\n",
    "                smsnet_df: dataframe with Score, Peptide, AAScore of SMSNet\n",
    "        \"\"\"\n",
    "    try:\n",
    "        titles = []\n",
    "        with open(mgf_file,'r') as fr:\n",
    "            lines = fr.readlines()\n",
    "            for line in lines:\n",
    "                if 'TITLE=' in line:\n",
    "                    titles.append(line.strip().split('TITLE=')[-1])\n",
    "        \n",
    "        with open(smsnet_path) as f, open(\n",
    "                smsnet_path + '_prob') as g:  # change _rescore and _prob to switch between rescoring and real\n",
    "            smsnet_peptide = pd.Series([line.rstrip() for line in f])\n",
    "            peptide_list = [x.replace(\" \", \"\").replace(\"I\", \"L\").replace(\"<s>\",\"\").replace(\"<unk>\",\"\") for x in\n",
    "                            smsnet_peptide]\n",
    "            smsnet_peptide = pd.DataFrame(peptide_list)\n",
    "            aa_score = g.readlines()\n",
    "            aa_score = [i.strip().split(' ') for i in aa_score]\n",
    "            score_sum = []\n",
    "            for i in range(len(aa_score)):\n",
    "                if not aa_score[i] == ['']:\n",
    "                    for j in range(len(aa_score[i])):\n",
    "                        aa_score[i][j] = int(float(np.exp(float(aa_score[i][j])) * 100))\n",
    "                else:\n",
    "                    aa_score[i] = [0]\n",
    "            for i in range(len(aa_score)):\n",
    "                if not aa_score[i] == [0]:\n",
    "                    score_sum.append(int(statistics.mean(aa_score[i])))\n",
    "                else:\n",
    "                    score_sum.append(0)\n",
    "            df = pd.DataFrame({'aaScore': aa_score, 'Peptide Score': score_sum})\n",
    "            smsnet_df = pd.concat([smsnet_peptide, df], axis=1)\n",
    "            smsnet_df['Spectrum Name'] = titles\n",
    "            smsnet_df.columns = ['SMSNet Peptide', 'SMSNet aaScore', 'SMSNet Score','Spectrum Name']\n",
    "            clist = ['Spectrum Name','SMSNet Peptide','SMSNet Score', 'SMSNet aaScore']\n",
    "            smsnet_df = smsnet_df[clist]\n",
    "            smsnet_df.index = range(1, len(smsnet_df) + 1)\n",
    "\n",
    "            smsnet_aascore = smsnet_df['SMSNet aaScore'].tolist()\n",
    "\n",
    "            # remove peptide predictions which look like \"<s><s><ink>SSSSLASSS\"\n",
    "\n",
    "            smsnet_df['SMSNet aaScore'] = [str(i).replace(' ', '').replace(',', ' '\n",
    "                                            ).replace('[', '').replace(']', '')\n",
    "                                           for i in\n",
    "                                           smsnet_aascore]\n",
    "            smsnet_df.to_csv(os.path.join(os.path.dirname(smsnet_path),'SMSNet_result.csv'), index=False)\n",
    "    except IOError:\n",
    "        logger.warning(f\"SMSNet results not accessible. Make sure they are placed in {smsnet_path}\")\n",
    "        return pd.DataFrame()\n",
    "\n",
    "process_SMSNet('/mnt/data_nas/jwb/AbNovoBench/denovo/SMSNet/131ab_EThcd/131ab_EThcd_forSMSNet','/mnt/data_nas/jwb/AbNovoBench/denovo/SMSNet/131ab_EThcd/131ab_EThcd_forSMSNet.mgf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09b108c0-4005-44ac-8315-d83a8f673121",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "# 主文件路径\n",
    "path = '/mnt/data_nas/jwb/AbNovoBench/denovo'\n",
    "db_search_df = pd.read_csv('/mnt/data_nas/jwb/antibody-gw-finalData/131_Ab_EThcD_filtered_DB_search_psm_only3mod_Trypsin_final.csv')\n",
    "db_search_df.set_index(\"Spectrum Name\", inplace=True)\n",
    "\n",
    "# 工具名称及对应文件夹名\n",
    "tools_list = [\n",
    "    \"AdaNovo\", \"CasanovoV1\", \"CasanovoV2\", \"ContraNovo\", \"DeepNovo\", \"InstaNovo\",\n",
    "    \"PepNet\", \"PGPointNovo\", \"pi-HelixNovo\", \"pi-PrimeNovo\", \"PointNovo\",'SMSNet'\n",
    "]\n",
    "tools_fold = [\n",
    "    \"adanovo\", \"casanovo3\", \"casanovo4\", \"contranovo\", \"deepnovo2\", \"instanovo_new\",\n",
    "    \"pepnet\", \"pgpointnovo\", \"pi-helixnovo\", \"pi-primenovo\", \"pointnovo\",'SMSNet'\n",
    "]\n",
    "\n",
    "# 合并每个工具的结果\n",
    "for tool_name, tool_fold in zip(tools_list, tools_fold):\n",
    "    print(tool_name)\n",
    "    result_file = os.path.join(path, tool_fold,'131ab_EThcd', f\"{tool_name}_result.csv\")\n",
    "    if not os.path.exists(result_file):\n",
    "        print(f\"❌ File not found: {result_file}, skipping.\")\n",
    "        continue\n",
    "\n",
    "    result_df = pd.read_csv(result_file)\n",
    "    result_df.set_index(\"Spectrum Name\", inplace=True)\n",
    "\n",
    "    # 只保留 DB_search 中的 Spectrum Name\n",
    "    result_df = result_df[result_df.index.isin(db_search_df.index)]\n",
    "\n",
    "    # 列名处理\n",
    "    peptide_col = f\"{tool_name} Peptide\"\n",
    "    score_col = f\"{tool_name} Score\"\n",
    "    aa_score_col = f\"{tool_name} aaScore\" if f\"{tool_name} aaScore\" in result_df.columns else None\n",
    "\n",
    "    # 填空处理\n",
    "    result_df[peptide_col] = result_df[peptide_col].fillna('').replace('', '')\n",
    "    result_df[score_col] = result_df[score_col].fillna(0)\n",
    "    if aa_score_col:\n",
    "        result_df[aa_score_col] = result_df[aa_score_col].fillna('').replace('', '')\n",
    "\n",
    "    # 强制 peptide 为空时 score=0，aaScore=''\n",
    "    empty_peptide_mask = result_df[peptide_col] == ''\n",
    "    result_df.loc[empty_peptide_mask, score_col] = 0\n",
    "    if aa_score_col:\n",
    "        result_df.loc[empty_peptide_mask, aa_score_col] = ''\n",
    "\n",
    "    # 要合并的列\n",
    "    merge_cols = [peptide_col, score_col]\n",
    "    if aa_score_col:\n",
    "        merge_cols.append(aa_score_col)\n",
    "\n",
    "    # 合并进主表\n",
    "    db_search_df = db_search_df.join(result_df[merge_cols], how=\"left\")\n",
    "\n",
    "# 写入合并后的文件\n",
    "output_path = os.path.join(path, \"EThcD_Trypsin_DB_search_merged.csv\")\n",
    "db_search_df.reset_index(inplace=True)\n",
    "db_search_df.to_csv(output_path, index=False)\n",
    "print(f\"✅ 合并完成，输出文件：{output_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a1a413c-0471-4944-a449-d561dc5b0b5d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
