{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c9f4cd3-1e38-4722-8311-dfc8e04bf6e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "path = '../../../ALPS_Assembly/'\n",
    "os.mkdir(path)\n",
    "tools_list = [\n",
    "    \"AdaNovo\", \"CasanovoV1\", \"CasanovoV2\", \"ContraNovo\", \"DeepNovo\",\n",
    "    \"PepNet\", \"PGPointNovo\", \"pi-HelixNovo\", \"pi-PrimeNovo\", \"pNovo3\", \"PointNovo\", 'SMSNet', 'InstaNovo'\n",
    "]\n",
    "mabs = os.listdir('../../../monoclonal_antibody/')\n",
    "for tool in tools_list:\n",
    "    path1 = path+tool\n",
    "    os.mkdir(path1)\n",
    "    for mab in mabs:\n",
    "        path2 = path1+'/'+mab\n",
    "        os.mkdir(path2)\n",
    "\n",
    "df = pd.read_csv('../../data/Tool_Confidence_Threshold.csv')\n",
    "SocreAll = pd.Series(df.Confidence_Threshold.values, index=df.Tool).to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db702015-070f-448d-bb48-02b9e48dcac1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "quality_cutoff_ALPS = 50\n",
    "tool = 'CasanovoV1'\n",
    "quality_cutoff_ALPS_local = SocreAll[tool]\n",
    "kmer = 7\n",
    "contigs_ALPS = 20\n",
    "def process_casanovoV1_ALPS(casanovo_path,mgf_file,resultdir):\n",
    "    \"\"\"Parse de novo results from Casanovo to dataframe.\n",
    "            :param\n",
    "                casanovo_path: path to the mztab file of Casanovo\n",
    "            :return\n",
    "                pnovo_df: dataframe with Score, Peptide, AAScore of Casanovo\n",
    "    \"\"\"\n",
    "    titles=[]\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'TITLE=' in line:\n",
    "                titles.append( line.strip().split('TITLE=')[-1] )\n",
    "    \n",
    "    casanovo_out = resultdir + tool + f'_confScoreThreshold_{str(quality_cutoff_ALPS)}_localScore_{str(quality_cutoff_ALPS_local)}.csv'\n",
    "    with open(casanovo_path) as f:           \n",
    "        # Identify header\n",
    "        for i, line in enumerate(f):\n",
    "            if line.startswith(\"PSH\"):\n",
    "                casanovo_header = i \n",
    "                break\n",
    "            \n",
    "    # Load the data into a DataFrame\n",
    "    df_deno = pd.read_csv(casanovo_path, sep=\"\\t\", header=casanovo_header)\n",
    "    df_denovo = df_deno[df_deno['sequence'].notna()].copy()\n",
    "    df_denovo['PSM_ID'] = df_denovo['PSM_ID'].astype(int)\n",
    "    df_denovo['PSM_ID'] = df_denovo['PSM_ID'].apply(lambda x: titles[x] if x < len(titles) else None)\n",
    "\n",
    "    sequence = df_denovo['sequence'].tolist()\n",
    "    PSM_ID = df_denovo['PSM_ID'].tolist()\n",
    "    Score = [int((i * 100 + 100)/2) for i in df_denovo['search_engine_score[1]'].tolist()]\n",
    "    aascore = df_denovo['opt_ms_run[1]_aa_scores'].tolist()\n",
    "    for i in range(len(aascore)):\n",
    "        aascore[i]=aascore[i].replace(',', ' ')\n",
    "    aascore= [i.replace(',', ' ') for i in aascore]\n",
    "    aascore = [i.split() for i in aascore]\n",
    "    aascore = [[str(int(float(j) * 100)) for j in i] for i in aascore]\n",
    "    df_denovo['aaScore'] = [\" \".join(i) for i in aascore]\n",
    "    aascore = df_denovo['aaScore'].tolist()\n",
    "    scan = df_denovo['PSM_ID'].tolist()\n",
    "    DF = np.empty((len(PSM_ID), 5),dtype='object')\n",
    "    DF[:,0] = PSM_ID\n",
    "    DF[:,1] = sequence\n",
    "    DF[:,3] = Score\n",
    "    DF[:,2] = aascore\n",
    "    DF[:,4] = 1 \n",
    "\n",
    "    name = ['Spectrum Name','CasanovoV1 Peptide','CasanovoV1 aaScore', 'CasanovoV1 Score','Area']\n",
    "    df = pd.DataFrame(DF, columns=name)\n",
    "    df['CasanovoV1 Score'] = df['CasanovoV1 Score'].astype(int)\n",
    "    df = df[df['CasanovoV1 Score'] >= SocreAll[tool]]\n",
    "    df.to_csv(casanovo_out,index=False)\n",
    "    subprocess.run(\n",
    "        ('java', '-jar', '../../../ALPS.jar', casanovo_out, str(kmer), str(contigs_ALPS), '>>',\n",
    "        resultdir + 'assembly.log'), stdout=subprocess.DEVNULL)\n",
    "\n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/casanovo3/'\n",
    "out_path = '../../../ALPS_Assembly/CasanovoV1/'\n",
    "mgf_path = '../../../monoclonal_antibody/'\n",
    "mabs = os.listdir(path)\n",
    "for mab in mabs:\n",
    "    casanovo_path = path + mab +'/'+'casanovo_denovo.mztab'\n",
    "    resultdir = out_path + mab +'/'\n",
    "    mgf_file = mgf_path + mab +'/process1/'+'spectrum_'+ Names[mab] +'_HCD.mgf'\n",
    "    process_casanovoV1_ALPS(casanovo_path,mgf_file,resultdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76e8dfaa-3d76-4711-ab2a-6b409c88df66",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "quality_cutoff_ALPS = 50\n",
    "tool = 'CasanovoV2'\n",
    "quality_cutoff_ALPS_local = SocreAll[tool]\n",
    "kmer = 7\n",
    "contigs_ALPS = 20\n",
    "\n",
    "def process_casanovoV2_ALPS(casanovo_path,mgf_file,resultdir):\n",
    "    \"\"\"Parse de novo results from Casanovo to dataframe.\n",
    "            :param\n",
    "                casanovo_path: path to the mztab file of Casanovo\n",
    "            :return\n",
    "                pnovo_df: dataframe with Score, Peptide, AAScore of Casanovo\n",
    "    \"\"\"\n",
    "    titles=[]\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'TITLE=' in line:\n",
    "                titles.append( line.strip().split('TITLE=')[-1] )\n",
    "    \n",
    "    casanovo_out = resultdir + tool + f'_confScoreThreshold_{str(quality_cutoff_ALPS)}_localScore_{str(quality_cutoff_ALPS_local)}.csv'\n",
    "    with open(casanovo_path) as f:           \n",
    "        # Identify header\n",
    "        for i, line in enumerate(f):\n",
    "            if line.startswith(\"PSH\"):\n",
    "                casanovo_header = i \n",
    "                break\n",
    "            \n",
    "    # Load the data into a DataFrame\n",
    "    df_deno = pd.read_csv(casanovo_path, sep=\"\\t\", header=casanovo_header)\n",
    "    df_denovo = df_deno[df_deno['sequence'].notna()].copy()\n",
    "               \n",
    "    df_denovo['PSM_ID'] = [int(item.replace('ms_run[1]:index=', '')) for item in df_denovo['spectra_ref']]\n",
    "    df_denovo['PSM_ID'] = df_denovo['PSM_ID'].apply(lambda x: titles[x] if x < len(titles) else None)\n",
    "\n",
    "    sequence = df_denovo['sequence'].tolist()\n",
    "    PSM_ID = df_denovo['PSM_ID'].tolist()\n",
    "    Score = [int((i * 100 + 100)/2) for i in df_denovo['search_engine_score[1]'].tolist()]\n",
    "    aascore = df_denovo['opt_ms_run[1]_aa_scores'].tolist()\n",
    "    for i in range(len(aascore)):\n",
    "        aascore[i]=aascore[i].replace(',', ' ')\n",
    "    aascore= [i.replace(',', ' ') for i in aascore]\n",
    "    aascore = [i.split() for i in aascore]\n",
    "    aascore = [[str(int(float(j) * 100)) for j in i] for i in aascore]\n",
    "    df_denovo['aaScore'] = [\" \".join(i) for i in aascore]\n",
    "    aascore = df_denovo['aaScore'].tolist()\n",
    "    scan = df_denovo['PSM_ID'].tolist()\n",
    "    DF = np.empty((len(PSM_ID), 5),dtype='object')\n",
    "    DF[:,0] = PSM_ID\n",
    "    DF[:,1] = sequence\n",
    "    DF[:,3] = Score\n",
    "    DF[:,2] = aascore\n",
    "    DF[:,4] = 1 \n",
    "\n",
    "    name = ['Spectrum Name','CasanovoV2 Peptide','CasanovoV2 aaScore', 'CasanovoV2 Score','Area']\n",
    "    df = pd.DataFrame(DF, columns=name)\n",
    "    df['CasanovoV2 Score'] = df['CasanovoV2 Score'].astype(int)\n",
    "    df = df[df['CasanovoV2 Score'] >= SocreAll[tool]]\n",
    "    df.to_csv(casanovo_out,index=False)\n",
    "    subprocess.run(\n",
    "        ('java', '-jar', '../../../ALPS.jar', casanovo_out, str(kmer), str(contigs_ALPS), '>>',\n",
    "        resultdir + 'assembly.log'), stdout=subprocess.DEVNULL)\n",
    "    \n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/casanovo4/'\n",
    "out_path = '../../../ALPS_Assembly/CasanovoV2/'\n",
    "mgf_path = '../../../monoclonal_antibody/'\n",
    "mabs = os.listdir(path)\n",
    "for mab in mabs:\n",
    "    casanovo_path = path + mab +'/'+'denovo.mztab'\n",
    "    resultdir = out_path + mab +'/'\n",
    "    mgf_file = mgf_path + mab +'/process1/'+'spectrum_'+ Names[mab] +'_HCD.mgf'\n",
    "    process_casanovoV2_ALPS(casanovo_path,mgf_file,resultdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0eac20fd-9946-465b-802c-5621e631f89a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "quality_cutoff_ALPS = 50\n",
    "tool = 'PepNet'\n",
    "quality_cutoff_ALPS_local = SocreAll[tool]\n",
    "kmer = 7\n",
    "contigs_ALPS = 20\n",
    "\n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/pepnet/'\n",
    "out_path = '../../../ALPS_Assembly/PepNet/'\n",
    "mgf_path = '../../../monoclonal_antibody/'\n",
    "mabs = os.listdir(path)\n",
    "for mab in mabs:\n",
    "    print(mab)\n",
    "    pepnet_path = path + mab +'/'+'denovo.tsv'\n",
    "    pepnet_df = pd.read_csv(pepnet_path, sep='\\t')\n",
    "    pepnet_title = pepnet_df['TITLE'].tolist()\n",
    "    pepnet_df['Spectrum Name']=pepnet_title\n",
    "    pepnet_peptide = pepnet_df['DENOVO'].tolist()\n",
    "    pepnet_df['PepNet Peptide']=[peptide.replace(\"m\", \"M(+15.99)\").replace(\"q\", \"Q(+.98)\").replace(\"n\", \"N(+.98)\").replace(\"C\", \"C(+57.02)\") for peptide in pepnet_peptide]\n",
    "    pepnet_score = [int(i*100) for i in pepnet_df['Score'].tolist()]\n",
    "    pepnet_df['PepNet Score']=pepnet_score\n",
    "    aascore = pepnet_df['Positional Score'].tolist()\n",
    "    aascore= [i.replace('[', '').replace(']', '').replace(',', ' ') for i in aascore]\n",
    "    aascore = [i.split() for i in aascore]\n",
    "    aascore = [[str(int(float(j)*100)) for j in i] for i in aascore]\n",
    "    pepnet_df['PepNet aaScore'] = [\" \".join(i) for i in aascore]\n",
    "    pepnet_df['Area'] = 1\n",
    "    pepnet_df = pepnet_df[['Spectrum Name', 'PepNet Peptide', 'PepNet aaScore', 'PepNet Score', 'Area']]\n",
    "    pepnet_df['PepNet Score'] = pepnet_df['PepNet Score'].astype(int)\n",
    "    pepnet_df = pepnet_df[pepnet_df['PepNet Score'] >= SocreAll[tool]]\n",
    "    resultdir = out_path + mab +'/'\n",
    "    pepnet_out = resultdir + tool + f'_confScoreThreshold_{str(quality_cutoff_ALPS)}_localScore_{str(quality_cutoff_ALPS_local)}.csv'\n",
    "    pepnet_df.to_csv(pepnet_out,index=False)\n",
    "    subprocess.run(\n",
    "        ('java', '-jar', '../../../ALPS.jar', pepnet_out, str(kmer), str(contigs_ALPS), '>>',\n",
    "        resultdir + 'assembly.log'), stdout=subprocess.DEVNULL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c16eaa0-dce9-4a96-9e0e-70ec7ed0ae19",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "quality_cutoff_Fusion = 50\n",
    "aa_cutoff_Fusion = 50\n",
    "tool = 'pNovo3'\n",
    "quality_cutoff_Fusion_local = SocreAll[tool]\n",
    "\n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/pNovo3/'\n",
    "out_path = '../../../ALPS_Assembly/pNovo3/'\n",
    "mabs = os.listdir(path)\n",
    "for mab in mabs:\n",
    "    print(mab)\n",
    "    resultdir = out_path + mab +'/'\n",
    "    pnovo_path = path + mab + '/HCD/results.res'\n",
    "    df = pd.read_csv(pnovo_path, sep=\"\t\", header=None)\n",
    "    df = df[[0, 1, 4, 5]]\n",
    "    df.columns = ['Spectrum Name', 'pNovo3 Peptide', 'pNovo3 Score', 'pNovo3 aaScore']\n",
    "\n",
    "    peps = list(df['pNovo3 Peptide'])\n",
    "    aas = list(df['pNovo3 aaScore'])\n",
    "    loc = [len(p) == len(a.split(',')) for p, a in zip(peps, aas)]\n",
    "    pnovo_df = df[loc].copy()  \n",
    "\n",
    "    pnovo_peptide = pnovo_df['pNovo3 Peptide'].tolist()\n",
    "    pnovo_peptide = [str(i).replace('I','L').replace('a', 'N(+.98)').replace('b', 'Q(+.98)').replace('B', 'Q(+.98)').replace('c', 'M(+15.99)').replace('C','C(+57.02)') for i in pnovo_peptide]\n",
    "    pnovo_df['pNovo3 Peptide']=pnovo_peptide\n",
    "    pnovo_score = [int(i) for i in pnovo_df['pNovo3 Score'].tolist()]\n",
    "    pnovo_df['pNovo3 Score']=pnovo_score\n",
    "    aascore = pnovo_df['pNovo3 aaScore'].tolist()\n",
    "    aascore= [i.replace(',', ' ') for i in aascore]\n",
    "    aascore = [i.split() for i in aascore]\n",
    "    aascore = [[str(int(float(j))) for j in i] for i in aascore]\n",
    "    pnovo_df['pNovo3 aaScore'] = [\" \".join(i) for i in aascore]\n",
    "    pnovo_df = pnovo_df[['Spectrum Name', 'pNovo3 Peptide', 'pNovo3 aaScore', 'pNovo3 Score']]\n",
    "    pnovo_df['pNovo3 Score'] = pnovo_df['pNovo3 Score'].astype(int)\n",
    "    pnovo_df = pnovo_df[pnovo_df['pNovo3 Score'] >= SocreAll[tool]]\n",
    "    pnovo_df['Area'] = 1\n",
    "    pnovo_out = resultdir + tool + '_confScoreThreshold_' + str(quality_cutoff_Fusion) + '_localScore_' + str(quality_cutoff_Fusion_local) + '.csv'\n",
    "    pnovo_df.to_csv(pnovo_out,index=False)\n",
    "    subprocess.run(\n",
    "        ('java', '-jar', '../../../ALPS.jar', pnovo_out, str(kmer), str(contigs_ALPS), '>>',\n",
    "        resultdir + 'assembly.log'), stdout=subprocess.DEVNULL)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a761658a-3676-4f48-8cbd-91516756bf26",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "quality_cutoff_ALPS = 50\n",
    "tool = 'pi-HelixNovo'\n",
    "quality_cutoff_ALPS_local = SocreAll[tool]\n",
    "kmer = 7\n",
    "contigs_ALPS = 20\n",
    "\n",
    "def process_pinovo_ALPS(pinovo_path,resultdir):\n",
    "    \"\"\"parse de novo results from Pi-HelixNovo to dataframe\n",
    "            :param\n",
    "                pinovo_path: path to the result file of Pi-HelixNovo\n",
    "            :return\n",
    "                pinovo_df: dataframe with Tiltle, Peptide, Score of Pi-HelixNovo\n",
    "    \"\"\"\n",
    "    pinovo_out = resultdir + tool + f'_confScoreThreshold_{str(quality_cutoff_ALPS)}_localScore_{str(quality_cutoff_ALPS_local)}.csv'\n",
    "    pinovo_df = pd.read_csv(pinovo_path, sep=\"\\t\", header=None)\n",
    "    pinovo_df = pinovo_df.dropna()\n",
    "    pinovo_df.columns = [\"Title\", \"sequence\", \"Score\"]\n",
    "    pinovo_title = pinovo_df['Title']\n",
    "    pinovo_df['Title'] = pinovo_title\n",
    "    pinovo_peptide = pinovo_df['sequence'].tolist()\n",
    "    pinovo_df['pi-HelixNovo Peptide'] = [str(i).replace('M(+15.99)', 'm').replace('Q(+.98)', 'q').replace('N(+.98)', 'n').replace(' ',\n",
    "                                        '').replace('C(+57.02)', 'C') for i in pinovo_peptide]\n",
    "            \n",
    "    # scale peptide score from 0 to 100\n",
    "\n",
    "    pinovo_score = pinovo_df['Score']\n",
    "    pinovo_df['pi-HelixNovo Score'] = [int(i * 100) for i in pinovo_score]\n",
    "    pinovo_df['pi-HelixNovo aaScore'] = pinovo_df.apply(lambda row: ' '.join([str(row['pi-HelixNovo Score'])] * len(row['pi-HelixNovo Peptide'])), axis=1)\n",
    "    pinovo_df['pi-HelixNovo Peptide'] = pinovo_peptide\n",
    "    pinovo_df['Spectrum Name'] = pinovo_df['Title']\n",
    "    pinovo_df['Area'] = 1\n",
    "    pinovo_df = pinovo_df[['Spectrum Name', 'pi-HelixNovo Peptide', 'pi-HelixNovo aaScore', 'pi-HelixNovo Score', 'Area']]\n",
    "    pinovo_df['pi-HelixNovo Score'] = pinovo_df['pi-HelixNovo Score'].astype(int)\n",
    "    pinovo_df = pinovo_df[pinovo_df['pi-HelixNovo Score'] >= SocreAll[tool]]\n",
    "    pinovo_df.to_csv(pinovo_out,index=False)\n",
    "    subprocess.run(\n",
    "        ('java', '-jar', '../../../ALPS.jar', pinovo_out, str(kmer), str(contigs_ALPS), '>>',\n",
    "        resultdir + 'assembly.log'), stdout=subprocess.DEVNULL)\n",
    "    \n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/pi-helixnovo/'\n",
    "out_path = '../../../ALPS_Assembly/pi-HelixNovo/'\n",
    "mgf_path = '../../../monoclonal_antibody/'\n",
    "mabs = os.listdir(path)\n",
    "for mab in mabs:\n",
    "    pinovo_path = path + mab +'/'+'denovo_denovo.txt'\n",
    "    resultdir = out_path + mab +'/'\n",
    "    process_pinovo_ALPS(pinovo_path, resultdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6e82b9e-cb6c-477a-98f1-ad5584694799",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "quality_cutoff_ALPS = 50\n",
    "tool = 'pi-PrimeNovo'\n",
    "quality_cutoff_ALPS_local = SocreAll[tool]\n",
    "kmer = 7\n",
    "contigs_ALPS = 20\n",
    "\n",
    "\n",
    "def process_primenovo_ALPS(primenovo_path, resultdir):\n",
    "    \"\"\"parse de novo results from pi-PrimeNovo to dataframe\n",
    "            :param\n",
    "                primenovo_path: path to the mztab file of pi-PrimeNovo\n",
    "            :return\n",
    "                pnovo_df: dataframe with Score, Peptide, AAScore of pi-PrimeNovo\n",
    "    \"\"\" \n",
    "    primenovo_out = resultdir + tool + f'_confScoreThreshold_{str(quality_cutoff_ALPS)}_localScore_{str(quality_cutoff_ALPS_local)}.csv'\n",
    "    primenovo_df = pd.read_table(primenovo_path)\n",
    "    primenovo_peptide = primenovo_df['prediction'].tolist()\n",
    "    primenovo_df['pi-PrimeNovo Peptide'] = [str(i).replace('M[+15.995]', 'm').replace('Q[+0.984]', 'q').replace('N[+0.984]', 'n').replace(' ',\n",
    "                    '').replace('C[+57.021]', 'C').replace('[-17.027]-Q', 'p').replace('-[17.027]-Q', 'p').replace('[+43.006-17.027]-','').replace('+[43.006-17.027]','').replace('[17.027]-','').replace('[-17.027]-','').replace('-[17.027]','').replace('[+43.006]-','').replace('+[43.006]','').replace('[+42.011]-','').replace('+[42.011]','') for i in primenovo_peptide]\n",
    "    score = primenovo_df['score']\n",
    "    primenovo_df['pi-PrimeNovo Score'] = [int(i * 100) for i in score]\n",
    "    primenovo_df['pi-PrimeNovo aaScore'] = primenovo_df.apply(lambda row: ' '.join([str(row['pi-PrimeNovo Score'])] * len(row['pi-PrimeNovo Peptide'])), axis=1)\n",
    "    primenovo_df['pi-PrimeNovo Peptide'] = [str(i).replace('M[+15.995]', 'M(+15.99)').replace('Q[+0.984]', 'Q(+.98)').replace('N[+0.984]', 'N(+.98)').replace(' ',\n",
    "                    '').replace('C[+57.021]', 'C(+57.02)').replace('[-17.027]-Q', 'Q(-17.03)').replace('-[17.027]-Q', 'Q(-17.03)').replace('[+43.006-17.027]-','').replace('+[43.006-17.027]','').replace('[17.027]-','').replace('[-17.027]-','').replace('-[17.027]','').replace('[+43.006]-','').replace('+[43.006]','').replace('[+42.011]-','').replace('+[42.011]','') for i in primenovo_peptide]\n",
    "    primenovo_df['Spectrum Name'] = primenovo_df['label']\n",
    "    primenovo_df['Area'] = 1\n",
    "    primenovo_df = primenovo_df[['Spectrum Name', 'pi-PrimeNovo Peptide', 'pi-PrimeNovo aaScore', 'pi-PrimeNovo Score', 'Area']]\n",
    "    primenovo_df['pi-PrimeNovo Score'] = primenovo_df['pi-PrimeNovo Score'].astype(int)\n",
    "    primenovo_df = primenovo_df[primenovo_df['pi-PrimeNovo Score'] >= SocreAll[tool]]\n",
    "    primenovo_df.to_csv(primenovo_out,index=False)\n",
    "    subprocess.run(\n",
    "        ('java', '-jar', '../../../ALPS.jar', primenovo_out, str(kmer), str(contigs_ALPS), '>>',\n",
    "        resultdir + 'assembly.log'), stdout=subprocess.DEVNULL)\n",
    "\n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/pi-primenovo/'\n",
    "out_path = '../../../ALPS_Assembly/pi-PrimeNovo/'\n",
    "mgf_path = '../../../monoclonal_antibody/'\n",
    "mabs = os.listdir(path)\n",
    "for mab in mabs:\n",
    "    primenovo_path = path + mab +'/'+'denovo.tsv'\n",
    "    resultdir = out_path + mab +'/'\n",
    "    process_primenovo_ALPS(primenovo_path, resultdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07ba9bd0-7ce3-42bc-868d-4e4579dd28d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "quality_cutoff_ALPS = 50\n",
    "tool = 'AdaNovo'\n",
    "quality_cutoff_ALPS_local = SocreAll[tool]\n",
    "kmer = 7\n",
    "contigs_ALPS = 20\n",
    "\n",
    "def process_adanovo_ALPS(adanovo_path, mgf_file, resultdir):\n",
    "    \"\"\"Parse de novo results from AdaNovo to dataframe.\n",
    "            :param\n",
    "                adanovo_path: path to the mztab file of AdaNovo\n",
    "            :return\n",
    "                pnovo_df: dataframe with Score, Peptide, AAScore of AdaNovo\n",
    "    \"\"\"\n",
    "    titles=[]\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'TITLE=' in line:\n",
    "                titles.append( line.strip().split('TITLE=')[-1] )\n",
    "    \n",
    "    adanovo_out = resultdir + tool + f'_confScoreThreshold_{str(quality_cutoff_ALPS)}_localScore_{str(quality_cutoff_ALPS_local)}.csv'\n",
    "    with open(adanovo_path) as f:           \n",
    "        # Identify header\n",
    "        for i, line in enumerate(f):\n",
    "            if line.startswith(\"PSH\"):\n",
    "                adanovo_header = i \n",
    "                break\n",
    "            \n",
    "    # Load the data into a DataFrame\n",
    "    df_deno = pd.read_csv(adanovo_path, sep=\"\\t\", header=adanovo_header)\n",
    "    df_denovo = df_deno[df_deno['sequence'].notna()].copy()\n",
    "\n",
    "    df_denovo['PSM_ID'] = [int(item.replace('ms_run[1]:index=', '')) for item in df_denovo['spectra_ref']]\n",
    "    df_denovo['PSM_ID'] = df_denovo['PSM_ID'].apply(lambda x: titles[x] if x < len(titles) else None)\n",
    "    PSM_ID = df_denovo['PSM_ID'].tolist()\n",
    "    sequence = df_denovo['sequence'].tolist()\n",
    "    Score = [int((i * 100 + 100)/2) for i in df_denovo['search_engine_score[1]'].tolist()]\n",
    "    aascore = df_denovo['opt_ms_run[1]_aa_scores'].tolist()\n",
    "    for i in range(len(aascore)):\n",
    "        aascore[i]=aascore[i].replace(',', ' ')\n",
    "    aascore= [i.replace(',', ' ') for i in aascore]\n",
    "    aascore = [i.split() for i in aascore]\n",
    "    aascore = [[str(int(float(j) * 100)) for j in i] for i in aascore]\n",
    "    df_denovo['aaScore'] = [\" \".join(i) for i in aascore]\n",
    "    aascore = df_denovo['aaScore'].tolist()\n",
    "\n",
    "    DF = np.empty((len(PSM_ID), 5),dtype='object')\n",
    "    DF[:,0] = PSM_ID\n",
    "    DF[:,1] = sequence\n",
    "    DF[:,3] = Score\n",
    "    DF[:,2] = aascore\n",
    "    DF[:,4] = 1 \n",
    "    name = ['Spectrum Name','AdaNovo Peptide','AdaNovo aaScore','AdaNovo Score','Area']\n",
    "    df = pd.DataFrame(DF, columns=name)\n",
    "    df['AdaNovo Score'] = df['AdaNovo Score'].astype(int)\n",
    "    df = df[df['AdaNovo Score'] >= SocreAll[tool]]\n",
    "    df.to_csv(adanovo_out,index=False)\n",
    "    subprocess.run(\n",
    "        ('java', '-jar', '../../../ALPS.jar', adanovo_out, str(kmer), str(contigs_ALPS), '>>',\n",
    "        resultdir + 'assembly.log'), stdout=subprocess.DEVNULL)\n",
    "\n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/adanovo/'\n",
    "out_path = '../../../ALPS_Assembly/AdaNovo/'\n",
    "mgf_path = '../../../monoclonal_antibody/'\n",
    "mabs = os.listdir(path)\n",
    "for mab in mabs:\n",
    "    adanovo_path = path + mab +'/'+'denovo.mztab'\n",
    "    resultdir = out_path + mab +'/'\n",
    "    mgf_file = mgf_path + mab +'/process1/'+'spectrum_'+ Names[mab] +'_HCD.mgf'\n",
    "    process_adanovo_ALPS(adanovo_path,mgf_file,resultdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02cf7664-86f7-4c06-aaff-f939f69f1d57",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "quality_cutoff_ALPS = 50\n",
    "tool = 'ContraNovo'\n",
    "quality_cutoff_ALPS_local = SocreAll[tool]\n",
    "kmer = 7\n",
    "contigs_ALPS = 20\n",
    "\n",
    "def process_contranovo_ALPS(contranovo_path, resultdir):\n",
    "    \"\"\"Parse de novo results from ContraNovo to dataframe.\n",
    "            :param\n",
    "                contranovo_path: path to the mztab file of ContraNovo\n",
    "            :return\n",
    "                contranovo_df: dataframe with Score, Peptide, AAScore of ContraNovo\n",
    "    \"\"\"\n",
    "    \n",
    "    contranovo_out = resultdir + tool + f'_confScoreThreshold_{str(quality_cutoff_ALPS)}_localScore_{str(quality_cutoff_ALPS_local)}.csv'\n",
    "    with open(contranovo_path) as f:           \n",
    "        # Identify header\n",
    "        for i, line in enumerate(f):\n",
    "            if line.startswith(\"PSH\"):\n",
    "                contranovo_header = i \n",
    "                break\n",
    "            \n",
    "    # Load the data into a DataFrame\n",
    "    df_deno = pd.read_csv(contranovo_path, sep=\"\\t\", header=contranovo_header)\n",
    "    df_denovo = df_deno[df_deno['sequence'].notna()].copy()\n",
    "    \n",
    "    df_denovo['PSM_ID'] = [i.replace('ms_run[','').replace(']','') for i in df_denovo['spectra_ref']]\n",
    "\n",
    "    sequence = df_denovo['sequence'].tolist()\n",
    "    #sequence = [str(i).replace('M+15.995', 'M(+15.99)').replace('Q+0.984', 'Q(+.98)').replace('N+0.984', 'N(+.98)').replace(' ','').replace('C+57.021', 'C(+57.02)').replace('-17.027Q','').replace('+43.006-17.027','').replace('-17.027','').replace('+43.006','').replace('+42.011','') for i in sequence]     \n",
    "    \n",
    "    PSM_ID = df_denovo['PSM_ID'].tolist()\n",
    "    Score = [int((i * 100 + 100)/2) for i in df_denovo['search_engine_score[1]'].tolist()]\n",
    "    aascore = df_denovo['opt_ms_run[1]_aa_scores'].tolist()\n",
    "    for i in range(len(aascore)):\n",
    "        aascore[i]=aascore[i].replace(',', ' ')\n",
    "    aascore= [i.replace(',', ' ') for i in aascore]\n",
    "    aascore = [i.split() for i in aascore]\n",
    "    aascore = [[str(int(float(j) * 100)) for j in i] for i in aascore]\n",
    "    df_denovo['aaScore'] = [\" \".join(i) for i in aascore]\n",
    "    aascore = df_denovo['aaScore'].tolist()\n",
    "    scan = df_denovo['PSM_ID'].tolist()\n",
    "    seq = []\n",
    "    for s, aa in zip(sequence, aascore):\n",
    "        # Replace known components in the sequence\n",
    "        s_modified = s.replace('M+15.995', 'm') \\\n",
    "                  .replace('Q+0.984', 'q') \\\n",
    "                  .replace('N+0.984', 'n') \\\n",
    "                  .replace('C+57.021', 'C') \\\n",
    "                  .replace('+42.011', 'g') \\\n",
    "                  .replace('+43.006-17.027', 'd') \\\n",
    "                  .replace('-17.027','e') \\\n",
    "                  .replace('+43.006','f')\n",
    "        # Split the aaScore to check lengths\n",
    "        aa_list = aa.split()\n",
    "        # If the sequence length and aaScore length don't match, replace 'd' with 'fe'\n",
    "        if len(s_modified) != len(aa_list):\n",
    "            s_modified = s_modified.replace('d', 'fe')  # If mismatch, replace 'd' with 'f' and 'e'\n",
    "\n",
    "        # Append the modified sequence\n",
    "        seq.append(s_modified)\n",
    "\n",
    "    # 过滤包含p, d, e, f, g的氨基酸，并同步移除对应位置的aascore\n",
    "    filtered_sequences = []\n",
    "    filtered_aascores = []\n",
    "\n",
    "    excluded_chars = {'p', 'd', 'e', 'f', 'g'}\n",
    "\n",
    "    idx = 0\n",
    "    for s, aa in zip(seq, aascore):\n",
    "        idx += 1\n",
    "        # 将氨基酸分数字符串转换为列表\n",
    "        aa_list = aa.split()\n",
    "    \n",
    "        # 检查序列和分数长度是否一致\n",
    "        if len(s) != len(aa_list):\n",
    "            raise ValueError(f\"序列长度与分数数量不匹配: 序列长度 {len(s)}, 分数数量 {len(aa_list)}\")\n",
    "    \n",
    "        # 同步过滤氨基酸和对应分数\n",
    "        filtered = [(char, score) for char, score in zip(s, aa_list) if char not in excluded_chars]\n",
    "    \n",
    "        # 分离过滤后的字符和分数\n",
    "        filtered_chars, filtered_scores = zip(*filtered) if filtered else ([], [])\n",
    "    \n",
    "        filtered_sequences.append(''.join(filtered_chars))\n",
    "        filtered_aascores.append(' '.join(filtered_scores))\n",
    "\n",
    "    filtered_sequences = [seqe.replace('C','C(+57.02)').replace('m','M(+15.99)').replace('q','Q(+.98)').replace('n','N(+.98)')  for seqe in filtered_sequences]\n",
    "    \n",
    "    DF = np.empty((len(PSM_ID), 5),dtype='object')\n",
    "    DF[:,0] = PSM_ID\n",
    "    DF[:,1] = filtered_sequences\n",
    "    DF[:,3] = Score\n",
    "    DF[:,2] = filtered_aascores\n",
    "    DF[:,4] = 1 \n",
    "    name = ['Spectrum Name','ContraNovo Peptide','ContraNovo aaScore','ContraNovo Score','Area']\n",
    "    df = pd.DataFrame(DF, columns=name)\n",
    "    df['ContraNovo Score'] = df['ContraNovo Score'].astype(int)\n",
    "    df = df[df['ContraNovo Score'] >= SocreAll[tool]]\n",
    "    df.to_csv(contranovo_out,index=False)\n",
    "    subprocess.run(\n",
    "        ('java', '-jar', '../../../ALPS.jar', contranovo_out, str(kmer), str(contigs_ALPS), '>>',\n",
    "        resultdir + 'assembly.log'), stdout=subprocess.DEVNULL)\n",
    "\n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/contranovo/'\n",
    "out_path = '../../../ALPS_Assembly/ContraNovo/'\n",
    "mabs = os.listdir(path)\n",
    "for mab in mabs:\n",
    "    print(mab)\n",
    "    contranovo_path = path + mab +'/'+'denovo.mztab'\n",
    "    resultdir = out_path + mab +'/'\n",
    "    process_contranovo_ALPS(contranovo_path, resultdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65fd68a3-de42-49df-a68c-d4d405c01b62",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "quality_cutoff_ALPS = 50\n",
    "tool = 'DeepNovo'\n",
    "quality_cutoff_ALPS_local = SocreAll[tool]\n",
    "kmer = 7\n",
    "contigs_ALPS = 20\n",
    "\n",
    "def process_deepnovo_ALPS(deepnovo_path, resultdir):\n",
    "    \"\"\"parse de novo results from DeepNoo to dataframe\n",
    "\n",
    "            :param\n",
    "                deepnovo_path: path to the result file of DeepNovo\n",
    "            :return\n",
    "                deepnovo_df: dataframe with Score, Peptide, AAScore of DeepNovo\n",
    "    \"\"\"\n",
    "    deepnovo_out = resultdir + tool + f'_confScoreThreshold_{str(quality_cutoff_ALPS)}_localScore_{str(quality_cutoff_ALPS_local)}.csv'\n",
    "    deepno_df = pd.read_csv(deepnovo_path, sep=\"\t\", header=0)\n",
    "    deepno_df = deepno_df[['scan', 'predicted_score', 'predicted_sequence',  'predicted_position_score']]\n",
    "    deepnovo_df = deepno_df[deepno_df['predicted_sequence'].notna()].copy()\n",
    "\n",
    "    # bug from deepnovo (v.PNAS) which causes result to be wrong aligned\n",
    "\n",
    "    '''\n",
    "    scan    output_score    output_seq  aa_scpre\n",
    "    1       NaN             NaN         NaN\n",
    "    NaN     24              S,E,L       12, 41, 12\n",
    "    2       NaN ...\n",
    "    '''\n",
    "    deepnovo_scan = deepnovo_df['scan'].tolist()\n",
    "    deepnovo_df['scan'] = deepnovo_scan\n",
    "\n",
    "    # change peptide from P,E,P,T,I,D to PEPTID\n",
    "\n",
    "    deepnovo_peptide = deepnovo_df['predicted_sequence'].tolist()\n",
    "    for i in range(len(deepnovo_peptide)):\n",
    "        deepnovo_peptide[i] = str(deepnovo_peptide[i])\n",
    "        deepnovo_peptide[i] = deepnovo_peptide[i].replace(\",\", \"\").replace(\"I\", \"L\").replace(\"Cmod\",\n",
    "            \"C(+57.02)\").replace(\"Mmod\", \"M(+15.99)\").replace(\"Nmod\", \"N(+.98)\").replace(\"Qmod\", \"Q(+.98)\")\n",
    "    deepnovo_df['predicted_sequence'] = deepnovo_peptide\n",
    "    deepnovo_df.columns = ['Spectrum Name','DeepNovo Score', 'DeepNovo Peptide', 'DeepNovo aaScore']\n",
    "\n",
    "    # scale peptide score from 0 to 100\n",
    "\n",
    "    deepnovo_score = deepnovo_df['DeepNovo Score']\n",
    "    deepnovo_df['DeepNovo Score'] = [int(np.exp(i) * 100) if not pd.isna(i) else 0 for i in deepnovo_score]\n",
    "\n",
    "    # scale amino acid score from 0 to 100\n",
    "\n",
    "    deepnovo_aascore = deepnovo_df['DeepNovo aaScore'].tolist()\n",
    "    deepnovo_aascore = [str(i).replace(\",\", \" \") for i in deepnovo_aascore]\n",
    "    deepnovo_aascore = [i.split() for i in deepnovo_aascore]\n",
    "    deepnovo_aascore = [[str(int(np.exp(float(j)) * 100)) if j not in ['nan', 'NaN', '', None] else '' for j in i] for i in deepnovo_aascore]\n",
    "    deepnovo_df['DeepNovo aaScore'] = [\" \".join(i) for i in deepnovo_aascore]\n",
    "    deepnovo_df['Area'] = 1\n",
    "    deepnovo_df['DeepNovo Score'] = deepnovo_df['DeepNovo Score'].astype(int)\n",
    "    deepnovo_df = deepnovo_df[deepnovo_df['DeepNovo Score'] >= SocreAll[tool]]\n",
    "    deepnovo_df = deepnovo_df[['Spectrum Name', 'DeepNovo Peptide', 'DeepNovo aaScore', 'DeepNovo Score', 'Area']]\n",
    "    deepnovo_df.to_csv(deepnovo_out,index=False)\n",
    "    subprocess.run(\n",
    "        ('java', '-jar', '../../../ALPS.jar', deepnovo_out, str(kmer), str(contigs_ALPS), '>>',\n",
    "        resultdir + 'assembly.log'), stdout=subprocess.DEVNULL)\n",
    "\n",
    "path = '../../../denovo/deepnovo2/'\n",
    "out_path = '../../../ALPS_Assembly/DeepNovo/'\n",
    "mabs = os.listdir(path)\n",
    "for mab in mabs:\n",
    "    print(mab)\n",
    "    deepnovo_path = path + mab +'/'+mab+'.deepnovo_denovo'\n",
    "    resultdir = out_path + mab +'/'\n",
    "    process_deepnovo_ALPS(deepnovo_path, resultdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96469906-27b0-4f4e-a8c7-bbe85fef3570",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "quality_cutoff_ALPS = 50\n",
    "tool = 'InstaNovo'\n",
    "quality_cutoff_ALPS_local = SocreAll[tool]\n",
    "kmer = 7\n",
    "contigs_ALPS = 20\n",
    "\n",
    "def process_instanovo_ALPS(instanovo_path,mgf_file,resultdir):\n",
    "    \"\"\"Parse de novo results from AdaNovo to dataframe.\n",
    "            :param\n",
    "                instanovo_path: path to the mztab file of InstaNovo\n",
    "            :return\n",
    "                pnovo_df: dataframe with Score, Peptide, AAScore of InstaNovo\n",
    "    \"\"\"\n",
    "    instanovo_out = resultdir + tool + f'_confScoreThreshold_{str(quality_cutoff_ALPS)}_localScore_{str(quality_cutoff_ALPS_local)}.csv'\n",
    "    titles=[]\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'TITLE=' in line:\n",
    "                titles.append( line.strip().split('TITLE=')[-1] )\n",
    "           \n",
    "    # Load the data into a DataFrame\n",
    "    df_deno = pd.read_csv(instanovo_path)\n",
    "    df_denovo = df_deno[df_deno['predictions'].notna()].copy()\n",
    "\n",
    "    df_denovo['PSM_ID'] = [int(item) for item in df_denovo['scan_number']]\n",
    "    df_denovo['PSM_ID'] = df_denovo['PSM_ID'].apply(lambda x: titles[x] if x < len(titles) else None)\n",
    "\n",
    "    sequence = df_denovo['predictions'].tolist()\n",
    "    PSM_ID = df_denovo['PSM_ID'].tolist()\n",
    "    Score = [int(np.exp(i) * 100) for i in df_denovo['log_probabilities'].tolist()]\n",
    "    aascore = df_denovo['token_log_probabilities'].tolist()\n",
    "    for i in range(len(aascore)):\n",
    "        aascore[i]=aascore[i].replace(',', ' ').replace('[','').replace(']','')\n",
    "    aascore= [i.replace(',', ' ') for i in aascore]\n",
    "    aascore = [i.split() for i in aascore]\n",
    "    #aascore = [[str(int(np.exp(float(j)) * 100)) for j in i] for i in aascore]\n",
    "    aascore = [[str(int(np.exp(float(i[j])) * 100)) for j in range(0, len(i)-1)] for i in aascore]\n",
    "\n",
    "    df_denovo['aaScore'] = [\" \".join(i) for i in aascore]\n",
    "    aascore = df_denovo['aaScore'].tolist()\n",
    "    scan = df_denovo['PSM_ID'].tolist()\n",
    "    seq = [str(i).replace('M[UNIMOD:35]', 'M(+15.99)').replace('Q[UNIMOD:7]', 'Q(+.98)').replace('N[UNIMOD:7]', 'N(+.98)').replace('C[UNIMOD:4]', 'C(+57.02)').replace('I','L') for i in sequence]\n",
    "    DF = np.empty((len(PSM_ID), 4),dtype='object')\n",
    "    DF[:,0] = PSM_ID\n",
    "    DF[:,1] = seq\n",
    "    DF[:,2] = Score\n",
    "    DF[:,3] = aascore \n",
    "\n",
    "    name = ['Spectrum Name','InstaNovo Peptide','InstaNovo Score','InstaNovo aaScore']\n",
    "    df = pd.DataFrame(DF, columns=name)\n",
    "    df.to_csv(instanovo_out,index=False)\n",
    "\n",
    "    df['Area'] = 1\n",
    "    df['InstaNovo Score'] = df['InstaNovo Score'].astype(int)\n",
    "    df = df[df['InstaNovo Score'] >= SocreAll[tool]]\n",
    "    df = df[['Spectrum Name', 'InstaNovo Peptide', 'InstaNovo aaScore', 'InstaNovo Score', 'Area']]\n",
    "    \n",
    "\n",
    "    peps = [i.replace('C(+57.02)','C').replace('M(+15.99)','m').replace('Q(+.98)','q').replace('N(+.98)','n') for i in df['InstaNovo Peptide']]\n",
    "    aas = list(df['InstaNovo aaScore'])\n",
    "    loc = [len(p) == len(a.split()) for p, a in zip(peps, aas)]\n",
    "    df1 = df[loc].copy()\n",
    "    \n",
    "    df1.to_csv(instanovo_out,index=False)\n",
    "    subprocess.run(\n",
    "        ('java', '-jar', '../../../ALPS.jar', instanovo_out, str(kmer), str(contigs_ALPS), '>>',\n",
    "        resultdir + 'assembly.log'), stdout=subprocess.DEVNULL)\n",
    "\n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/instanovo_new/'\n",
    "out_path = '../../../ALPS_Assembly/InstaNovo/'\n",
    "mgf_path = '../../../monoclonal_antibody/'\n",
    "mabs = os.listdir(path)\n",
    "for mab in mabs:\n",
    "    instanovo_path = path + mab +'/'+'predictions_before_refinement.csv'\n",
    "    resultdir = out_path + mab +'/'\n",
    "    mgf_file = mgf_path + mab +'/process1/'+'spectrum_'+ Names[mab] +'_HCD.mgf'\n",
    "    process_instanovo_ALPS(instanovo_path,mgf_file,resultdir)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0958ea35-97da-4c54-9f46-e6af5d80e06c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "quality_cutoff_ALPS = 50\n",
    "tool = 'PointNovo'\n",
    "quality_cutoff_ALPS_local = SocreAll[tool]\n",
    "kmer = 7\n",
    "contigs_ALPS = 20\n",
    "\n",
    "def process_pointnovo_ALPS(pointnovo_path, mgf_file, resultdir):\n",
    "    \"\"\"parse de novo results from PointNovo to dataframe\n",
    "            :param\n",
    "                pointnovo_path: path to the result file of PointNovo\n",
    "            :return\n",
    "                pointnovo_df: dataframe with ScanNum, Score, Peptide, AAScore of PointNovo\n",
    "    \"\"\"\n",
    "\n",
    "    pointnovo_out = resultdir + tool + f'_confScoreThreshold_{str(quality_cutoff_ALPS)}_localScore_{str(quality_cutoff_ALPS_local)}.csv'\n",
    "    titles=[]\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'TITLE=' in line:\n",
    "                titles.append( line.strip().split('TITLE=')[-1] )\n",
    "    pointno_df = pd.read_csv(pointnovo_path, sep=\"\\t\", header=0)\n",
    "    pointno_df['PSM_ID'] = [int(item) for item in pointno_df['feature_id']]\n",
    "    pointno_df['PSM_ID'] = pointno_df['PSM_ID'].apply(lambda x: titles[x-1] if x < len(titles) else None)\n",
    "    \n",
    "    pointno_df = pointno_df[['PSM_ID', 'predicted_sequence', 'predicted_score', 'predicted_position_score']]\n",
    "    pointnovo_df = pointno_df[pointno_df['predicted_sequence'].notna()].copy()\n",
    "\n",
    "    # change peptide from P,E,P,T,I,D to PEPTID\n",
    "\n",
    "    pointnovo_peptide = pointnovo_df['predicted_sequence'].tolist()\n",
    "    for i in range(len(pointnovo_peptide)):\n",
    "        pointnovo_peptide[i] = str(pointnovo_peptide[i])\n",
    "        pointnovo_peptide[i] = pointnovo_peptide[i].replace(\",\", \"\").replace(\"I\", \"L\").replace(\"N(Deamidation)\",\n",
    "        \"N(+.98)\").replace(\"Q(Deamidation)\", \"Q(+.98)\").replace(\"C(Carbamidomethylation)\", \"C(+57.02)\").replace(\"M(Oxidation)\", \"M(+15.99)\")\n",
    "    pointnovo_df['predicted_sequence'] = pointnovo_peptide\n",
    "    pointnovo_df.columns = ['Spectrum Name','PointNovo Peptide', 'PointNovo Score', 'PointNovo aaScore']\n",
    "\n",
    "    # scale peptide score from 0 to 100\n",
    "\n",
    "    pointnovo_score = pointnovo_df['PointNovo Score']\n",
    "    pointnovo_df['PointNovo Score'] = [int(np.exp(i) * 100) if not pd.isna(i) else 0 for i in pointnovo_score]\n",
    "\n",
    "    # scale AAscore from 0 to 100\n",
    "\n",
    "    pointnovo_aascore = pointnovo_df['PointNovo aaScore'].tolist()\n",
    "    pointnovo_aascore = [str(i).replace(\",\", \" \") for i in pointnovo_aascore]\n",
    "    pointnovo_aascore = [i.split() for i in pointnovo_aascore]\n",
    "    pointnovo_aascore = [[str(int(np.exp(float(j)) * 100)) if j not in ['nan', 'NaN', '', None] else '' for j in i] for i in pointnovo_aascore]\n",
    "    pointnovo_df['PointNovo aaScore'] = [\" \".join(i) for i in pointnovo_aascore]\n",
    "\n",
    "    pointnovo_df = pointnovo_df[['Spectrum Name','PointNovo Peptide', 'PointNovo aaScore', 'PointNovo Score']]\n",
    "    pointnovo_df['Area'] = 1\n",
    "\n",
    "    pointnovo_df['PointNovo Score'] = pointnovo_df['PointNovo Score'].astype(int)\n",
    "    pointnovo_df = pointnovo_df[pointnovo_df['PointNovo Score'] >= SocreAll[tool]]\n",
    "    pointnovo_df.to_csv(pointnovo_out,index=False)\n",
    "    subprocess.run(\n",
    "        ('java', '-jar', '../../../ALPS.jar', pointnovo_out, str(kmer), str(contigs_ALPS), '>>',\n",
    "        resultdir + 'assembly.log'), stdout=subprocess.DEVNULL)\n",
    "            \n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/pointnovo/'\n",
    "out_path = '../../../ALPS_Assembly/PointNovo/'\n",
    "mgf_path = '../../../monoclonal_antibody/'\n",
    "mabs = os.listdir(path)\n",
    "for mab in mabs:\n",
    "    pointnovo_path = path + mab +'/'+mab+'.deepnovo_denovo'\n",
    "    resultdir = out_path + mab +'/'\n",
    "    mgf_file = mgf_path + mab +'/process1/'+'spectrum_'+ Names[mab] +'_HCD.mgf'\n",
    "    process_pointnovo_ALPS(pointnovo_path,mgf_file,resultdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ede1c4e9-6a4e-4a52-9246-63ee9d9997b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "quality_cutoff_ALPS = 50\n",
    "tool = 'PGPointNovo'\n",
    "quality_cutoff_ALPS_local = SocreAll[tool]\n",
    "kmer = 7\n",
    "contigs_ALPS = 20\n",
    "\n",
    "def process_pgpointnovo_ALPS(pgpointnovo_path,mgf_file, resultdir):\n",
    "    \"\"\"parse de novo results from PointNovo to dataframe\n",
    "            :param\n",
    "                pgpointnovo_path: path to the result file of PointNovo\n",
    "            :return\n",
    "                pgpointnovo_df: dataframe with ScanNum, Score, Peptide, AAScore of PointNovo\n",
    "    \"\"\"\n",
    "    pgpointnovo_out = resultdir + tool + f'_confScoreThreshold_{str(quality_cutoff_ALPS)}_localScore_{str(quality_cutoff_ALPS_local)}.csv'\n",
    "    titles=[]\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'TITLE=' in line:\n",
    "                titles.append( line.strip().split('TITLE=')[-1] )\n",
    "    pgpointno_df = pd.read_csv(pgpointnovo_path, sep=\"\\t\", header=0)\n",
    "\n",
    "    pgpointno_df['PSM_ID'] = [int(item) for item in pgpointno_df['feature_id']]\n",
    "    pgpointno_df['Spectrum Name'] = pgpointno_df['PSM_ID'].apply(lambda x: titles[x-1] if x < len(titles) else None)\n",
    "    \n",
    "    pgpointno_df = pgpointno_df[['Spectrum Name', 'predicted_sequence', 'predicted_score', 'predicted_position_score']]\n",
    "    pgpointnovo_df = pgpointno_df[pgpointno_df['predicted_sequence'].notna()].copy()\n",
    "\n",
    "    # change peptide from P,E,P,T,I,D to PEPTID\n",
    "\n",
    "    pgpointnovo_peptide = pgpointnovo_df['predicted_sequence'].tolist()\n",
    "    for i in range(len(pgpointnovo_peptide)):\n",
    "        pgpointnovo_peptide[i] = str(pgpointnovo_peptide[i])\n",
    "        pgpointnovo_peptide[i] = pgpointnovo_peptide[i].replace(\",\", \"\").replace(\"I\", \"L\").replace(\"N(Deamidation)\",\n",
    "        \"N(+.98)\").replace(\"Q(Deamidation)\", \"Q(+.98)\").replace(\"C(Carbamidomethylation)\", \"C(+57.02)\").replace(\"M(Oxidation)\", \"M(+15.99)\")\n",
    "    pgpointnovo_df['predicted_sequence'] = pgpointnovo_peptide\n",
    "    pgpointnovo_df.columns = ['Spectrum Name', 'PGPointNovo Peptide', 'PGPointNovo Score', 'PGPointNovo aaScore']\n",
    "\n",
    "    # scale peptide score from 0 to 100\n",
    "\n",
    "    pgpointnovo_score = pgpointnovo_df['PGPointNovo Score']\n",
    "    pgpointnovo_df['PGPointNovo Score'] = [int(np.exp(i) * 100) if not pd.isna(i) else 0 for i in pgpointnovo_score]\n",
    "\n",
    "    # scale AAscore from 0 to 100\n",
    "\n",
    "    pgpointnovo_aascore = pgpointnovo_df['PGPointNovo aaScore'].tolist()\n",
    "    pgpointnovo_aascore = [str(i).replace(\",\", \" \") for i in pgpointnovo_aascore]\n",
    "    pgpointnovo_aascore = [i.split() for i in pgpointnovo_aascore]\n",
    "    pgpointnovo_aascore = [[str(int(np.exp(float(j)) * 100)) if j not in ['nan', 'NaN', '', None] else '' for j in i] for i in pgpointnovo_aascore]\n",
    "\n",
    "    pgpointnovo_df['PGPointNovo aaScore'] = [\" \".join(i) for i in pgpointnovo_aascore]\n",
    "\n",
    "    pgpointnovo_df = pgpointnovo_df[['Spectrum Name','PGPointNovo Peptide', 'PGPointNovo aaScore', 'PGPointNovo Score']]\n",
    "    pgpointnovo_df['Area'] = 1\n",
    "\n",
    "    pgpointnovo_df['PGPointNovo Score'] = pgpointnovo_df['PGPointNovo Score'].astype(int)\n",
    "    pgpointnovo_df = pgpointnovo_df[pgpointnovo_df['PGPointNovo Score'] >= SocreAll[tool]]\n",
    "    pgpointnovo_df.to_csv(pgpointnovo_out,index=False)\n",
    "    subprocess.run(\n",
    "        ('java', '-jar', '../../../ALPS.jar', pgpointnovo_out, str(kmer), str(contigs_ALPS), '>>',\n",
    "        resultdir + 'assembly.log'), stdout=subprocess.DEVNULL)\n",
    "\n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/pgpointnovo/'\n",
    "out_path = '../../../ALPS_Assembly/PGPointNovo/'\n",
    "mgf_path = '../../../monoclonal_antibody/'\n",
    "mabs = os.listdir(path)\n",
    "for mab in mabs:\n",
    "    pgpointnovo_path = path + mab +'/'+mab+'.deepnovo_denovo'\n",
    "    resultdir = out_path + mab +'/'\n",
    "    mgf_file = mgf_path + mab +'/process1/'+'spectrum_'+ Names[mab] +'_HCD.mgf'\n",
    "    process_pgpointnovo_ALPS(pgpointnovo_path,mgf_file,resultdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a760d36-fa4d-49ab-b466-e18574593eba",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "import statistics\n",
    "quality_cutoff_ALPS = 50\n",
    "tool = 'SMSNet'\n",
    "quality_cutoff_ALPS_local = SocreAll[tool]\n",
    "kmer = 7\n",
    "contigs_ALPS = 20\n",
    "\n",
    "def process_smsnet_ALPS(smsnet_path, mgf_file, resultdir):\n",
    "    \"\"\"parse de novo results from SMSNet to dataframe\n",
    "\n",
    "            :param\n",
    "                smsnet_path: path to the result file of SMSNet\n",
    "            :return\n",
    "                smsnet_df: dataframe with Score, Peptide, AAScore of SMSNet\n",
    "        \"\"\"\n",
    "    smsnet_out = resultdir + tool + f'_confScoreThreshold_{str(quality_cutoff_ALPS)}_localScore_{str(quality_cutoff_ALPS_local)}.csv'\n",
    "    titles=[]\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'TITLE=' in line:\n",
    "                titles.append( line.strip().split('TITLE=')[-1] )\n",
    "                \n",
    "    with open(smsnet_path) as f, open(\n",
    "            smsnet_path + '_prob') as g:  # change _rescore and _prob to switch between rescoring and real\n",
    "        smsnet_peptide = pd.Series([line.rstrip() for line in f])\n",
    "        peptide_list = [x.replace(\" \", \"\").replace(\"I\", \"L\") for x in smsnet_peptide]\n",
    "        \n",
    "        smsnet_peptide = pd.DataFrame(peptide_list)\n",
    "        aa_score = g.readlines()\n",
    "        aa_score = [i.strip().split(' ') for i in aa_score]\n",
    "        score_sum = []\n",
    "        for i in range(len(aa_score)):\n",
    "            if not aa_score[i] == ['']:\n",
    "                for j in range(len(aa_score[i])):\n",
    "                    aa_score[i][j] = int(float(np.exp(float(aa_score[i][j])) * 100))\n",
    "            else:\n",
    "                aa_score[i] = [0]\n",
    "        for i in range(len(aa_score)):\n",
    "            if not aa_score[i] == [0]:\n",
    "                score_sum.append(int(statistics.mean(aa_score[i])))\n",
    "            else:\n",
    "                score_sum.append(0)\n",
    "        df = pd.DataFrame({'aaScore': aa_score, 'Peptide Score': score_sum})\n",
    "        smsnet_df = pd.concat([smsnet_peptide, df], axis=1)\n",
    "        smsnet_df['Spectrum Name'] = titles\n",
    "        smsnet_df.columns = ['SMSNet Peptide', 'SMSNet aaScore', 'SMSNet Score','Spectrum Name']\n",
    "        clist = ['Spectrum Name','SMSNet Peptide', 'SMSNet aaScore', 'SMSNet Score']\n",
    "        smsnet_df = smsnet_df[clist]\n",
    "        smsnet_aascore = smsnet_df['SMSNet aaScore'].tolist()\n",
    "        smsnet_df['SMSNet aaScore'] = [str(i).replace(' ', '').replace(',', ' ').replace('[', '').replace(']', '') for i in smsnet_aascore]\n",
    "        smsnet_df['Area'] = 1\n",
    "        smsnet_df['SMSNet Peptide'] = smsnet_df['SMSNet Peptide'].astype(str)\n",
    "        smsnet_df['SMSNet Peptide'] = smsnet_df['SMSNet Peptide'].str.strip()\n",
    "        smsnet_df1 = smsnet_df[smsnet_df['SMSNet Peptide'].str.strip() != ''].copy()\n",
    "        #smsnet_df1 = smsnet_df[smsnet_df['SMSNet Peptide'].notna()].copy()\n",
    "        smsnet_df1 = smsnet_df1[~smsnet_df1['SMSNet Peptide'].str.contains('<unk>|<s>', na=False)]\n",
    "        smsnet_df1['SMSNet Peptide'] = [i.replace('C','C(+57.02)').replace('m','M(+15.99)').replace('n','N(+.98)').replace('q','Q(+.98)') for i in smsnet_df1['SMSNet Peptide']]\n",
    "        smsnet_df1['SMSNet Score'] = smsnet_df1['SMSNet Score'].astype(int)\n",
    "        smsnet_df1 = smsnet_df1[smsnet_df1['SMSNet Score'] >= SocreAll[tool]]\n",
    "        smsnet_df1 = smsnet_df1[smsnet_df1['SMSNet Peptide'].notna()]\n",
    "\n",
    "        smsnet_df1.to_csv(smsnet_out,index=False)\n",
    "        subprocess.run(\n",
    "            ('java', '-jar', '../../../ALPS.jar', smsnet_out, str(kmer), str(contigs_ALPS), '>>',\n",
    "            resultdir + 'assembly.log'), stdout=subprocess.DEVNULL)\n",
    "\n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/SMSNet/'\n",
    "out_path = '../../../ALPS_Assembly/SMSNet/'\n",
    "mgf_path = '../../../monoclonal_antibody/'\n",
    "mabs = os.listdir(path)\n",
    "for mab in mabs:\n",
    "    print(mab)\n",
    "    smsnet_path = path + mab +'/'+'spectrum_'+Names[mab]+'_HCD_forSMSNet'\n",
    "    resultdir = out_path + mab +'/'\n",
    "    mgf_file = mgf_path + mab +'/process1/'+'spectrum_'+ Names[mab] +'_HCD.mgf'\n",
    "    process_smsnet_ALPS(smsnet_path,mgf_file,resultdir)       "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}