{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8198d05e-4859-49ca-b830-b78953aaf348",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "path = '../../../Stitch_Assembly/'\n",
    "os.mkdir(path)\n",
    "tools_list = [\n",
    "    \"AdaNovo\", \"CasanovoV1\", \"CasanovoV2\", \"ContraNovo\", \"DeepNovo\",\n",
    "    \"PepNet\", \"PGPointNovo\", \"pi-HelixNovo\", \"pi-PrimeNovo\", \"pNovo3\", \"PointNovo\", 'SMSNet', 'InstaNovo'\n",
    "]\n",
    "mabs = os.listdir('../../../monoclonal_antibody/')\n",
    "for tool in tools_list:\n",
    "    path1 = path+tool\n",
    "    try:\n",
    "        os.mkdir(path1)\n",
    "    except:\n",
    "        print(path1+'existed')\n",
    "    for mab in mabs:\n",
    "        try:\n",
    "            path2 = path1+'/'+mab\n",
    "            os.mkdir(path2)\n",
    "        except:\n",
    "            print(path2+'existed')\n",
    "\n",
    "df = pd.read_csv('../../data/Tool_Confidence_Threshold_90.csv')\n",
    "SocreAll = pd.Series(df.Confidence_Threshold.values, index=df.Tool).to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fac94558-576b-4390-b851-6fd8685a3284",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "mass_H = 1.0078\n",
    "mass_N_terminus = 1.0078\n",
    "mass_C_terminus = 17.0027\n",
    "mass_AA = {\n",
    "    \"A\": 71.03711,  # 0\n",
    "    \"R\": 156.10111,  # 1\n",
    "    \"N\": 114.04293,  # 2\n",
    "    \"n\": 115.02695,\n",
    "    \"D\": 115.02694,  # 3\n",
    "    \"C\": 160.03065,  # 103.00919,  # 4\n",
    "    \"E\": 129.04259,  # 5\n",
    "    \"Q\": 128.05858,  # 6\n",
    "    \"q\": 129.0426,\n",
    "    \"G\": 57.02146,  # 7\n",
    "    \"H\": 137.05891,  # 8\n",
    "    \"I\": 113.08406,  # 9\n",
    "    \"L\": 113.08406,  # 10\n",
    "    \"K\": 128.09496,  # 11\n",
    "    \"M\": 131.04049,  # 12\n",
    "    \"m\": 147.0354,\n",
    "    \"F\": 147.06841,  # 13\n",
    "    \"P\": 97.05276,  # 14\n",
    "    \"S\": 87.03203,  # 15\n",
    "    \"T\": 101.04768,  # 16\n",
    "    \"W\": 186.07931,  # 17\n",
    "    \"Y\": 163.06333,  # 18\n",
    "    \"V\": 99.06841,  # 19\n",
    "    \"d\": 25.980265,\n",
    "    \"e\": -17.026549,\n",
    "    \"f\": 43.005814,\n",
    "    \"g\": 42.010565,\n",
    "    \"p\": 111.032,\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aea70159-4751-41d7-8a33-16013bebdc38",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "import re\n",
    "quality_cutoff_Stitch = 90\n",
    "tool = 'CasanovoV1'\n",
    "quality_cutoff_Stitch_local = SocreAll[tool]\n",
    "def process_casanovoV1_Stitch(casanovo_path,mgf_file,batchfile_dir, resultdir):\n",
    "    \"\"\"Parse de novo results from Casanovo to dataframe.\n",
    "            :param\n",
    "                casanovo_path: path to the mztab file of Casanovo\n",
    "            :return\n",
    "                pnovo_df: dataframe with Score, Peptide, AAScore of Casanovo\n",
    "    \"\"\"\n",
    "    titles=[]\n",
    "    exp_mass_to_charge = [] \n",
    "    charge = []\n",
    "    RT = []\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'TITLE=' in line:\n",
    "                titles.append( line.strip().split('TITLE=')[-1] )\n",
    "            elif 'PEPMASS=' in line:\n",
    "                exp_mass_to_charge.append(line.strip().split('PEPMASS=')[-1])\n",
    "            elif 'CHARGE=' in line:\n",
    "                charge.append(line.strip().split('CHARGE=')[-1])\n",
    "            elif 'RTINSECONDS=' in line:\n",
    "                RT.append(line.strip().split('RTINSECONDS=')[-1])\n",
    "                \n",
    "    exp_mass_to_charge = [float(i) for i in exp_mass_to_charge]\n",
    "    charge = [int(re.match(r'\\d+', i).group()) for i in charge]    \n",
    "    \n",
    "    casanovo_out = resultdir + tool + '_confScoreThreshold_' + str(quality_cutoff_Stitch) + '_localScore_' + str(quality_cutoff_Stitch_local) + '.csv'\n",
    "    with open(casanovo_path) as f:           \n",
    "        # Identify header\n",
    "        for i, line in enumerate(f):\n",
    "            if line.startswith(\"PSH\"):\n",
    "                casanovo_header = i \n",
    "                break\n",
    "            \n",
    "    # Load the data into a DataFrame\n",
    "    df_deno = pd.read_csv(casanovo_path, sep=\"\\t\", header=casanovo_header)\n",
    "    df_denovo = df_deno[df_deno['sequence'].notna()].copy()\n",
    "    df_denovo['PSM_ID'] = df_denovo['PSM_ID'].astype(int)\n",
    "    df_denovo['exp_mass_to_charge'] = df_denovo['PSM_ID'].apply(lambda x: exp_mass_to_charge[x] if x < len(exp_mass_to_charge) else None)\n",
    "    df_denovo['charge'] = df_denovo['PSM_ID'].apply(lambda x: charge[x] if x < len(charge) else None)\n",
    "    df_denovo['RT'] = df_denovo['PSM_ID'].apply(lambda x: RT[x] if x < len(RT) else None)\n",
    "    df_denovo['PSM_ID'] = df_denovo['PSM_ID'].apply(lambda x: titles[x] if x < len(titles) else None)\n",
    "\n",
    "    sequence = df_denovo['sequence'].tolist()\n",
    "    PSM_ID = df_denovo['PSM_ID'].tolist()\n",
    "    Score = [int((i * 100 + 100)/2) for i in df_denovo['search_engine_score[1]'].tolist()]\n",
    "    aascore = df_denovo['opt_ms_run[1]_aa_scores'].tolist()\n",
    "    for i in range(len(aascore)):\n",
    "        aascore[i]=aascore[i].replace(',', ' ')\n",
    "    aascore= [i.replace(',', ' ') for i in aascore]\n",
    "    aascore = [i.split() for i in aascore]\n",
    "    aascore = [[str(int(float(j) * 100)) for j in i] for i in aascore]\n",
    "    df_denovo['aaScore'] = [\" \".join(i) for i in aascore]\n",
    "    aascore = df_denovo['aaScore'].tolist()\n",
    "    scan = df_denovo['PSM_ID'].tolist()\n",
    "\n",
    "    mode = ['HCD' for x in PSM_ID]\n",
    "    seq = [str(i).replace('M(+15.99)', 'm').replace('Q(+.98)', 'q').replace('N(+.98)', 'n').replace('Q(-17.03)','q').replace(' ','').replace('C(+57.02)', 'C') for i in sequence]\n",
    "    length = [len(x) for x in seq]\n",
    "    Mass_peptide = []\n",
    "    for se in seq:\n",
    "        Mass_peptide.append(sum(mass_AA[aa] for aa in se)+mass_N_terminus + mass_C_terminus)\n",
    "    #Mass_exp = exp_mass_to_charge * charge - charge * mass_H\n",
    "    #ppm = 10**6 * (Mass_exp - Mass_peptide) / Mass_peptide\n",
    "\n",
    "    Mass_exp = [\n",
    "        exp_mass_to_charge[i] * charge[i] - charge[i] * mass_H for i in range(len(exp_mass_to_charge))\n",
    "    ]\n",
    "    ppm = [\n",
    "        10**6 * (Mass_exp[i] - Mass_peptide[i]) / Mass_peptide[i]\n",
    "        for i in range(len(Mass_peptide))\n",
    "    ]\n",
    "    \n",
    "    DF = np.empty((len(PSM_ID), 20),dtype='object')\n",
    "    DF[:,1] = PSM_ID\n",
    "    DF[:,3] = sequence \n",
    "    DF[:,4] = scan \n",
    "    DF[:,5] = length \n",
    "    DF[:,6] = Score \n",
    "    DF[:,7] = Score\n",
    "    DF[:,8] = length\n",
    "    DF[:,9] = df_denovo['exp_mass_to_charge']\n",
    "    DF[:,10] = df_denovo['charge'] \n",
    "    DF[:,11] = df_denovo['RT'] \n",
    "    DF[:,12] = '-' \n",
    "    DF[:,13] = 1\n",
    "    DF[:,14] = Mass_peptide \n",
    "    DF[:,15] = ppm\n",
    "    DF[:,17] = aascore \n",
    "    DF[:,18] = sequence \n",
    "    DF[:,19] = mode \n",
    "    name = ['Fraction','Source File','Feature','Peptide','Scan','Tag Length','Denovo Score','ALC (%)','length','m/z','z','RT','Predict RT','Area','Mass','ppm','PTM','local confidence (%)','tag (>=0%)','mode']\n",
    "    df = pd.DataFrame(DF, columns=name)\n",
    "    df['Denovo Score'] = df['Denovo Score'].astype(int)\n",
    "    df = df[df['Denovo Score'] >= SocreAll[tool]]\n",
    "    df.to_csv(casanovo_out,index=False)\n",
    "\n",
    "    with open(batchfile_dir,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        newfile = resultdir + batchfile_dir.split('/')[-1]\n",
    "        with open(newfile,'w') as fw:\n",
    "            for line in lines:\n",
    "                if 'Path     : ' in line:\n",
    "                    fw.write(line.replace('../datasets/200305_HER_test_04_DENOVO.csv', casanovo_out))\n",
    "                elif 'CutoffALC: ' in line:\n",
    "                    fw.write(line.replace('95', str(quality_cutoff_Stitch_local)))\n",
    "                else:\n",
    "                    fw.write(line)\n",
    "\n",
    "    subprocess.run(\n",
    "    ['../../../stitch-v1.5.0-linux/stitch.bin', 'run', newfile], stdout=subprocess.DEVNULL)\n",
    "    \n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "path = '../../../denovo/casanovo3/'\n",
    "out_path = '../../../Stitch_Assembly/CasanovoV1/'\n",
    "mgf_path = '../../../monoclonal_antibody/'\n",
    "mabs = os.listdir(path)\n",
    "mabs = [mab for mab in mabs if '131ab_EThcd' not in mab and '.yaml' not in mab and 'Herceptin' not in mab and 'anti-FLAG-M2' not in mab]\n",
    "for mab in mabs:\n",
    "    print(mab)\n",
    "    casanovo_path = path + mab +'/'+'casanovo_denovo.mztab'\n",
    "    resultdir = out_path + mab +'/'\n",
    "    mgf_file = mgf_path + mab +'/process1/'+'spectrum_'+ Names[mab] +'_HCD.mgf'\n",
    "    batchfile_dir = '../../../stitch-v1.5.0-linux/batchfiles/' + mab + '.txt'\n",
    "    process_casanovoV1_Stitch(casanovo_path,mgf_file,batchfile_dir,resultdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b17fd49-0561-44dd-bdcb-7fd014ff6f82",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "import re\n",
    "quality_cutoff_Stitch = 90\n",
    "tool = 'CasanovoV2'\n",
    "quality_cutoff_Stitch_local = SocreAll[tool]\n",
    "def process_casanovoV2_Stitch(casanovo_path,mgf_file,batchfile_dir, resultdir):\n",
    "    \"\"\"Parse de novo results from Casanovo to dataframe.\n",
    "            :param\n",
    "                casanovo_path: path to the mztab file of Casanovo\n",
    "            :return\n",
    "                pnovo_df: dataframe with Score, Peptide, AAScore of Casanovo\n",
    "    \"\"\"\n",
    "    titles=[]\n",
    "    exp_mass_to_charge = [] \n",
    "    charge = []\n",
    "    RT = []\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'TITLE=' in line:\n",
    "                titles.append( line.strip().split('TITLE=')[-1] )\n",
    "            elif 'PEPMASS=' in line:\n",
    "                exp_mass_to_charge.append(line.strip().split('PEPMASS=')[-1])\n",
    "            elif 'CHARGE=' in line:\n",
    "                charge.append(line.strip().split('CHARGE=')[-1])\n",
    "            elif 'RTINSECONDS=' in line:\n",
    "                RT.append(line.strip().split('RTINSECONDS=')[-1])\n",
    "                \n",
    "    exp_mass_to_charge = [float(i) for i in exp_mass_to_charge]\n",
    "    charge = [int(re.match(r'\\d+', i).group()) for i in charge]    \n",
    "    \n",
    "    casanovo_out = resultdir + tool + '_confScoreThreshold_' + str(quality_cutoff_Stitch) + '_localScore_' + str(quality_cutoff_Stitch_local) + '.csv'\n",
    "    with open(casanovo_path) as f:           \n",
    "        # Identify header\n",
    "        for i, line in enumerate(f):\n",
    "            if line.startswith(\"PSH\"):\n",
    "                casanovo_header = i \n",
    "                break\n",
    "            \n",
    "    # Load the data into a DataFrame\n",
    "    df_deno = pd.read_csv(casanovo_path, sep=\"\\t\", header=casanovo_header)\n",
    "    df_denovo = df_deno[df_deno['sequence'].notna()].copy()\n",
    "    df_denovo['PSM_ID'] = [int(item.replace('ms_run[1]:index=', '')) for item in df_denovo['spectra_ref']]\n",
    "    df_denovo['exp_mass_to_charge'] = df_denovo['PSM_ID'].apply(lambda x: exp_mass_to_charge[x] if x < len(exp_mass_to_charge) else None)\n",
    "    df_denovo['charge'] = df_denovo['PSM_ID'].apply(lambda x: charge[x] if x < len(charge) else None)\n",
    "    df_denovo['RT'] = df_denovo['PSM_ID'].apply(lambda x: RT[x] if x < len(RT) else None)\n",
    "    df_denovo['PSM_ID'] = df_denovo['PSM_ID'].apply(lambda x: titles[x] if x < len(titles) else None)\n",
    "\n",
    "    sequence = df_denovo['sequence'].tolist()\n",
    "    PSM_ID = df_denovo['PSM_ID'].tolist()\n",
    "    Score = [int((i * 100 + 100)/2) for i in df_denovo['search_engine_score[1]'].tolist()]\n",
    "    aascore = df_denovo['opt_ms_run[1]_aa_scores'].tolist()\n",
    "    for i in range(len(aascore)):\n",
    "        aascore[i]=aascore[i].replace(',', ' ')\n",
    "    aascore= [i.replace(',', ' ') for i in aascore]\n",
    "    aascore = [i.split() for i in aascore]\n",
    "    aascore = [[str(int(float(j) * 100)) for j in i] for i in aascore]\n",
    "    df_denovo['aaScore'] = [\" \".join(i) for i in aascore]\n",
    "    aascore = df_denovo['aaScore'].tolist()\n",
    "    scan = df_denovo['PSM_ID'].tolist()\n",
    "\n",
    "    mode = ['HCD' for x in PSM_ID]\n",
    "    seq = [str(i).replace('M(+15.99)', 'm').replace('Q(+.98)', 'q').replace('N(+.98)', 'n').replace('Q(-17.03)','q').replace(' ','').replace('C(+57.02)', 'C') for i in sequence]\n",
    "    length = [len(x) for x in seq]\n",
    "    Mass_peptide = []\n",
    "    for se in seq:\n",
    "        Mass_peptide.append(sum(mass_AA[aa] for aa in se)+mass_N_terminus + mass_C_terminus)\n",
    "    #Mass_exp = exp_mass_to_charge * charge - charge * mass_H\n",
    "    #ppm = 10**6 * (Mass_exp - Mass_peptide) / Mass_peptide\n",
    "\n",
    "    Mass_exp = [\n",
    "        exp_mass_to_charge[i] * charge[i] - charge[i] * mass_H for i in range(len(exp_mass_to_charge))\n",
    "    ]\n",
    "    ppm = [\n",
    "        10**6 * (Mass_exp[i] - Mass_peptide[i]) / Mass_peptide[i]\n",
    "        for i in range(len(Mass_peptide))\n",
    "    ]\n",
    "    \n",
    "    DF = np.empty((len(PSM_ID), 20),dtype='object')\n",
    "    DF[:,1] = PSM_ID\n",
    "    DF[:,3] = sequence \n",
    "    DF[:,4] = scan \n",
    "    DF[:,5] = length \n",
    "    DF[:,6] = Score \n",
    "    DF[:,7] = Score\n",
    "    DF[:,8] = length\n",
    "    DF[:,9] = df_denovo['exp_mass_to_charge']\n",
    "    DF[:,10] = df_denovo['charge'] \n",
    "    DF[:,11] = df_denovo['RT'] \n",
    "    DF[:,12] = '-' \n",
    "    DF[:,13] = 1\n",
    "    DF[:,14] = Mass_peptide \n",
    "    DF[:,15] = ppm\n",
    "    DF[:,17] = aascore \n",
    "    DF[:,18] = sequence \n",
    "    DF[:,19] = mode \n",
    "    name = ['Fraction','Source File','Feature','Peptide','Scan','Tag Length','Denovo Score','ALC (%)','length','m/z','z','RT','Predict RT','Area','Mass','ppm','PTM','local confidence (%)','tag (>=0%)','mode']\n",
    "    df = pd.DataFrame(DF, columns=name)\n",
    "    df['Denovo Score'] = df['Denovo Score'].astype(int)\n",
    "    df = df[df['Denovo Score'] >= SocreAll[tool]]\n",
    "    df.to_csv(casanovo_out,index=False)\n",
    "\n",
    "    with open(batchfile_dir,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        newfile = resultdir + batchfile_dir.split('/')[-1]\n",
    "        with open(newfile,'w') as fw:\n",
    "            for line in lines:\n",
    "                if 'Path     : ' in line:\n",
    "                    fw.write(line.replace('../datasets/200305_HER_test_04_DENOVO.csv', casanovo_out))\n",
    "                elif 'CutoffALC: ' in line:\n",
    "                    fw.write(line.replace('95', str(quality_cutoff_Stitch_local)))\n",
    "                else:\n",
    "                    fw.write(line)\n",
    "\n",
    "    subprocess.run(\n",
    "    ['../../../stitch-v1.5.0-linux/stitch.bin', 'run', newfile], stdout=subprocess.DEVNULL)\n",
    "    \n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/casanovo4/'\n",
    "out_path = '../../../Stitch_Assembly/CasanovoV2/'\n",
    "mgf_path = '../../../monoclonal_antibody/'\n",
    "mabs = os.listdir(path)\n",
    "mabs = [mab for mab in mabs if '131ab_EThcd' not in mab and '.yaml' not in mab and 'Herceptin' not in mab and 'anti-FLAG-M2' not in mab]\n",
    "for mab in mabs:\n",
    "    print(mab)\n",
    "    casanovo_path = path + mab +'/'+'denovo.mztab'\n",
    "    resultdir = out_path + mab +'/'\n",
    "    mgf_file = mgf_path + mab +'/process1/'+'spectrum_'+ Names[mab] +'_HCD.mgf'\n",
    "    batchfile_dir = '../../../stitch-v1.5.0-linux/batchfiles/' + mab + '.txt'\n",
    "    process_casanovoV2_Stitch(casanovo_path,mgf_file,batchfile_dir,resultdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1661e38b-a0f3-4121-a399-9d4b0f7afa24",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "quality_cutoff_Stitch = 90\n",
    "tool = 'PepNet'\n",
    "quality_cutoff_Stitch_local = SocreAll[tool]\n",
    "\n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/pepnet/'\n",
    "out_path = '../../../Stitch_Assembly/PepNet/'\n",
    "mgf_path = '../../../monoclonal_antibody/'\n",
    "mabs = os.listdir(path)\n",
    "mabs = [mab for mab in mabs if '131ab_EThcd' not in mab and '.yaml' not in mab and 'Herceptin' not in mab and 'anti-FLAG-M2' not in mab and 'results' not in mab]\n",
    "for mab in mabs:\n",
    "    print(mab)\n",
    "    mgf_file = mgf_path + mab +'/process1/'+'spectrum_'+ Names[mab] +'_HCD.mgf'\n",
    "    resultdir= out_path + mab +'/'\n",
    "    pepnet_path = path + mab +'/'+'denovo.tsv'\n",
    "    pepnet_df = pd.read_csv(pepnet_path, sep='\\t')\n",
    "    pepnet_title = pepnet_df['TITLE'].tolist()\n",
    "    pepnet_df['Spectrum Name']=pepnet_title\n",
    "\n",
    "    dat = pd.read_csv(mgf_path + mab +'/process1/'+'spectrum_id_HCD_old.csv')\n",
    "    titles = list(dat['TITLE'])\n",
    "\n",
    "    exp_mass_to_charge = [] \n",
    "    charge = []\n",
    "    RT = []\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'PEPMASS=' in line:\n",
    "                mz = line.strip().split('PEPMASS=')[-1]\n",
    "            elif 'CHARGE=' in line:\n",
    "                z = line.strip().split('CHARGE=')[-1]\n",
    "            elif 'RTINSECONDS=' in line:\n",
    "                rt = line.strip().split('RTINSECONDS=')[-1]\n",
    "                exp_mass_to_charge.append(mz)\n",
    "                charge.append(z)\n",
    "                RT.append(float(rt))\n",
    "                \n",
    "    exp_mass_to_charge = [float(i) for i in exp_mass_to_charge]\n",
    "    charge = [int(re.match(r'\\d+', i).group()) for i in charge] \n",
    "\n",
    "    loc = [titles.index(i) for i in pepnet_df['Spectrum Name']]\n",
    "    exp_mass_to_charge = exp_mass_to_charge[loc]\n",
    "    charge = charge[loc]\n",
    "    RT = RT[loc]\n",
    "\n",
    "    if len(exp_mass_to_charge) != len(pepnet_df['Spectrum Name']):\n",
    "        print('error')\n",
    "    \n",
    "    pepnet_peptide = pepnet_df['DENOVO'].tolist()\n",
    "    Mass_peptide = []\n",
    "    for se in pepnet_peptide:\n",
    "        Mass_peptide.append(sum(mass_AA[aa] for aa in se)+mass_N_terminus + mass_C_terminus)\n",
    "\n",
    "    Mass_exp = [\n",
    "        exp_mass_to_charge[i] * charge[i] - charge[i] * mass_H for i in range(len(exp_mass_to_charge))\n",
    "    ]\n",
    "    ppm = [\n",
    "        10**6 * (Mass_exp[i] - Mass_peptide[i]) / Mass_peptide[i]\n",
    "        for i in range(len(Mass_peptide))\n",
    "    ]   \n",
    "    length = [len(x) for x in pepnet_peptide]\n",
    "    pepnet_df['PepNet Peptide']=[peptide.replace(\"m\", \"M(+15.99)\").replace(\"q\", \"Q(+.98)\").replace(\"n\", \"N(+.98)\").replace(\"C\", \"C(+57.02)\") for peptide in pepnet_peptide]\n",
    "    pepnet_score = [int(i*100) for i in pepnet_df['Score'].tolist()]\n",
    "    aascore = pepnet_df['Positional Score'].tolist()\n",
    "    aascore= [i.replace('[', '').replace(']', '').replace(',', ' ') for i in aascore]\n",
    "    aascore = [i.split() for i in aascore]\n",
    "    aascore = [[str(int(float(j)*100)) for j in i] for i in aascore]\n",
    "    pepnet_df['PepNet aaScore'] = [\" \".join(i) for i in aascore]\n",
    "\n",
    "    DF = np.empty((len(pepnet_df), 20),dtype='object')\n",
    "    DF[:,1] = pepnet_df['Spectrum Name']\n",
    "    DF[:,3] = pepnet_df['PepNet Peptide']\n",
    "    DF[:,4] = pepnet_df['Spectrum Name']\n",
    "    DF[:,5] = length \n",
    "    DF[:,6] = pepnet_score\n",
    "    DF[:,7] = pepnet_score\n",
    "    DF[:,8] = length\n",
    "    DF[:,9] = exp_mass_to_charge\n",
    "    DF[:,10] = charge\n",
    "    DF[:,11] = RT\n",
    "    DF[:,12] = '-' \n",
    "    DF[:,13] = 1\n",
    "    DF[:,14] = Mass_peptide \n",
    "    DF[:,15] = ppm\n",
    "    DF[:,17] = pepnet_df['PepNet aaScore'] \n",
    "    DF[:,18] = pepnet_df['PepNet Peptide'] \n",
    "    DF[:,19] = 'HCD'\n",
    "    name = ['Fraction','Source File','Feature','Peptide','Scan','Tag Length','Denovo Score','ALC (%)','length','m/z','z','RT','Predict RT','Area','Mass','ppm','PTM','local confidence (%)','tag (>=0%)','mode']\n",
    "    df = pd.DataFrame(DF, columns=name)\n",
    "    df['Denovo Score'] = df['Denovo Score'].astype(int)\n",
    "    df = df[df['Denovo Score'] >= SocreAll[tool]]\n",
    "    pepnet_out = resultdir + tool + '_confScoreThreshold_' + str(quality_cutoff_Stitch) + '_localScore_' + str(quality_cutoff_Stitch_local) + '.csv'\n",
    "    df.to_csv(pepnet_out,index=False)\n",
    "    batchfile_dir = '../../../stitch-v1.5.0-linux/batchfiles/' + mab + '.txt'\n",
    "    with open(batchfile_dir,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        newfile = resultdir + batchfile_dir.split('/')[-1]\n",
    "        with open(newfile,'w') as fw:\n",
    "            for line in lines:\n",
    "                if 'Path     : ' in line:\n",
    "                    fw.write(line.replace('../datasets/200305_HER_test_04_DENOVO.csv', pepnet_out))\n",
    "                elif 'CutoffALC: ' in line:\n",
    "                    fw.write(line.replace('95', str(quality_cutoff_Stitch_local)))\n",
    "                else:\n",
    "                    fw.write(line)\n",
    "\n",
    "    subprocess.run(\n",
    "    ['../../../stitch-v1.5.0-linux/stitch.bin', 'run', newfile], stdout=subprocess.DEVNULL) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f68de998-ed0e-4a1a-b16a-69ccbdb78e0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "quality_cutoff_Stitch = 90\n",
    "tool = 'pNovo3'\n",
    "quality_cutoff_Stitch_local = SocreAll[tool]\n",
    "\n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/pNovo3/'\n",
    "out_path = '../../../Stitch_Assembly/pNovo3/'\n",
    "mgf_path = '../../../monoclonal_antibody/'\n",
    "mabs = os.listdir(path)\n",
    "mabs = [mab for mab in mabs if '131ab_EThcd' not in mab and '.yaml' not in mab and 'Herceptin' not in mab and 'anti-FLAG-M2' not in mab]\n",
    "for mab in mabs:\n",
    "    print(mab)\n",
    "    resultdir = out_path + mab +'/'\n",
    "    pnovo_path = path + mab + '/HCD/results.res'\n",
    "    mgf_file = mgf_path + mab +'/process1/'+'spectrum_'+ Names[mab] +'_HCD.mgf'\n",
    "    df = pd.read_csv(pnovo_path, sep=\"\t\", header=None)\n",
    "    df = df[[0, 1, 4, 5]]\n",
    "    df.columns = ['Spectrum Name', 'pNovo3 Peptide', 'pNovo3 Score', 'pNovo3 aaScore']\n",
    "\n",
    "    peps = list(df['pNovo3 Peptide'])\n",
    "    aas = list(df['pNovo3 aaScore'])\n",
    "    loc = [len(p) == len(a.split(',')) for p, a in zip(peps, aas)]\n",
    "    pnovo_df = df[loc].copy()  \n",
    "    pnovo_peptide = pnovo_df['pNovo3 Peptide'].tolist()\n",
    "    pnovo_peptide = [str(i).replace('I','L').replace('a', 'N(+.98)').replace('b', 'Q(+.98)').replace('B', 'Q(+.98)').replace('c', 'M(+15.99)').replace('C','C(+57.02)') for i in pnovo_peptide]\n",
    "    pnovo_df['pNovo3 Peptide']=pnovo_peptide\n",
    "    seq = [i.replace('N(+.98)','n').replace('Q(+.98)','q').replace('M(+15.99)','m').replace('C(+57.02)','C') for i in pnovo_df['pNovo3 Peptide']]\n",
    "    pnovo_score = [int(i) for i in pnovo_df['pNovo3 Score'].tolist()]\n",
    "    pnovo_df['pNovo3 Score']=pnovo_score\n",
    "    aascore = pnovo_df['pNovo3 aaScore'].tolist()\n",
    "    aascore= [i.replace(',', ' ') for i in aascore]\n",
    "    aascore = [i.split() for i in aascore]\n",
    "    aascore = [[str(int(float(j))) for j in i] for i in aascore]\n",
    "    pnovo_df['pNovo3 aaScore'] = [\" \".join(i) for i in aascore]\n",
    "\n",
    "    titles = []\n",
    "    exp_mass_to_charge = [] \n",
    "    charge = []\n",
    "    RT = []\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'TITLE=' in line:\n",
    "                titles.append(line.strip().split('TITLE=')[-1])\n",
    "            elif 'PEPMASS=' in line:\n",
    "                mz = line.strip().split('PEPMASS=')[-1]\n",
    "            elif 'CHARGE=' in line:\n",
    "                z = line.strip().split('CHARGE=')[-1]\n",
    "            elif 'RTINSECONDS=' in line:\n",
    "                rt = line.strip().split('RTINSECONDS=')[-1]\n",
    "                exp_mass_to_charge.append(mz)\n",
    "                charge.append(z)\n",
    "                RT.append(float(rt))\n",
    "\n",
    "    exp_mass_to_charge = [float(i) for i in exp_mass_to_charge]\n",
    "    charge = [int(re.match(r'\\d+', i).group()) for i in charge] \n",
    "\n",
    "    loc = [titles.index(i) for i in pnovo_df['Spectrum Name']]\n",
    "    exp_mass_to_charge = np.array(exp_mass_to_charge)[loc].tolist()\n",
    "    charge = np.array(charge)[loc].tolist()\n",
    "    RT = np.array(RT)[loc].tolist()\n",
    "\n",
    "    if len(exp_mass_to_charge) != len(pnovo_df['Spectrum Name']):\n",
    "        print('error')\n",
    "\n",
    "    Mass_peptide = []\n",
    "    for se in seq:\n",
    "        Mass_peptide.append(sum(mass_AA[aa] for aa in se)+mass_N_terminus + mass_C_terminus)\n",
    "\n",
    "    Mass_exp = [\n",
    "        exp_mass_to_charge[i] * charge[i] - charge[i] * mass_H for i in range(len(exp_mass_to_charge))\n",
    "    ]\n",
    "    ppm = [\n",
    "        10**6 * (Mass_exp[i] - Mass_peptide[i]) / Mass_peptide[i]\n",
    "        for i in range(len(Mass_peptide))\n",
    "    ]   \n",
    "    length = [len(x) for x in seq]\n",
    "\n",
    "    DF = np.empty((len(pnovo_df['Spectrum Name']), 20),dtype='object')\n",
    "    DF[:,1] = pnovo_df['Spectrum Name']\n",
    "    DF[:,3] = pnovo_df['pNovo3 Peptide']\n",
    "    DF[:,4] = pnovo_df['Spectrum Name']\n",
    "    DF[:,5] = length \n",
    "    DF[:,6] = pnovo_df['pNovo3 Score']\n",
    "    DF[:,7] = pnovo_df['pNovo3 Score']\n",
    "    DF[:,8] = length\n",
    "    DF[:,9] = exp_mass_to_charge\n",
    "    DF[:,10] = charge\n",
    "    DF[:,11] = RT\n",
    "    DF[:,12] = '-' \n",
    "    DF[:,13] = 1\n",
    "    DF[:,14] = Mass_peptide \n",
    "    DF[:,15] = ppm\n",
    "    DF[:,17] = pnovo_df['pNovo3 aaScore'] \n",
    "    DF[:,18] = pnovo_df['pNovo3 Peptide'] \n",
    "    DF[:,19] = 'HCD'\n",
    "    name = ['Fraction','Source File','Feature','Peptide','Scan','Tag Length','Denovo Score','ALC (%)','length','m/z','z','RT','Predict RT','Area','Mass','ppm','PTM','local confidence (%)','tag (>=0%)','mode']\n",
    "    df = pd.DataFrame(DF, columns=name)\n",
    "    df['Denovo Score'] = df['Denovo Score'].astype(int)\n",
    "    df = df[df['Denovo Score'] >= SocreAll[tool]]\n",
    "    pnovo_out = resultdir + tool + '_confScoreThreshold_' + str(quality_cutoff_Stitch) + '_localScore_' + str(quality_cutoff_Stitch_local) + '.csv'\n",
    "    df.to_csv(pnovo_out,index=False)\n",
    "\n",
    "    batchfile_dir = '../../../stitch-v1.5.0-linux/batchfiles/' + mab + '.txt'\n",
    "    with open(batchfile_dir,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        newfile = resultdir + batchfile_dir.split('/')[-1]\n",
    "        with open(newfile,'w') as fw:\n",
    "            for line in lines:\n",
    "                if 'Path     : ' in line:\n",
    "                    fw.write(line.replace('../datasets/200305_HER_test_04_DENOVO.csv', pnovo_out))\n",
    "                elif 'CutoffALC: ' in line:\n",
    "                    fw.write(line.replace('95', str(quality_cutoff_Stitch_local)))\n",
    "                else:\n",
    "                    fw.write(line)\n",
    "\n",
    "    subprocess.run(\n",
    "    ['../../../stitch-v1.5.0-linux/stitch.bin', 'run', newfile], stdout=subprocess.DEVNULL) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81dc9291-699c-40a0-b6a8-5fd39b6b2fe2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "quality_cutoff_Stitch = 90\n",
    "tool = 'pi-HelixNovo'\n",
    "quality_cutoff_Stitch_local = SocreAll[tool]\n",
    "\n",
    "def process_pinovo_Stitch(pinovo_path,mgf_file,batchfile_dir,resultdir,mab):\n",
    "    \"\"\"parse de novo results from Pi-HelixNovo to dataframe\n",
    "            :param\n",
    "                pinovo_path: path to the result file of Pi-HelixNovo\n",
    "            :return\n",
    "                pinovo_df: dataframe with Tiltle, Peptide, Score of Pi-HelixNovo\n",
    "    \"\"\"\n",
    "    pinovo_out = resultdir + tool + '_confScoreThreshold_' + str(quality_cutoff_Stitch) + '_localScore_' + str(quality_cutoff_Stitch_local) + '.csv'\n",
    "    pinovo_df = pd.read_csv(pinovo_path, sep=\"\\t\", header=None)\n",
    "    pinovo_df = pinovo_df.dropna()\n",
    "    pinovo_df.columns = [\"Title\", \"sequence\", \"Score\"]\n",
    "    pinovo_title = pinovo_df['Title']\n",
    "    pinovo_df['Title'] = pinovo_title\n",
    "    pinovo_peptide = pinovo_df['sequence'].tolist()\n",
    "    seq = [str(i).replace('M(+15.99)', 'm').replace('Q(+.98)', 'q').replace('N(+.98)', 'n').replace(' ',\n",
    "                                        '').replace('C(+57.02)', 'C') for i in pinovo_peptide]\n",
    "    pinovo_df['pi-HelixNovo Peptide'] = seq        \n",
    "    # scale peptide score from 0 to 100\n",
    "    pinovo_score = pinovo_df['Score']\n",
    "    pinovo_df['pi-HelixNovo Score'] = [int(i * 100) for i in pinovo_score]\n",
    "    pinovo_df['pi-HelixNovo aaScore'] = pinovo_df.apply(lambda row: ' '.join([str(row['pi-HelixNovo Score'])] * len(row['pi-HelixNovo Peptide'])), axis=1)\n",
    "    pinovo_df['pi-HelixNovo Peptide'] = pinovo_peptide\n",
    "    pinovo_df['Spectrum Name'] = pinovo_df['Title']\n",
    "    pinovo_df = pinovo_df[['Spectrum Name', 'pi-HelixNovo Peptide', 'pi-HelixNovo Score', 'pi-HelixNovo aaScore']]\n",
    "\n",
    "    dat = pd.read_csv(mgf_path + mab +'/process1/'+'spectrum_id_HCD_old.csv')\n",
    "    titles = list(dat['TITLE'])\n",
    "\n",
    "    exp_mass_to_charge = [] \n",
    "    charge = []\n",
    "    RT = []\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'PEPMASS=' in line:\n",
    "                mz = line.strip().split('PEPMASS=')[-1]\n",
    "            elif 'CHARGE=' in line:\n",
    "                z = line.strip().split('CHARGE=')[-1]\n",
    "            elif 'RTINSECONDS=' in line:\n",
    "                rt = line.strip().split('RTINSECONDS=')[-1]\n",
    "                exp_mass_to_charge.append(mz)\n",
    "                charge.append(z)\n",
    "                RT.append(float(rt))\n",
    "                \n",
    "    exp_mass_to_charge = [float(i) for i in exp_mass_to_charge]\n",
    "    charge = [int(re.match(r'\\d+', i).group()) for i in charge] \n",
    "\n",
    "    loc = [titles.index(i) for i in pinovo_df['Spectrum Name']]\n",
    "    exp_mass_to_charge = np.array(exp_mass_to_charge)[loc].tolist()\n",
    "    charge = np.array(charge)[loc].tolist()\n",
    "    RT = np.array(RT)[loc].tolist()\n",
    "\n",
    "    if len(exp_mass_to_charge) != len(pinovo_df['Spectrum Name']):\n",
    "        print('error')\n",
    "                \n",
    "\n",
    "    Mass_peptide = []\n",
    "    for se in seq:\n",
    "        Mass_peptide.append(sum(mass_AA[aa] for aa in se)+mass_N_terminus + mass_C_terminus)\n",
    "\n",
    "    Mass_exp = [\n",
    "        exp_mass_to_charge[i] * charge[i] - charge[i] * mass_H for i in range(len(exp_mass_to_charge))\n",
    "    ]\n",
    "    ppm = [\n",
    "        10**6 * (Mass_exp[i] - Mass_peptide[i]) / Mass_peptide[i]\n",
    "        for i in range(len(Mass_peptide))\n",
    "    ]   \n",
    "    length = [len(x) for x in seq]\n",
    "\n",
    "    DF = np.empty((len(pinovo_df['Spectrum Name']), 20),dtype='object')\n",
    "    DF[:,1] = pinovo_df['Spectrum Name']\n",
    "    DF[:,3] = pinovo_df['pi-HelixNovo Peptide']\n",
    "    DF[:,4] = pinovo_df['Spectrum Name']\n",
    "    DF[:,5] = length \n",
    "    DF[:,6] = pinovo_df['pi-HelixNovo Score']\n",
    "    DF[:,7] = pinovo_df['pi-HelixNovo Score']\n",
    "    DF[:,8] = length\n",
    "    DF[:,9] = exp_mass_to_charge\n",
    "    DF[:,10] = charge\n",
    "    DF[:,11] = RT\n",
    "    DF[:,12] = '-' \n",
    "    DF[:,13] = 1\n",
    "    DF[:,14] = Mass_peptide \n",
    "    DF[:,15] = ppm\n",
    "    DF[:,17] = pinovo_df['pi-HelixNovo aaScore'] \n",
    "    DF[:,18] = pinovo_df['pi-HelixNovo Peptide']\n",
    "    DF[:,19] = 'HCD'\n",
    "    name = ['Fraction','Source File','Feature','Peptide','Scan','Tag Length','Denovo Score','ALC (%)','length','m/z','z','RT','Predict RT','Area','Mass','ppm','PTM','local confidence (%)','tag (>=0%)','mode']\n",
    "    df = pd.DataFrame(DF, columns=name)\n",
    "    df['Denovo Score'] = df['Denovo Score'].astype(int)\n",
    "    df = df[df['Denovo Score'] >= SocreAll[tool]]\n",
    "    pinovo_out = resultdir + tool + '_confScoreThreshold_' + str(quality_cutoff_Stitch) + '_localScore_' + str(quality_cutoff_Stitch_local) + '.csv'\n",
    "    df.to_csv(pinovo_out,index=False)\n",
    "\n",
    "    batchfile_dir = '../../../stitch-v1.5.0-linux/batchfiles/' + mab + '.txt'\n",
    "    with open(batchfile_dir,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "    newfile = resultdir + batchfile_dir.split('/')[-1]\n",
    "    with open(newfile,'w') as fw:\n",
    "        for line in lines:\n",
    "            if 'Path     : ' in line:\n",
    "                fw.write(line.replace('../datasets/200305_HER_test_04_DENOVO.csv', pinovo_out))\n",
    "            elif 'CutoffALC: ' in line:\n",
    "                fw.write(line.replace('95', str(quality_cutoff_Stitch_local)))\n",
    "            else:\n",
    "                fw.write(line)\n",
    "\n",
    "    subprocess.run(\n",
    "    ['../../../stitch-v1.5.0-linux/stitch.bin', 'run', newfile], stdout=subprocess.DEVNULL) \n",
    "\n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/pi-helixnovo/'\n",
    "out_path = '../../../Stitch_Assembly/pi-HelixNovo/'\n",
    "mgf_path = '../../../monoclonal_antibody/'\n",
    "mabs = os.listdir(path)\n",
    "mabs = [mab for mab in mabs if '131ab_EThcd' not in mab and '.yaml' not in mab and 'Herceptin' not in mab and 'anti-FLAG-M2' not in mab and '131ab_EThcd' not in mab]\n",
    "for mab in mabs:\n",
    "    print(mab)\n",
    "    pinovo_path = path + mab +'/'+'denovo_denovo.txt'\n",
    "    mgf_file = mgf_path + mab +'/process1/'+'spectrum_'+ Names[mab] +'_HCD.mgf'\n",
    "    batchfile_dir = '../../../stitch-v1.5.0-linux/batchfiles/' + mab + '.txt'\n",
    "    resultdir = out_path + mab +'/'\n",
    "    process_pinovo_Stitch(pinovo_path,mgf_file,batchfile_dir,resultdir,mab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bee5e3e7-d5d5-468f-b655-a0bfefae384e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "quality_cutoff_Stitch = 90\n",
    "tool = 'pi-PrimeNovo'\n",
    "quality_cutoff_Stitch_local = SocreAll[tool]\n",
    "\n",
    "def process_primenovo_Stitch(primenovo_path,mgf_file,batchfile_dir,resultdir,mab):\n",
    "    \"\"\"parse de novo results from pi-PrimeNovo to dataframe\n",
    "            :param\n",
    "                primenovo_path: path to the mztab file of pi-PrimeNovo\n",
    "            :return\n",
    "                pnovo_df: dataframe with Score, Peptide, AAScore of pi-PrimeNovo\n",
    "    \"\"\" \n",
    "    primenovo_out = resultdir + tool + '_confScoreThreshold_' + str(quality_cutoff_Stitch) + '_localScore_' + str(quality_cutoff_Stitch_local) + '.csv'\n",
    "    primenovo_df = pd.read_table(primenovo_path)\n",
    "    primenovo_peptide = primenovo_df['prediction'].tolist()\n",
    "    seq = [str(i).replace('M[+15.995]', 'm').replace('Q[+0.984]', 'q').replace('N[+0.984]', 'n').replace(' ',\n",
    "                                        '').replace('C[+57.021]', 'C').replace('[-17.027]-Q', 'p').replace('-[17.027]-Q', 'p').replace('[+43.006-17.027]-','d').replace('+[43.006-17.027]','d').replace('-[17.027]','e').replace('[17.027]-','e').replace('[-17.027]-','e').replace('[+43.006]-','f').replace('+[43.006]','f').replace('[+42.011]-','g').replace('+[42.011]','g') for i in primenovo_peptide]\n",
    "    \n",
    "    Mass_peptide = []\n",
    "    for se in seq:\n",
    "        Mass_peptide.append(sum(mass_AA[aa] for aa in se)+mass_N_terminus + mass_C_terminus)\n",
    "    seq = [str(i).replace('M[+15.995]', 'm').replace('Q[+0.984]', 'q').replace('N[+0.984]', 'n').replace(' ',\n",
    "                                        '').replace('C[+57.021]', 'C').replace('[-17.027]-Q', 'p').replace('-[17.027]-Q', 'p').replace('[+43.006-17.027]-','').replace('+[43.006-17.027]','').replace('-[17.027]','').replace('[17.027]-','').replace('[-17.027]-','').replace('[+43.006]-','').replace('+[43.006]','').replace('[+42.011]-','').replace('+[42.011]','') for i in primenovo_peptide]\n",
    "    primenovo_df['pi-PrimeNovo Peptide'] = seq\n",
    "    score = primenovo_df['score']\n",
    "    primenovo_df['pi-PrimeNovo Score'] = [int(i * 100) for i in score]\n",
    "    primenovo_df['pi-PrimeNovo aaScore'] = primenovo_df.apply(lambda row: ' '.join([str(row['pi-PrimeNovo Score'])] * len(row['pi-PrimeNovo Peptide'])), axis=1)\n",
    "    primenovo_df['pi-PrimeNovo Peptide'] = [str(i).replace('M[+15.995]', 'M(+15.99)').replace('Q[+0.984]', 'Q(+.98)').replace('N[+0.984]', 'N(+.98)').replace(' ',\n",
    "                    '').replace('C[+57.021]', 'C(+57.02)').replace('[-17.027]-Q', 'Q(-17.03)').replace('-[17.027]-Q', 'Q(-17.03)').replace('[+43.006-17.027]-','').replace('+[43.006-17.027]','').replace('-[17.027]','').replace('[17.027]-','').replace('[-17.027]-','').replace('[+43.006]-','').replace('+[43.006]','').replace('[+42.011]-','').replace('+[42.011]','') for i in primenovo_peptide]\n",
    "    \n",
    "    primenovo_df['Spectrum Name'] = primenovo_df['label']\n",
    "\n",
    "    dat = pd.read_csv(mgf_path + mab +'/process1/'+'spectrum_id_HCD_old.csv')\n",
    "    titles = list(dat['TITLE'])\n",
    "\n",
    "    exp_mass_to_charge = [] \n",
    "    charge = []\n",
    "    RT = []\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'PEPMASS=' in line:\n",
    "                mz = line.strip().split('PEPMASS=')[-1]\n",
    "            elif 'CHARGE=' in line:\n",
    "                z = line.strip().split('CHARGE=')[-1]\n",
    "            elif 'RTINSECONDS=' in line:\n",
    "                rt = line.strip().split('RTINSECONDS=')[-1]\n",
    "                exp_mass_to_charge.append(mz)\n",
    "                charge.append(z)\n",
    "                RT.append(float(rt))\n",
    "                \n",
    "    exp_mass_to_charge = [float(i) for i in exp_mass_to_charge]\n",
    "    charge = [int(re.match(r'\\d+', i).group()) for i in charge] \n",
    "\n",
    "    loc = [titles.index(i) for i in primenovo_df['Spectrum Name']]\n",
    "    exp_mass_to_charge = np.array(exp_mass_to_charge)[loc].tolist()\n",
    "    charge = np.array(charge)[loc].tolist()\n",
    "    RT = np.array(RT)[loc].tolist()\n",
    "\n",
    "    if len(exp_mass_to_charge) != len(primenovo_df['Spectrum Name']):\n",
    "        print('error')\n",
    "                \n",
    "\n",
    "    Mass_peptide = []\n",
    "    for se in seq:\n",
    "        Mass_peptide.append(sum(mass_AA[aa] for aa in se)+mass_N_terminus + mass_C_terminus)\n",
    "\n",
    "    Mass_exp = [\n",
    "        exp_mass_to_charge[i] * charge[i] - charge[i] * mass_H for i in range(len(exp_mass_to_charge))\n",
    "    ]\n",
    "    ppm = [\n",
    "        10**6 * (Mass_exp[i] - Mass_peptide[i]) / Mass_peptide[i]\n",
    "        for i in range(len(Mass_peptide))\n",
    "    ]   \n",
    "    length = [len(x) for x in seq]\n",
    "\n",
    "    DF = np.empty((len(primenovo_df['Spectrum Name']), 20),dtype='object')\n",
    "    DF[:,1] = primenovo_df['Spectrum Name']\n",
    "    DF[:,3] = primenovo_df['pi-PrimeNovo Peptide']\n",
    "    DF[:,4] = primenovo_df['Spectrum Name']\n",
    "    DF[:,5] = length \n",
    "    DF[:,6] = primenovo_df['pi-PrimeNovo Score']\n",
    "    DF[:,7] = primenovo_df['pi-PrimeNovo Score']\n",
    "    DF[:,8] = length\n",
    "    DF[:,9] = exp_mass_to_charge\n",
    "    DF[:,10] = charge\n",
    "    DF[:,11] = RT\n",
    "    DF[:,12] = '-' \n",
    "    DF[:,13] = 1\n",
    "    DF[:,14] = Mass_peptide \n",
    "    DF[:,15] = ppm\n",
    "    DF[:,17] = primenovo_df['pi-PrimeNovo aaScore'] \n",
    "    DF[:,18] = primenovo_df['pi-PrimeNovo Peptide']\n",
    "    DF[:,19] = 'HCD'\n",
    "    name = ['Fraction','Source File','Feature','Peptide','Scan','Tag Length','Denovo Score','ALC (%)','length','m/z','z','RT','Predict RT','Area','Mass','ppm','PTM','local confidence (%)','tag (>=0%)','mode']\n",
    "    df = pd.DataFrame(DF, columns=name)\n",
    "    df['Denovo Score'] = df['Denovo Score'].astype(int)\n",
    "    df = df[df['Denovo Score'] >= SocreAll[tool]]\n",
    "    primenovo_out = resultdir + tool + '_confScoreThreshold_' + str(quality_cutoff_Stitch) + '_localScore_' + str(quality_cutoff_Stitch_local) + '.csv'\n",
    "    df.to_csv(primenovo_out,index=False)\n",
    "\n",
    "    batchfile_dir = '../../../stitch-v1.5.0-linux/batchfiles/' + mab + '.txt'\n",
    "    with open(batchfile_dir,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "    newfile = resultdir + batchfile_dir.split('/')[-1]\n",
    "    with open(newfile,'w') as fw:\n",
    "        for line in lines:\n",
    "            if 'Path     : ' in line:\n",
    "                fw.write(line.replace('../datasets/200305_HER_test_04_DENOVO.csv', primenovo_out))\n",
    "            elif 'CutoffALC: ' in line:\n",
    "                fw.write(line.replace('95', str(quality_cutoff_Stitch_local)))\n",
    "            else:\n",
    "                fw.write(line)\n",
    "\n",
    "    subprocess.run(\n",
    "    ['../../../stitch-v1.5.0-linux/stitch.bin', 'run', newfile], stdout=subprocess.DEVNULL) \n",
    "\n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/pi-primenovo/'\n",
    "out_path = '../../../Stitch_Assembly/pi-PrimeNovo/'\n",
    "mgf_path = '../../../monoclonal_antibody/'\n",
    "mabs = os.listdir(path)\n",
    "mabs = [mab for mab in mabs if '131ab_EThcd' not in mab and '.yaml' not in mab and 'Herceptin' not in mab and 'anti-FLAG-M2' not in mab and '131ab_EThcd' not in mab]\n",
    "for mab in mabs:\n",
    "    print(mab)\n",
    "    primenovo_path = path + mab +'/'+'denovo.tsv'\n",
    "    mgf_file = mgf_path + mab +'/process1/'+'spectrum_'+ Names[mab] +'_HCD.mgf'\n",
    "    resultdir = out_path + mab +'/'\n",
    "    process_primenovo_Stitch(primenovo_path, mgf_file,batchfile_dir,resultdir,mab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df87d3cc-56a3-4f75-b917-97ab44d5cf88",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "quality_cutoff_Stitch = 90\n",
    "tool = 'AdaNovo'\n",
    "quality_cutoff_Stitch_local = SocreAll[tool]\n",
    "\n",
    "def process_adanovo_Stitch(adanovo_path, mgf_file, batchfile_dir, resultdir):\n",
    "    \"\"\"Parse de novo results from AdaNovo to dataframe.\n",
    "            :param\n",
    "                adanovo_path: path to the mztab file of AdaNovo\n",
    "            :return\n",
    "                pnovo_df: dataframe with Score, Peptide, AAScore of AdaNovo\n",
    "    \"\"\"\n",
    "    titles=[]\n",
    "    exp_mass_to_charge = [] \n",
    "    charge = []\n",
    "    RT = []\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'TITLE=' in line:\n",
    "                titles.append( line.strip().split('TITLE=')[-1] )\n",
    "            elif 'PEPMASS=' in line:\n",
    "                exp_mass_to_charge.append(line.strip().split('PEPMASS=')[-1])\n",
    "            elif 'CHARGE=' in line:\n",
    "                charge.append(line.strip().split('CHARGE=')[-1])\n",
    "            elif 'RTINSECONDS=' in line:\n",
    "                RT.append(line.strip().split('RTINSECONDS=')[-1])\n",
    "                \n",
    "    exp_mass_to_charge = [float(i) for i in exp_mass_to_charge]\n",
    "    charge = [int(re.match(r'\\d+', i).group()) for i in charge]  \n",
    "    \n",
    "    adanovo_out = resultdir + tool + '_confScoreThreshold_' + str(quality_cutoff_Stitch) + '_localScore_' + str(quality_cutoff_Stitch_local) + '.csv'\n",
    "    with open(adanovo_path) as f:           \n",
    "        # Identify header\n",
    "        for i, line in enumerate(f):\n",
    "            if line.startswith(\"PSH\"):\n",
    "                adanovo_header = i \n",
    "                break\n",
    "            \n",
    "    # Load the data into a DataFrame\n",
    "    df_deno = pd.read_csv(adanovo_path, sep=\"\\t\", header=adanovo_header)\n",
    "    df_denovo = df_deno[df_deno['sequence'].notna()].copy()\n",
    "\n",
    "    df_denovo['PSM_ID'] = [int(item.replace('ms_run[1]:index=', '')) for item in df_denovo['spectra_ref']]\n",
    "    df_denovo['exp_mass_to_charge'] = df_denovo['PSM_ID'].apply(lambda x: exp_mass_to_charge[x] if x < len(exp_mass_to_charge) else None)\n",
    "    df_denovo['charge'] = df_denovo['PSM_ID'].apply(lambda x: charge[x] if x < len(charge) else None)\n",
    "    df_denovo['RT'] = df_denovo['PSM_ID'].apply(lambda x: RT[x] if x < len(RT) else None)\n",
    "    df_denovo['PSM_ID'] = df_denovo['PSM_ID'].apply(lambda x: titles[x] if x < len(titles) else None)\n",
    "    PSM_ID = df_denovo['PSM_ID'].tolist()\n",
    "    sequence = df_denovo['sequence'].tolist()\n",
    "    Score = [int((i * 100 + 100)/2) for i in df_denovo['search_engine_score[1]'].tolist()]\n",
    "    aascore = df_denovo['opt_ms_run[1]_aa_scores'].tolist()\n",
    "    for i in range(len(aascore)):\n",
    "        aascore[i]=aascore[i].replace(',', ' ')\n",
    "    aascore= [i.replace(',', ' ') for i in aascore]\n",
    "    aascore = [i.split() for i in aascore]\n",
    "    aascore = [[str(int(float(j) * 100)) for j in i] for i in aascore]\n",
    "    df_denovo['aaScore'] = [\" \".join(i) for i in aascore]\n",
    "    aascore = df_denovo['aaScore'].tolist()\n",
    "    scan = df_denovo['PSM_ID'].tolist()\n",
    "\n",
    "    mode = ['HCD' for x in PSM_ID]\n",
    "    seq = [str(i).replace('M(+15.99)', 'm').replace('Q(+.98)', 'q').replace('N(+.98)', 'n').replace('Q(-17.03)','q').replace(' ','').replace('C(+57.02)', 'C') for i in sequence]\n",
    "    length = [len(x) for x in seq]\n",
    "    Mass_peptide = []\n",
    "    for se in seq:\n",
    "        Mass_peptide.append(sum(mass_AA[aa] for aa in se)+mass_N_terminus + mass_C_terminus)\n",
    "    #Mass_exp = exp_mass_to_charge * charge - charge * mass_H\n",
    "    #ppm = 10**6 * (Mass_exp - Mass_peptide) / Mass_peptide\n",
    "\n",
    "    Mass_exp = [\n",
    "        exp_mass_to_charge[i] * charge[i] - charge[i] * mass_H for i in range(len(exp_mass_to_charge))\n",
    "    ]\n",
    "    ppm = [\n",
    "        10**6 * (Mass_exp[i] - Mass_peptide[i]) / Mass_peptide[i]\n",
    "        for i in range(len(Mass_peptide))\n",
    "    ]\n",
    "\n",
    "    DF = np.empty((len(PSM_ID), 20),dtype='object')\n",
    "    DF[:,1] = PSM_ID\n",
    "    DF[:,3] = sequence \n",
    "    DF[:,4] = scan \n",
    "    DF[:,5] = length \n",
    "    DF[:,6] = Score \n",
    "    DF[:,7] = Score\n",
    "    DF[:,8] = length\n",
    "    DF[:,9] = df_denovo['exp_mass_to_charge']\n",
    "    DF[:,10] = df_denovo['charge'] \n",
    "    DF[:,11] = df_denovo['RT'] \n",
    "    DF[:,12] = '-' \n",
    "    DF[:,13] = 1\n",
    "    DF[:,14] = Mass_peptide \n",
    "    DF[:,15] = ppm\n",
    "    DF[:,17] = aascore \n",
    "    DF[:,18] = sequence \n",
    "    DF[:,19] = mode \n",
    "    name = ['Fraction','Source File','Feature','Peptide','Scan','Tag Length','Denovo Score','ALC (%)','length','m/z','z','RT','Predict RT','Area','Mass','ppm','PTM','local confidence (%)','tag (>=0%)','mode']\n",
    "    df = pd.DataFrame(DF, columns=name)\n",
    "    df['Denovo Score'] = df['Denovo Score'].astype(int)\n",
    "    df = df[df['Denovo Score'] >= SocreAll[tool]]\n",
    "    df.to_csv(adanovo_out,index=False)\n",
    "\n",
    "    with open(batchfile_dir,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        newfile = resultdir + batchfile_dir.split('/')[-1]\n",
    "        with open(newfile,'w') as fw:\n",
    "            for line in lines:\n",
    "                if 'Path     : ' in line:\n",
    "                    fw.write(line.replace('../datasets/200305_HER_test_04_DENOVO.csv', adanovo_out))\n",
    "                elif 'CutoffALC: ' in line:\n",
    "                    fw.write(line.replace('95', str(quality_cutoff_Stitch_local)))\n",
    "                else:\n",
    "                    fw.write(line)\n",
    "\n",
    "    subprocess.run(\n",
    "    ['../../../stitch-v1.5.0-linux/stitch.bin', 'run', newfile], stdout=subprocess.DEVNULL)\n",
    "\n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/adanovo/'\n",
    "out_path = '../../../Stitch_Assembly/AdaNovo/'\n",
    "mgf_path = '../../../monoclonal_antibody/'\n",
    "mabs = os.listdir(path)\n",
    "mabs = [mab for mab in mabs if '131ab_EThcd' not in mab and '.yaml' not in mab and 'Herceptin' not in mab and 'anti-FLAG-M2' not in mab]\n",
    "for mab in mabs:\n",
    "    print(mab)\n",
    "    adanovo_path = path + mab +'/'+'denovo.mztab'\n",
    "    resultdir = out_path + mab +'/'\n",
    "    mgf_file = mgf_path + mab +'/process1/'+'spectrum_'+ Names[mab] +'_HCD.mgf'\n",
    "    batchfile_dir = '../../../stitch-v1.5.0-linux/batchfiles/' + mab + '.txt'\n",
    "    process_adanovo_Stitch(adanovo_path,mgf_file,batchfile_dir,resultdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "447635c3-b7c9-4d3b-912b-1455564d6eb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "quality_cutoff_Stitch = 90\n",
    "tool = 'ContraNovo'\n",
    "quality_cutoff_Stitch_local = SocreAll[tool]\n",
    "\n",
    "def process_contranovo_Stitch(contranovo_path,mgf_file,batchfile_dir, resultdir, mab):\n",
    "    \"\"\"Parse de novo results from ContraNovo to dataframe.\n",
    "            :param\n",
    "                contranovo_path: path to the mztab file of ContraNovo\n",
    "            :return\n",
    "                contranovo_df: dataframe with Score, Peptide, AAScore of ContraNovo\n",
    "    \"\"\"\n",
    "    \n",
    "    \n",
    "    contranovo_out = resultdir + tool + '_confScoreThreshold_' + str(quality_cutoff_Stitch) + '_localScore_' + str(quality_cutoff_Stitch_local) + '.csv'\n",
    "    with open(contranovo_path) as f:           \n",
    "        # Identify header\n",
    "        for i, line in enumerate(f):\n",
    "            if line.startswith(\"PSH\"):\n",
    "                contranovo_header = i \n",
    "                break\n",
    "            \n",
    "    # Load the data into a DataFrame\n",
    "    df_deno = pd.read_csv(contranovo_path, sep=\"\\t\", header=contranovo_header)\n",
    "    df_denovo = df_deno[df_deno['sequence'].notna()].copy()\n",
    "    \n",
    "    df_denovo['PSM_ID'] = [i.replace('ms_run[','').replace(']','') for i in df_denovo['spectra_ref']]\n",
    "\n",
    "    sequence = df_denovo['sequence'].tolist()\n",
    "    #sequence = [str(i).replace('M+15.995', 'M(+15.99)').replace('Q+0.984', 'Q(+.98)').replace('N+0.984', 'N(+.98)').replace(' ','').replace('C+57.021', 'C(+57.02)').replace('-17.027Q','').replace('+43.006-17.027','').replace('-17.027','').replace('+43.006','').replace('+42.011','') for i in sequence]     \n",
    "    \n",
    "    PSM_ID = df_denovo['PSM_ID'].tolist()\n",
    "    Score = [int((i * 100 + 100)/2) for i in df_denovo['search_engine_score[1]'].tolist()]\n",
    "    aascore = df_denovo['opt_ms_run[1]_aa_scores'].tolist()\n",
    "    for i in range(len(aascore)):\n",
    "        aascore[i]=aascore[i].replace(',', ' ')\n",
    "    aascore= [i.replace(',', ' ') for i in aascore]\n",
    "    aascore = [i.split() for i in aascore]\n",
    "    aascore = [[str(int(float(j) * 100)) for j in i] for i in aascore]\n",
    "    df_denovo['aaScore'] = [\" \".join(i) for i in aascore]\n",
    "    aascore = df_denovo['aaScore'].tolist()\n",
    "    scan = df_denovo['PSM_ID'].tolist()\n",
    "    seq = []\n",
    "    for s, aa in zip(sequence, aascore):\n",
    "        # Replace known components in the sequence\n",
    "        s_modified = s.replace('M+15.995', 'm') \\\n",
    "                  .replace('Q+0.984', 'q') \\\n",
    "                  .replace('N+0.984', 'n') \\\n",
    "                  .replace('C+57.021', 'C') \\\n",
    "                  .replace('+42.011', 'g') \\\n",
    "                  .replace('+43.006-17.027', 'd') \\\n",
    "                  .replace('-17.027','e') \\\n",
    "                  .replace('+43.006','f')\n",
    "        # Split the aaScore to check lengths\n",
    "        aa_list = aa.split()\n",
    "        # If the sequence length and aaScore length don't match, replace 'd' with 'fe'\n",
    "        if len(s_modified) != len(aa_list):\n",
    "            s_modified = s_modified.replace('d', 'fe')  # If mismatch, replace 'd' with 'f' and 'e'\n",
    "\n",
    "        # Append the modified sequence\n",
    "        seq.append(s_modified)\n",
    "\n",
    "    # p, d, e, f, gaascore\n",
    "    filtered_sequences = []\n",
    "    filtered_aascores = []\n",
    "\n",
    "    excluded_chars = {'p', 'd', 'e', 'f', 'g'}\n",
    "\n",
    "    idx = 0\n",
    "    for s, aa in zip(seq, aascore):\n",
    "        idx += 1\n",
    "        # \n",
    "        aa_list = aa.split()\n",
    "    \n",
    "        # \n",
    "        if len(s) != len(aa_list):\n",
    "            raise ValueError(\":  %d,  %d\" % (len(s), len(aa_list)))\n",
    "    \n",
    "        # \n",
    "        filtered = [(char, score) for char, score in zip(s, aa_list) if char not in excluded_chars]\n",
    "    \n",
    "        # \n",
    "        filtered_chars, filtered_scores = zip(*filtered) if filtered else ([], [])\n",
    "    \n",
    "        filtered_sequences.append(''.join(filtered_chars))\n",
    "        filtered_aascores.append(' '.join(filtered_scores))\n",
    "\n",
    "    length = [len(seqe)  for seqe in filtered_sequences]\n",
    "    seq = [i for i in filtered_sequences]\n",
    "    filtered_sequences = [seqe.replace('C','C(+57.02)').replace('m','M(+15.99)').replace('q','Q(+.98)').replace('n','N(+.98)')  for seqe in filtered_sequences]\n",
    "    \n",
    "    DF = np.empty((len(PSM_ID), 4),dtype='object')\n",
    "    DF[:,0] = PSM_ID\n",
    "    DF[:,1] = filtered_sequences\n",
    "    DF[:,2] = Score\n",
    "    DF[:,3] = filtered_aascores\n",
    "    name = ['Scan', 'Peptide', 'Score', 'aaScore']\n",
    "    df_denovo = pd.DataFrame(DF, columns=name)\n",
    "\n",
    "    dat = pd.read_csv(mgf_path + mab +'/process1/'+'spectrum_id_HCD_old.csv')\n",
    "    titles = list(dat['TITLE'])\n",
    "\n",
    "    exp_mass_to_charge = [] \n",
    "    charge = []\n",
    "    RT = []\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'PEPMASS=' in line:\n",
    "                mz = line.strip().split('PEPMASS=')[-1]\n",
    "            elif 'CHARGE=' in line:\n",
    "                z = line.strip().split('CHARGE=')[-1]\n",
    "            elif 'RTINSECONDS=' in line:\n",
    "                rt = line.strip().split('RTINSECONDS=')[-1]\n",
    "                exp_mass_to_charge.append(mz)\n",
    "                charge.append(z)\n",
    "                RT.append(float(rt))\n",
    "                \n",
    "    exp_mass_to_charge = [float(i) for i in exp_mass_to_charge]\n",
    "    charge = [int(re.match(r'\\d+', i).group()) for i in charge] \n",
    "\n",
    "    loc = [titles.index(i) for i in df_denovo['Scan']]\n",
    "    exp_mass_to_charge = np.array(exp_mass_to_charge)[loc].tolist()\n",
    "    charge = np.array(charge)[loc].tolist()\n",
    "    RT = np.array(RT)[loc].tolist()\n",
    "\n",
    "    if len(exp_mass_to_charge) != len(df_denovo['Scan']):\n",
    "        print('error')\n",
    "\n",
    "    Mass_peptide = []\n",
    "    for se in seq:\n",
    "        Mass_peptide.append(sum(mass_AA[aa] for aa in se)+mass_N_terminus + mass_C_terminus)\n",
    "\n",
    "    Mass_exp = [\n",
    "        exp_mass_to_charge[i] * charge[i] - charge[i] * mass_H for i in range(len(exp_mass_to_charge))\n",
    "    ]\n",
    "    ppm = [\n",
    "        10**6 * (Mass_exp[i] - Mass_peptide[i]) / Mass_peptide[i]\n",
    "        for i in range(len(Mass_peptide))\n",
    "    ]   \n",
    "    length = [len(x) for x in seq]\n",
    "\n",
    "    DF = np.empty((len(df_denovo['Scan']), 20),dtype='object')\n",
    "    DF[:,1] = df_denovo['Scan']\n",
    "    DF[:,3] = df_denovo['Peptide']\n",
    "    DF[:,4] = df_denovo['Scan']\n",
    "    DF[:,5] = length \n",
    "    DF[:,6] = df_denovo['Score']\n",
    "    DF[:,7] = df_denovo['Score']\n",
    "    DF[:,8] = length\n",
    "    DF[:,9] = exp_mass_to_charge\n",
    "    DF[:,10] = charge\n",
    "    DF[:,11] = RT\n",
    "    DF[:,12] = '-' \n",
    "    DF[:,13] = 1\n",
    "    DF[:,14] = Mass_peptide \n",
    "    DF[:,15] = ppm\n",
    "    DF[:,17] = df_denovo['aaScore'] \n",
    "    DF[:,18] = df_denovo['Peptide']\n",
    "    DF[:,19] = 'HCD'\n",
    "    name = ['Fraction','Source File','Feature','Peptide','Scan','Tag Length','Denovo Score','ALC (%)','length','m/z','z','RT','Predict RT','Area','Mass','ppm','PTM','local confidence (%)','tag (>=0%)','mode']\n",
    "    df = pd.DataFrame(DF, columns=name)\n",
    "    df['Denovo Score'] = df['Denovo Score'].astype(int)\n",
    "    df = df[df['Denovo Score'] >= SocreAll[tool]]\n",
    "    df.to_csv(contranovo_out,index=False)\n",
    "\n",
    "    batchfile_dir = '../../../stitch-v1.5.0-linux/batchfiles/' + mab + '.txt'\n",
    "    with open(batchfile_dir,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "    newfile = resultdir + batchfile_dir.split('/')[-1]\n",
    "    with open(newfile,'w') as fw:\n",
    "        for line in lines:\n",
    "            if 'Path     : ' in line:\n",
    "                fw.write(line.replace('../datasets/200305_HER_test_04_DENOVO.csv', contranovo_out))\n",
    "            elif 'CutoffALC: ' in line:\n",
    "                fw.write(line.replace('95', str(quality_cutoff_Stitch_local)))\n",
    "            else:\n",
    "                fw.write(line)\n",
    "\n",
    "    subprocess.run(\n",
    "    ['../../../stitch-v1.5.0-linux/stitch.bin', 'run', newfile], stdout=subprocess.DEVNULL) \n",
    "\n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/contranovo/'\n",
    "out_path = '../../../Stitch_Assembly/ContraNovo/'\n",
    "mgf_path = '../../../monoclonal_antibody/'\n",
    "mabs = os.listdir(path)\n",
    "mabs = [mab for mab in mabs if '131ab_EThcd' not in mab and '.yaml' not in mab and 'Herceptin' not in mab and 'anti-FLAG-M2' not in mab and '131ab_EThcd' not in mab]\n",
    "for mab in mabs:\n",
    "    print(mab)\n",
    "    contranovo_path = path + mab +'/'+'denovo.mztab'\n",
    "    mgf_file = mgf_path + mab +'/process1/'+'spectrum_'+ Names[mab] +'_HCD.mgf'\n",
    "    resultdir = out_path + mab +'/'\n",
    "    process_contranovo_Stitch(contranovo_path,mgf_file,batchfile_dir,resultdir,mab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8863b48c-d3bf-4dca-840a-5c55aa8f38d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "quality_cutoff_Stitch = 90\n",
    "tool = 'DeepNovo'\n",
    "quality_cutoff_Stitch_local = SocreAll[tool]\n",
    "\n",
    "def process_deepnovo_Stitch(deepnovo_path, mgf_file,batchfile_dir,resultdir):\n",
    "    \"\"\"parse de novo results from DeepNoo to dataframe\n",
    "\n",
    "            :param\n",
    "                deepnovo_path: path to the result file of DeepNovo\n",
    "            :return\n",
    "                deepnovo_df: dataframe with Score, Peptide, AAScore of DeepNovo\n",
    "    \"\"\"\n",
    "    titles=[]\n",
    "    exp_mass_to_charge = [] \n",
    "    charge = []\n",
    "    RT = []\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'TITLE=' in line:\n",
    "                titles.append( line.strip().split('TITLE=')[-1] )\n",
    "            elif 'PEPMASS=' in line:\n",
    "                exp_mass_to_charge.append(line.strip().split('PEPMASS=')[-1])\n",
    "            elif 'CHARGE=' in line:\n",
    "                charge.append(line.strip().split('CHARGE=')[-1])\n",
    "            elif 'RTINSECONDS=' in line:\n",
    "                RT.append(line.strip().split('RTINSECONDS=')[-1])\n",
    "                \n",
    "    exp_mass_to_charge = [float(i) for i in exp_mass_to_charge]\n",
    "    charge = [int(re.match(r'\\d+', i).group()) for i in charge]\n",
    "    \n",
    "    deepnovo_out = resultdir + tool + '_confScoreThreshold_' + str(quality_cutoff_Stitch) + '_localScore_' + str(quality_cutoff_Stitch_local) + '.csv'\n",
    "    deepno_df = pd.read_csv(deepnovo_path, sep=\"\t\", header=0)\n",
    "    deepno_df = deepno_df[['scan', 'predicted_score', 'predicted_sequence',  'predicted_position_score']]\n",
    "    deepnovo_df = deepno_df[deepno_df['predicted_sequence'].notna()].copy()\n",
    "\n",
    "    # bug from deepnovo (v.PNAS) which causes result to be wrong aligned\n",
    "\n",
    "    '''\n",
    "    scan    output_score    output_seq  aa_scpre\n",
    "    1       NaN             NaN         NaN\n",
    "    NaN     24              S,E,L       12, 41, 12\n",
    "    2       NaN ...\n",
    "    '''\n",
    "    # change peptide from P,E,P,T,I,D to PEPTID\n",
    "\n",
    "    deepnovo_peptide = deepnovo_df['predicted_sequence'].tolist()\n",
    "    for i in range(len(deepnovo_peptide)):\n",
    "        deepnovo_peptide[i] = str(deepnovo_peptide[i])\n",
    "        deepnovo_peptide[i] = deepnovo_peptide[i].replace(\",\", \"\").replace(\"I\", \"L\").replace(\"Cmod\",\n",
    "            \"C(+57.02)\").replace(\"Mmod\", \"M(+15.99)\").replace(\"Nmod\", \"N(+.98)\").replace(\"Qmod\", \"Q(+.98)\")\n",
    "    deepnovo_df['predicted_sequence'] = deepnovo_peptide\n",
    "    deepnovo_df.columns = ['Spectrum Name','DeepNovo Score', 'DeepNovo Peptide', 'DeepNovo aaScore']\n",
    "\n",
    "    # scale peptide score from 0 to 100\n",
    "\n",
    "    deepnovo_score = deepnovo_df['DeepNovo Score']\n",
    "    deepnovo_df['DeepNovo Score'] = [int(np.exp(i) * 100) if not pd.isna(i) else 0 for i in deepnovo_score]\n",
    "\n",
    "    # scale amino acid score from 0 to 100\n",
    "\n",
    "    deepnovo_aascore = deepnovo_df['DeepNovo aaScore'].tolist()\n",
    "    deepnovo_aascore = [str(i).replace(\",\", \" \") for i in deepnovo_aascore]\n",
    "    deepnovo_aascore = [i.split() for i in deepnovo_aascore]\n",
    "    deepnovo_aascore = [[str(int(np.exp(float(j)) * 100)) if j not in ['nan', 'NaN', '', None] else '' for j in i] for i in deepnovo_aascore]\n",
    "    deepnovo_df['DeepNovo aaScore'] = [\" \".join(i) for i in deepnovo_aascore]\n",
    "    deepnovo_df['DeepNovo Score'] = deepnovo_df['DeepNovo Score'].astype(int)\n",
    "\n",
    "    deepnovo_df['PSM_ID'] = [int(item)-1 for item in deepnovo_df['Spectrum Name']]\n",
    "    deepnovo_df['exp_mass_to_charge'] = deepnovo_df['PSM_ID'].apply(lambda x: exp_mass_to_charge[x] if x < len(exp_mass_to_charge) else None)\n",
    "    exp_mass_to_charge = list(deepnovo_df['exp_mass_to_charge'])\n",
    "    deepnovo_df['charge'] = deepnovo_df['PSM_ID'].apply(lambda x: charge[x] if x < len(charge) else None)\n",
    "    charge = [int(i) for i in deepnovo_df['charge']]\n",
    "    deepnovo_df['RT'] = deepnovo_df['PSM_ID'].apply(lambda x: RT[x] if x < len(RT) else None)\n",
    "    deepnovo_df['PSM_ID'] = deepnovo_df['PSM_ID'].apply(lambda x: titles[x] if x < len(titles) else None)\n",
    "    seq = [i.replace('C(+57.02)','C').replace('N(+.98)','n').replace('Q(+.98)','q').replace('M(+15.99)','m') for i in deepnovo_df['DeepNovo Peptide']]\n",
    "    length = [len(x) for x in seq]\n",
    "    Mass_peptide = []\n",
    "    for se in seq:\n",
    "        Mass_peptide.append(sum(mass_AA[aa] for aa in se)+mass_N_terminus + mass_C_terminus)\n",
    "    #Mass_exp = exp_mass_to_charge * charge - charge * mass_H\n",
    "    #ppm = 10**6 * (Mass_exp - Mass_peptide) / Mass_peptide\n",
    "\n",
    "    Mass_exp = [\n",
    "        exp_mass_to_charge[i] * charge[i] - charge[i] * mass_H for i in range(len(exp_mass_to_charge))\n",
    "    ]\n",
    "    ppm = [\n",
    "        10**6 * (Mass_exp[i] - Mass_peptide[i]) / Mass_peptide[i]\n",
    "        for i in range(len(Mass_peptide))\n",
    "    ]\n",
    "    \n",
    "    DF = np.empty((len(deepnovo_df['PSM_ID']), 20),dtype='object')\n",
    "    DF[:,1] = deepnovo_df['PSM_ID']\n",
    "    DF[:,3] = deepnovo_df['DeepNovo Peptide']\n",
    "    DF[:,4] = deepnovo_df['PSM_ID'] \n",
    "    DF[:,5] = length \n",
    "    DF[:,6] = deepnovo_df['DeepNovo Score'] \n",
    "    DF[:,7] = deepnovo_df['DeepNovo Score'] \n",
    "    DF[:,8] = length\n",
    "    DF[:,9] = exp_mass_to_charge\n",
    "    DF[:,10] = charge\n",
    "    DF[:,11] = deepnovo_df['RT'] \n",
    "    DF[:,12] = '-' \n",
    "    DF[:,13] = 1\n",
    "    DF[:,14] = Mass_peptide \n",
    "    DF[:,15] = ppm\n",
    "    DF[:,17] = deepnovo_df['DeepNovo aaScore'] \n",
    "    DF[:,18] = deepnovo_df['DeepNovo Peptide']\n",
    "    DF[:,19] = 'HCD' \n",
    "    name = ['Fraction','Source File','Feature','Peptide','Scan','Tag Length','Denovo Score','ALC (%)','length','m/z','z','RT','Predict RT','Area','Mass','ppm','PTM','local confidence (%)','tag (>=0%)','mode']\n",
    "    df = pd.DataFrame(DF, columns=name)\n",
    "    df['Denovo Score'] = df['Denovo Score'].astype(int)\n",
    "    df = df[df['Denovo Score'] >= SocreAll[tool]]\n",
    "    df.to_csv(deepnovo_out,index=False)\n",
    "\n",
    "    with open(batchfile_dir,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        newfile = resultdir + batchfile_dir.split('/')[-1]\n",
    "        with open(newfile,'w') as fw:\n",
    "            for line in lines:\n",
    "                if 'Path     : ' in line:\n",
    "                    fw.write(line.replace('../datasets/200305_HER_test_04_DENOVO.csv', deepnovo_out))\n",
    "                elif 'CutoffALC: ' in line:\n",
    "                    fw.write(line.replace('95', str(quality_cutoff_Stitch_local)))\n",
    "                else:\n",
    "                    fw.write(line)\n",
    "\n",
    "    subprocess.run(\n",
    "    ['../../../stitch-v1.5.0-linux/stitch.bin', 'run', newfile], stdout=subprocess.DEVNULL)\n",
    "    \n",
    "path = '../../../denovo/deepnovo2/'\n",
    "out_path = '../../../Stitch_Assembly/DeepNovo/'\n",
    "mabs = os.listdir(path)\n",
    "mabs = [mab for mab in mabs if '131ab_EThcd' not in mab and '.yaml' not in mab and 'Herceptin' not in mab and 'anti-FLAG-M2' not in mab]\n",
    "for mab in mabs:\n",
    "    print(mab)\n",
    "    deepnovo_path = path + mab +'/'+mab+'.deepnovo_denovo'\n",
    "    resultdir = out_path + mab +'/'\n",
    "    mgf_file = mgf_path + mab +'/process1/'+'spectrum_'+ Names[mab] +'_HCD.mgf'\n",
    "    batchfile_dir = '../../../stitch-v1.5.0-linux/batchfiles/' + mab + '.txt'\n",
    "    process_deepnovo_Stitch(deepnovo_path, mgf_file,batchfile_dir,resultdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea6592e3-1f9d-492a-a16d-cd1acd7df612",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "quality_cutoff_Stitch = 90\n",
    "tool = 'InstaNovo'\n",
    "quality_cutoff_Stitch_local = SocreAll[tool]\n",
    "\n",
    "def process_instanovo_Stitch(instanovo_path,mgf_file,batchfile_dir,resultdir):\n",
    "    \"\"\"Parse de novo results from AdaNovo to dataframe.\n",
    "            :param\n",
    "                instanovo_path: path to the mztab file of InstaNovo\n",
    "            :return\n",
    "                pnovo_df: dataframe with Score, Peptide, AAScore of InstaNovo\n",
    "    \"\"\"\n",
    "    titles=[]\n",
    "    exp_mass_to_charge = [] \n",
    "    charge = []\n",
    "    RT = []\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'TITLE=' in line:\n",
    "                titles.append( line.strip().split('TITLE=')[-1] )\n",
    "            elif 'PEPMASS=' in line:\n",
    "                exp_mass_to_charge.append(line.strip().split('PEPMASS=')[-1])\n",
    "            elif 'CHARGE=' in line:\n",
    "                charge.append(line.strip().split('CHARGE=')[-1])\n",
    "            elif 'RTINSECONDS=' in line:\n",
    "                RT.append(line.strip().split('RTINSECONDS=')[-1])\n",
    "                \n",
    "    exp_mass_to_charge = [float(i) for i in exp_mass_to_charge]\n",
    "    charge = [int(re.match(r'\\d+', i).group()) for i in charge]   \n",
    "\n",
    "    instanovo_out = resultdir + tool + '_confScoreThreshold_' + str(quality_cutoff_Stitch) + '_localScore_' + str(quality_cutoff_Stitch_local) + '.csv'\n",
    "           \n",
    "    # Load the data into a DataFrame\n",
    "    df_deno = pd.read_csv(instanovo_path)\n",
    "    df_denovo = df_deno[df_deno['predictions'].notna()].copy()\n",
    "\n",
    "    df_denovo['PSM_ID'] = [int(item) for item in df_denovo['scan_number']]\n",
    "    df_denovo['exp_mass_to_charge'] = df_denovo['PSM_ID'].apply(lambda x: exp_mass_to_charge[x] if x < len(exp_mass_to_charge) else None)\n",
    "    df_denovo['charge'] = df_denovo['PSM_ID'].apply(lambda x: charge[x] if x < len(charge) else None)\n",
    "    df_denovo['RT'] = df_denovo['PSM_ID'].apply(lambda x: RT[x] if x < len(RT) else None)   \n",
    "    df_denovo['PSM_ID'] = df_denovo['PSM_ID'].apply(lambda x: titles[x] if x < len(titles) else None)\n",
    "\n",
    "    sequence = df_denovo['predictions'].tolist()\n",
    "    PSM_ID = df_denovo['PSM_ID'].tolist()\n",
    "    Score = [int(np.exp(i) * 100) for i in df_denovo['log_probabilities'].tolist()]\n",
    "    df_denovo['Score'] = Score\n",
    "    aascore = df_denovo['token_log_probabilities'].tolist()\n",
    "    for i in range(len(aascore)):\n",
    "        aascore[i]=aascore[i].replace(',', ' ').replace('[','').replace(']','')\n",
    "    aascore= [i.replace(',', ' ') for i in aascore]\n",
    "    aascore = [i.split() for i in aascore]\n",
    "    #aascore = [[str(int(np.exp(float(j)) * 100)) for j in i] for i in aascore]\n",
    "    aascore = [[str(int(np.exp(float(i[j])) * 100)) for j in range(0, len(i)-1)] for i in aascore]\n",
    "\n",
    "    df_denovo['aaScore'] = [\" \".join(i) for i in aascore]\n",
    "    aascore = df_denovo['aaScore'].tolist()\n",
    "    scan = df_denovo['PSM_ID'].tolist()\n",
    "    df_denovo['Peptide'] = [str(i).replace('M[UNIMOD:35]', 'M(+15.99)').replace('Q[UNIMOD:7]', 'Q(+.98)').replace('N[UNIMOD:7]', 'N(+.98)').replace('C[UNIMOD:4]', 'C(+57.02)').replace('I','L') for i in sequence]\n",
    "    seq = [str(i).replace('M(+15.99)', 'm').replace('Q(+.98)', 'q').replace('N(+.98)', 'n').replace('Q(-17.03)','q').replace(' ','').replace('C(+57.02)', 'C') for i in df_denovo['Peptide']]\n",
    "    length = [len(x) for x in seq]\n",
    "    Mass_peptide = []\n",
    "    for se in seq:\n",
    "        Mass_peptide.append(sum(mass_AA[aa] for aa in se)+mass_N_terminus + mass_C_terminus)\n",
    "    #Mass_exp = exp_mass_to_charge * charge - charge * mass_H\n",
    "    #ppm = 10**6 * (Mass_exp - Mass_peptide) / Mass_peptide\n",
    "    exp_mass_to_charge = list(df_denovo['exp_mass_to_charge'])\n",
    "    charge = list(df_denovo['charge'])\n",
    "    RT = df_denovo['RT']\n",
    "    Mass_exp = [\n",
    "        exp_mass_to_charge[i] * charge[i] - charge[i] * mass_H for i in range(len(exp_mass_to_charge))\n",
    "    ]\n",
    "    ppm = [\n",
    "        10**6 * (Mass_exp[i] - Mass_peptide[i]) / Mass_peptide[i]\n",
    "        for i in range(len(Mass_peptide))\n",
    "    ]\n",
    "\n",
    "    DF = np.empty((len(scan), 20),dtype='object')\n",
    "    DF[:,1] = scan\n",
    "    DF[:,3] = df_denovo['Peptide']\n",
    "    DF[:,4] = scan \n",
    "    DF[:,5] = length \n",
    "    DF[:,6] = df_denovo['Score']\n",
    "    DF[:,7] = df_denovo['Score']\n",
    "    DF[:,8] = length\n",
    "    DF[:,9] = df_denovo['exp_mass_to_charge']\n",
    "    DF[:,10] = df_denovo['charge'].astype(int)\n",
    "    DF[:,11] = df_denovo['RT'] \n",
    "    DF[:,12] = '-' \n",
    "    DF[:,13] = 1\n",
    "    DF[:,14] = Mass_peptide \n",
    "    DF[:,15] = ppm\n",
    "    DF[:,17] = df_denovo['aaScore']\n",
    "    DF[:,18] = df_denovo['Peptide'] \n",
    "    DF[:,19] = 'HCD'\n",
    "    name = ['Fraction','Source File','Feature','Peptide','Scan','Tag Length','Denovo Score','ALC (%)','length','m/z','z','RT','Predict RT','Area','Mass','ppm','PTM','local confidence (%)','tag (>=0%)','mode']\n",
    "    df = pd.DataFrame(DF, columns=name)\n",
    "    df['Denovo Score'] = df['Denovo Score'].astype(int)\n",
    "    df = df[df['Denovo Score'] >= SocreAll[tool]]\n",
    "    df.to_csv(instanovo_out,index=False)\n",
    "\n",
    "    with open(batchfile_dir,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        newfile = resultdir + batchfile_dir.split('/')[-1]\n",
    "        with open(newfile,'w') as fw:\n",
    "            for line in lines:\n",
    "                if 'Path     : ' in line:\n",
    "                    fw.write(line.replace('../datasets/200305_HER_test_04_DENOVO.csv', instanovo_out))\n",
    "                elif 'CutoffALC: ' in line:\n",
    "                    fw.write(line.replace('95', str(quality_cutoff_Stitch_local)))\n",
    "                else:\n",
    "                    fw.write(line)\n",
    "\n",
    "    subprocess.run(\n",
    "    ['../../../stitch-v1.5.0-linux/stitch.bin', 'run', newfile], stdout=subprocess.DEVNULL)\n",
    "\n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/instanovo_new/'\n",
    "out_path = '../../../Stitch_Assembly/InstaNovo/'\n",
    "mgf_path = '../../../monoclonal_antibody/'\n",
    "mabs = os.listdir(path)\n",
    "mabs = [mab for mab in mabs if '131ab_EThcd' not in mab and '.yaml' not in mab and 'Herceptin' not in mab and 'anti-FLAG-M2' not in mab]\n",
    "for mab in mabs:\n",
    "    print(mab)\n",
    "    instanovo_path = path + mab +'/'+'predictions_before_refinement.csv'\n",
    "    resultdir = out_path + mab +'/'\n",
    "    mgf_file = mgf_path + mab +'/process1/'+'spectrum_'+ Names[mab] +'_HCD.mgf'\n",
    "    batchfile_dir = '../../../stitch-v1.5.0-linux/batchfiles/' + mab + '.txt'\n",
    "    process_instanovo_Stitch(instanovo_path,mgf_file,batchfile_dir,resultdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14faf8eb-42db-4066-ab9c-1c377cd72ba7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "quality_cutoff_Stitch = 90\n",
    "tool = 'PointNovo'\n",
    "quality_cutoff_Stitch_local = SocreAll[tool]\n",
    "\n",
    "def process_pointnovo_Stitch(pointnovo_path, mgf_file, batchfile_dir, resultdir):\n",
    "    \"\"\"parse de novo results from PointNovo to dataframe\n",
    "            :param\n",
    "                pointnovo_path: path to the result file of PointNovo\n",
    "            :return\n",
    "                pointnovo_df: dataframe with ScanNum, Score, Peptide, AAScore of PointNovo\n",
    "    \"\"\"\n",
    "    titles=[]\n",
    "    exp_mass_to_charge = [] \n",
    "    charge = []\n",
    "    RT = []\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'TITLE=' in line:\n",
    "                titles.append( line.strip().split('TITLE=')[-1] )\n",
    "            elif 'PEPMASS=' in line:\n",
    "                exp_mass_to_charge.append(line.strip().split('PEPMASS=')[-1])\n",
    "            elif 'CHARGE=' in line:\n",
    "                charge.append(line.strip().split('CHARGE=')[-1])\n",
    "            elif 'RTINSECONDS=' in line:\n",
    "                RT.append(line.strip().split('RTINSECONDS=')[-1])\n",
    "                \n",
    "    exp_mass_to_charge = [float(i) for i in exp_mass_to_charge]\n",
    "    charge = [int(re.match(r'\\d+', i).group()) for i in charge]  \n",
    "    pointnovo_out = resultdir + tool + '_confScoreThreshold_' + str(quality_cutoff_Stitch) + '_localScore_' + str(quality_cutoff_Stitch_local) + '.csv'\n",
    "    \n",
    "    pointno_df = pd.read_csv(pointnovo_path, sep=\"\\t\", header=0)\n",
    "    pointno_df['PSM_ID'] = [int(item)-1 for item in pointno_df['feature_id']]\n",
    "    pointno_df = pointno_df[['PSM_ID', 'predicted_sequence', 'predicted_score', 'predicted_position_score']]\n",
    "    pointnovo_df = pointno_df[pointno_df['predicted_sequence'].notna()].copy()\n",
    "\n",
    "    # change peptide from P,E,P,T,I,D to PEPTID\n",
    "\n",
    "    pointnovo_peptide = pointnovo_df['predicted_sequence'].tolist()\n",
    "    for i in range(len(pointnovo_peptide)):\n",
    "        pointnovo_peptide[i] = str(pointnovo_peptide[i])\n",
    "        pointnovo_peptide[i] = pointnovo_peptide[i].replace(\",\", \"\").replace(\"I\", \"L\").replace(\"N(Deamidation)\",\n",
    "        \"N(+.98)\").replace(\"Q(Deamidation)\", \"Q(+.98)\").replace(\"C(Carbamidomethylation)\", \"C(+57.02)\").replace(\"M(Oxidation)\", \"M(+15.99)\")\n",
    "    pointnovo_df['predicted_sequence'] = pointnovo_peptide\n",
    "    pointnovo_df.columns = ['PSM_ID','PointNovo Peptide', 'PointNovo Score', 'PointNovo aaScore']\n",
    "\n",
    "    # scale peptide score from 0 to 100\n",
    "\n",
    "    pointnovo_score = pointnovo_df['PointNovo Score']\n",
    "    pointnovo_df['PointNovo Score'] = [int(np.exp(i) * 100) if not pd.isna(i) else 0 for i in pointnovo_score]\n",
    "\n",
    "    # scale AAscore from 0 to 100\n",
    "\n",
    "    pointnovo_aascore = pointnovo_df['PointNovo aaScore'].tolist()\n",
    "    pointnovo_aascore = [str(i).replace(\",\", \" \") for i in pointnovo_aascore]\n",
    "    pointnovo_aascore = [i.split() for i in pointnovo_aascore]\n",
    "    pointnovo_aascore = [[str(int(np.exp(float(j)) * 100)) if j not in ['nan', 'NaN', '', None] else '' for j in i] for i in pointnovo_aascore]\n",
    "    pointnovo_df['PointNovo aaScore'] = [\" \".join(i) for i in pointnovo_aascore]\n",
    "    pointnovo_df = pointnovo_df[['PSM_ID','PointNovo Peptide', 'PointNovo Score', 'PointNovo aaScore']]\n",
    "\n",
    "    pointnovo_df['exp_mass_to_charge'] = pointnovo_df['PSM_ID'].apply(lambda x: exp_mass_to_charge[x] if x < len(exp_mass_to_charge) else None)\n",
    "    exp_mass_to_charge = list(pointnovo_df['exp_mass_to_charge'])\n",
    "    pointnovo_df['charge'] = pointnovo_df['PSM_ID'].apply(lambda x: charge[x] if x < len(charge) else None)\n",
    "    charge = list(pointnovo_df['charge'])\n",
    "    pointnovo_df['RT'] = pointnovo_df['PSM_ID'].apply(lambda x: RT[x] if x < len(RT) else None)\n",
    "    pointnovo_df['PSM_ID'] = pointnovo_df['PSM_ID'].apply(lambda x: titles[x] if x < len(titles) else None)\n",
    "    seq = [i.replace('C(+57.02)','C').replace('N(+.98)','n').replace('Q(+.98)','q').replace('M(+15.99)','m') for i in pointnovo_df['PointNovo Peptide']]\n",
    "    length = [len(x) for x in seq]\n",
    "    Mass_peptide = []\n",
    "    for se in seq:\n",
    "        Mass_peptide.append(sum(mass_AA[aa] for aa in se)+mass_N_terminus + mass_C_terminus)\n",
    "    #Mass_exp = exp_mass_to_charge * charge - charge * mass_H\n",
    "    #ppm = 10**6 * (Mass_exp - Mass_peptide) / Mass_peptide\n",
    "    Mass_exp = [\n",
    "        exp_mass_to_charge[i] * charge[i] - charge[i] * mass_H for i in range(len(exp_mass_to_charge))\n",
    "    ]\n",
    "    ppm = [\n",
    "        10**6 * (Mass_exp[i] - Mass_peptide[i]) / Mass_peptide[i]\n",
    "        for i in range(len(Mass_peptide))\n",
    "    ]\n",
    "\n",
    "    DF = np.empty((len(pointnovo_df['PSM_ID']), 20),dtype='object')\n",
    "    DF[:,1] = pointnovo_df['PSM_ID']\n",
    "    DF[:,3] = pointnovo_df['PointNovo Peptide']\n",
    "    DF[:,4] = pointnovo_df['PSM_ID'] \n",
    "    DF[:,5] = length \n",
    "    DF[:,6] = pointnovo_df['PointNovo Score'] \n",
    "    DF[:,7] = pointnovo_df['PointNovo Score'] \n",
    "    DF[:,8] = length\n",
    "    DF[:,9] = exp_mass_to_charge\n",
    "    DF[:,10] = charge\n",
    "    DF[:,11] = pointnovo_df['RT'] \n",
    "    DF[:,12] = '-' \n",
    "    DF[:,13] = 1\n",
    "    DF[:,14] = Mass_peptide \n",
    "    DF[:,15] = ppm\n",
    "    DF[:,17] = pointnovo_df['PointNovo aaScore'] \n",
    "    DF[:,18] = pointnovo_df['PointNovo Peptide']\n",
    "    DF[:,19] = 'HCD' \n",
    "    name = ['Fraction','Source File','Feature','Peptide','Scan','Tag Length','Denovo Score','ALC (%)','length','m/z','z','RT','Predict RT','Area','Mass','ppm','PTM','local confidence (%)','tag (>=0%)','mode']\n",
    "    df = pd.DataFrame(DF, columns=name)\n",
    "    df['Denovo Score'] = df['Denovo Score'].astype(int)\n",
    "    df = df[df['Denovo Score'] >= SocreAll[tool]]\n",
    "    df.to_csv(pointnovo_out,index=False)\n",
    "\n",
    "    with open(batchfile_dir,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        newfile = resultdir + batchfile_dir.split('/')[-1]\n",
    "        with open(newfile,'w') as fw:\n",
    "            for line in lines:\n",
    "                if 'Path     : ' in line:\n",
    "                    fw.write(line.replace('../datasets/200305_HER_test_04_DENOVO.csv', pointnovo_out))\n",
    "                elif 'CutoffALC: ' in line:\n",
    "                    fw.write(line.replace('95', str(quality_cutoff_Stitch_local)))\n",
    "                else:\n",
    "                    fw.write(line)\n",
    "\n",
    "    subprocess.run(\n",
    "    ['../../../stitch-v1.5.0-linux/stitch.bin', 'run', newfile], stdout=subprocess.DEVNULL)\n",
    "   \n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/pointnovo/'\n",
    "out_path = '../../../Stitch_Assembly/PointNovo/'\n",
    "mgf_path = '../../../monoclonal_antibody/'\n",
    "mabs = os.listdir(path)\n",
    "mabs = [mab for mab in mabs if '131ab_EThcd' not in mab and '.yaml' not in mab and 'Herceptin' not in mab and 'anti-FLAG-M2' not in mab]\n",
    "for mab in mabs:\n",
    "    print(mab)\n",
    "    pointnovo_path = path + mab +'/'+mab+'.deepnovo_denovo'\n",
    "    resultdir = out_path + mab +'/'\n",
    "    mgf_file = mgf_path + mab +'/process1/'+'spectrum_'+ Names[mab] +'_HCD.mgf'\n",
    "    batchfile_dir = '../../../stitch-v1.5.0-linux/batchfiles/' + mab + '.txt'\n",
    "    process_pointnovo_Stitch(pointnovo_path,mgf_file,batchfile_dir,resultdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a43a50c-ed7f-413a-8532-b90565ed219c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "quality_cutoff_Stitch = 90\n",
    "tool = 'PGPointNovo'\n",
    "quality_cutoff_Stitch_local = SocreAll[tool]\n",
    "\n",
    "def process_pgpointnovo_Stitch(pgpointnovo_path,mgf_file,batchfile_dir,resultdir):\n",
    "    \"\"\"parse de novo results from PointNovo to dataframe\n",
    "            :param\n",
    "                pgpointnovo_path: path to the result file of PointNovo\n",
    "            :return\n",
    "                pgpointnovo_df: dataframe with ScanNum, Score, Peptide, AAScore of PointNovo\n",
    "    \"\"\"\n",
    "\n",
    "    titles=[]\n",
    "    exp_mass_to_charge = [] \n",
    "    charge = []\n",
    "    RT = []\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'TITLE=' in line:\n",
    "                titles.append( line.strip().split('TITLE=')[-1] )\n",
    "            elif 'PEPMASS=' in line:\n",
    "                exp_mass_to_charge.append(line.strip().split('PEPMASS=')[-1])\n",
    "            elif 'CHARGE=' in line:\n",
    "                charge.append(line.strip().split('CHARGE=')[-1])\n",
    "            elif 'RTINSECONDS=' in line:\n",
    "                RT.append(line.strip().split('RTINSECONDS=')[-1])\n",
    "                \n",
    "    exp_mass_to_charge = [float(i) for i in exp_mass_to_charge]\n",
    "    charge = [int(re.match(r'\\d+', i).group()) for i in charge]\n",
    "    \n",
    "    pgpointnovo_out = resultdir + tool + '_confScoreThreshold_' + str(quality_cutoff_Stitch) + '_localScore_' + str(quality_cutoff_Stitch_local) + '.csv'\n",
    "    \n",
    "    pgpointno_df = pd.read_csv(pgpointnovo_path, sep=\"\\t\", header=0)\n",
    "\n",
    "    pgpointno_df['PSM_ID'] = [int(item)-1 for item in pgpointno_df['feature_id']]\n",
    "    pgpointno_df = pgpointno_df[['PSM_ID', 'predicted_sequence', 'predicted_score', 'predicted_position_score']]\n",
    "    pgpointnovo_df = pgpointno_df[pgpointno_df['predicted_sequence'].notna()].copy()\n",
    "\n",
    "    # change peptide from P,E,P,T,I,D to PEPTID\n",
    "\n",
    "    pgpointnovo_peptide = pgpointnovo_df['predicted_sequence'].tolist()\n",
    "    for i in range(len(pgpointnovo_peptide)):\n",
    "        pgpointnovo_peptide[i] = str(pgpointnovo_peptide[i])\n",
    "        pgpointnovo_peptide[i] = pgpointnovo_peptide[i].replace(\",\", \"\").replace(\"I\", \"L\").replace(\"N(Deamidation)\",\n",
    "        \"N(+.98)\").replace(\"Q(Deamidation)\", \"Q(+.98)\").replace(\"C(Carbamidomethylation)\", \"C(+57.02)\").replace(\"M(Oxidation)\", \"M(+15.99)\")\n",
    "    pgpointnovo_df['predicted_sequence'] = pgpointnovo_peptide\n",
    "    pgpointnovo_df.columns = ['PSM_ID', 'PGPointNovo Peptide', 'PGPointNovo Score', 'PGPointNovo aaScore']\n",
    "\n",
    "    # scale peptide score from 0 to 100\n",
    "\n",
    "    pgpointnovo_score = pgpointnovo_df['PGPointNovo Score']\n",
    "    pgpointnovo_df['PGPointNovo Score'] = [int(np.exp(i) * 100) if not pd.isna(i) else 0 for i in pgpointnovo_score]\n",
    "\n",
    "    # scale AAscore from 0 to 100\n",
    "\n",
    "    pgpointnovo_aascore = pgpointnovo_df['PGPointNovo aaScore'].tolist()\n",
    "    pgpointnovo_aascore = [str(i).replace(\",\", \" \") for i in pgpointnovo_aascore]\n",
    "    pgpointnovo_aascore = [i.split() for i in pgpointnovo_aascore]\n",
    "    pgpointnovo_aascore = [[str(int(np.exp(float(j)) * 100)) if j not in ['nan', 'NaN', '', None] else '' for j in i] for i in pgpointnovo_aascore]\n",
    "\n",
    "    pgpointnovo_df['PGPointNovo aaScore'] = [\" \".join(i) for i in pgpointnovo_aascore]\n",
    "    pgpointnovo_df['PGPointNovo Score'] = pgpointnovo_df['PGPointNovo Score'].astype(int)\n",
    "\n",
    "    pgpointnovo_df['exp_mass_to_charge'] = pgpointnovo_df['PSM_ID'].apply(lambda x: exp_mass_to_charge[x] if x < len(exp_mass_to_charge) else None)\n",
    "    exp_mass_to_charge = list(pgpointnovo_df['exp_mass_to_charge'])\n",
    "    pgpointnovo_df['charge'] = pgpointnovo_df['PSM_ID'].apply(lambda x: charge[x] if x < len(charge) else None)\n",
    "    charge = list(pgpointnovo_df['charge'])\n",
    "    pgpointnovo_df['RT'] = pgpointnovo_df['PSM_ID'].apply(lambda x: RT[x] if x < len(RT) else None)\n",
    "    pgpointnovo_df['PSM_ID'] = pgpointnovo_df['PSM_ID'].apply(lambda x: titles[x] if x < len(titles) else None)\n",
    "    seq = [i.replace('C(+57.02)','C').replace('N(+.98)','n').replace('Q(+.98)','q').replace('M(+15.99)','m') for i in pgpointnovo_df['PGPointNovo Peptide']]\n",
    "    length = [len(x) for x in seq]\n",
    "    Mass_peptide = []\n",
    "    for se in seq:\n",
    "        Mass_peptide.append(sum(mass_AA[aa] for aa in se)+mass_N_terminus + mass_C_terminus)\n",
    "    #Mass_exp = exp_mass_to_charge * charge - charge * mass_H\n",
    "    #ppm = 10**6 * (Mass_exp - Mass_peptide) / Mass_peptide\n",
    "    Mass_exp = [\n",
    "        exp_mass_to_charge[i] * charge[i] - charge[i] * mass_H for i in range(len(exp_mass_to_charge))\n",
    "    ]\n",
    "    ppm = [\n",
    "        10**6 * (Mass_exp[i] - Mass_peptide[i]) / Mass_peptide[i]\n",
    "        for i in range(len(Mass_peptide))\n",
    "    ]\n",
    "\n",
    "    DF = np.empty((len(pgpointnovo_df['PSM_ID']), 20),dtype='object')\n",
    "    DF[:,1] = pgpointnovo_df['PSM_ID']\n",
    "    DF[:,3] = pgpointnovo_df['PGPointNovo Peptide']\n",
    "    DF[:,4] = pgpointnovo_df['PSM_ID'] \n",
    "    DF[:,5] = length \n",
    "    DF[:,6] = pgpointnovo_df['PGPointNovo Score'] \n",
    "    DF[:,7] = pgpointnovo_df['PGPointNovo Score'] \n",
    "    DF[:,8] = length\n",
    "    DF[:,9] = exp_mass_to_charge\n",
    "    DF[:,10] = charge\n",
    "    DF[:,11] = pgpointnovo_df['RT'] \n",
    "    DF[:,12] = '-' \n",
    "    DF[:,13] = 1\n",
    "    DF[:,14] = Mass_peptide \n",
    "    DF[:,15] = ppm\n",
    "    DF[:,17] = pgpointnovo_df['PGPointNovo aaScore'] \n",
    "    DF[:,18] = pgpointnovo_df['PGPointNovo Peptide']\n",
    "    DF[:,19] = 'HCD' \n",
    "    name = ['Fraction','Source File','Feature','Peptide','Scan','Tag Length','Denovo Score','ALC (%)','length','m/z','z','RT','Predict RT','Area','Mass','ppm','PTM','local confidence (%)','tag (>=0%)','mode']\n",
    "    df = pd.DataFrame(DF, columns=name)\n",
    "    df['Denovo Score'] = df['Denovo Score'].astype(int)\n",
    "    df = df[df['Denovo Score'] >= SocreAll[tool]]\n",
    "    df.to_csv(pgpointnovo_out,index=False)\n",
    "\n",
    "    with open(batchfile_dir,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        newfile = resultdir + batchfile_dir.split('/')[-1]\n",
    "        with open(newfile,'w') as fw:\n",
    "            for line in lines:\n",
    "                if 'Path     : ' in line:\n",
    "                    fw.write(line.replace('../datasets/200305_HER_test_04_DENOVO.csv', pgpointnovo_out))\n",
    "                elif 'CutoffALC: ' in line:\n",
    "                    fw.write(line.replace('95', str(quality_cutoff_Stitch_local)))\n",
    "                else:\n",
    "                    fw.write(line)\n",
    "\n",
    "    subprocess.run(\n",
    "    ['../../../stitch-v1.5.0-linux/stitch.bin', 'run', newfile], stdout=subprocess.DEVNULL)\n",
    "\n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/pgpointnovo/'\n",
    "out_path = '../../../Stitch_Assembly/PGPointNovo/'\n",
    "mgf_path = '../../../monoclonal_antibody/'\n",
    "mabs = os.listdir(path)\n",
    "mabs = [mab for mab in mabs if '131ab_EThcd' not in mab and '.yaml' not in mab and 'Herceptin' not in mab and 'anti-FLAG-M2' not in mab]\n",
    "for mab in mabs:\n",
    "    print(mab)\n",
    "    pgpointnovo_path = path + mab +'/'+mab+'.deepnovo_denovo'\n",
    "    resultdir = out_path + mab +'/'\n",
    "    mgf_file = mgf_path + mab +'/process1/'+'spectrum_'+ Names[mab] +'_HCD.mgf'\n",
    "    batchfile_dir = '../../../stitch-v1.5.0-linux/batchfiles/' + mab + '.txt'\n",
    "    process_pgpointnovo_Stitch(pgpointnovo_path,mgf_file,batchfile_dir,resultdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e282ffc-4107-4c1d-a5ba-a49e5fab5dd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess \n",
    "import statistics\n",
    "\n",
    "quality_cutoff_Stitch = 90\n",
    "tool = 'SMSNet'\n",
    "quality_cutoff_Stitch_local = SocreAll[tool]\n",
    "\n",
    "def process_smsnet_Stitch(smsnet_path, mgf_file, batchfile_dir, resultdir):\n",
    "    \"\"\"parse de novo results from SMSNet to dataframe\n",
    "\n",
    "            :param\n",
    "                smsnet_path: path to the result file of SMSNet\n",
    "            :return\n",
    "                smsnet_df: dataframe with Score, Peptide, AAScore of SMSNet\n",
    "        \"\"\"\n",
    "    titles=[]\n",
    "    exp_mass_to_charge = [] \n",
    "    charge = []\n",
    "    RT = []\n",
    "    with open(mgf_file,'r') as fr:\n",
    "        lines = fr.readlines()\n",
    "        for line in lines:\n",
    "            if 'TITLE=' in line:\n",
    "                titles.append( line.strip().split('TITLE=')[-1] )\n",
    "            elif 'PEPMASS=' in line:\n",
    "                exp_mass_to_charge.append(line.strip().split('PEPMASS=')[-1])\n",
    "            elif 'CHARGE=' in line:\n",
    "                charge.append(line.strip().split('CHARGE=')[-1])\n",
    "            elif 'RTINSECONDS=' in line:\n",
    "                RT.append(line.strip().split('RTINSECONDS=')[-1])\n",
    "                \n",
    "    exp_mass_to_charge = [float(i) for i in exp_mass_to_charge]\n",
    "    charge = [int(re.match(r'\\d+', i).group()) for i in charge] \n",
    "\n",
    "    smsnet_out = resultdir + tool + '_confScoreThreshold_' + str(quality_cutoff_Stitch) + '_localScore_' + str(quality_cutoff_Stitch_local) + '.csv'\n",
    "                \n",
    "    with open(smsnet_path) as f, open(\n",
    "            smsnet_path + '_prob') as g:  # change _rescore and _prob to switch between rescoring and real\n",
    "        smsnet_peptide = pd.Series([line.rstrip() for line in f])\n",
    "        peptide_list = [x.replace(\" \", \"\").replace(\"I\", \"L\") for x in smsnet_peptide]\n",
    "        \n",
    "        smsnet_peptide = pd.DataFrame(peptide_list)\n",
    "        aa_score = g.readlines()\n",
    "        aa_score = [i.strip().split(' ') for i in aa_score]\n",
    "        score_sum = []\n",
    "        for i in range(len(aa_score)):\n",
    "            if not aa_score[i] == ['']:\n",
    "                for j in range(len(aa_score[i])):\n",
    "                    aa_score[i][j] = int(float(np.exp(float(aa_score[i][j])) * 100))\n",
    "            else:\n",
    "                aa_score[i] = [0]\n",
    "        for i in range(len(aa_score)):\n",
    "            if not aa_score[i] == [0]:\n",
    "                #score_sum.append(int(statistics.mean(aa_score[i]))) #python3\n",
    "                score_sum.append(int(np.mean(aa_score[i])))\n",
    "            else:\n",
    "                score_sum.append(0)\n",
    "        df = pd.DataFrame({'aaScore': aa_score, 'Peptide Score': score_sum})\n",
    "        smsnet_df = pd.concat([smsnet_peptide, df], axis=1)\n",
    "        smsnet_df['Spectrum Name'] = titles\n",
    "        smsnet_df.columns = ['Peptide', 'aaScore', 'Score','Scan']\n",
    "  \n",
    "        smsnet_aascore = smsnet_df['aaScore'].tolist()\n",
    "        smsnet_df['aaScore'] = [str(i).replace(' ', '').replace(',', ' ').replace('[', '').replace(']', '') for i in smsnet_aascore]\n",
    "        smsnet_df['Peptide'] = smsnet_df['Peptide'].astype(str)\n",
    "        smsnet_df['Peptide'] = smsnet_df['Peptide'].str.strip()\n",
    "\n",
    "        smsnet_df['exp_mass_to_charge'] = exp_mass_to_charge\n",
    "        smsnet_df['charge'] = charge\n",
    "        smsnet_df['RT'] = RT\n",
    "                \n",
    "        smsnet_df1 = smsnet_df[smsnet_df['Peptide'].str.strip() != ''].copy()\n",
    "        #smsnet_df1 = smsnet_df[smsnet_df['SMSNet Peptide'].notna()].copy()\n",
    "        smsnet_df1 = smsnet_df1[~smsnet_df1['Peptide'].str.contains('<unk>|<s>', na=False)]\n",
    "        smsnet_df1['Peptide'] = [i.replace('C','C(+57.02)').replace('m','M(+15.99)').replace('n','N(+.98)').replace('q','Q(+.98)') for i in smsnet_df1['Peptide']]\n",
    "        seq = [i.replace('C(+57.02)','C').replace('N(+.98)','n').replace('Q(+.98)','q').replace('M(+15.99)','m') for i in smsnet_df1['Peptide']]\n",
    "        length = [len(x) for x in seq]\n",
    "        smsnet_df1 = smsnet_df1.reset_index(drop=True)\n",
    "        Mass_peptide = []\n",
    "        for se in seq:\n",
    "            Mass_peptide.append(sum(mass_AA[aa] for aa in se)+mass_N_terminus + mass_C_terminus)\n",
    "        #Mass_exp = exp_mass_to_charge * charge - charge * mass_H\n",
    "        #ppm = 10**6 * (Mass_exp - Mass_peptide) / Mass_peptide\n",
    "        exp_mass_to_charge = smsnet_df1['exp_mass_to_charge'].astype(float)\n",
    "        charge = smsnet_df1['charge'].astype(int)\n",
    "        Mass_exp = [\n",
    "            exp_mass_to_charge[i] * charge[i] - charge[i] * mass_H for i in range(len(exp_mass_to_charge))\n",
    "        ]\n",
    "        ppm = [\n",
    "            10**6 * (Mass_exp[i] - Mass_peptide[i]) / Mass_peptide[i]\n",
    "            for i in range(len(Mass_peptide))\n",
    "        ]\n",
    "\n",
    "        DF = np.empty((len(smsnet_df1['Scan']), 20),dtype='object')\n",
    "        DF[:,1] = smsnet_df1['Scan']\n",
    "        DF[:,3] = smsnet_df1['Peptide']\n",
    "        DF[:,4] = smsnet_df1['Scan']\n",
    "        DF[:,5] = length\n",
    "        DF[:,6] = smsnet_df1['Score'] \n",
    "        DF[:,7] = smsnet_df1['Score'] \n",
    "        DF[:,8] = length\n",
    "        DF[:,9] = exp_mass_to_charge\n",
    "        DF[:,10] = charge\n",
    "        DF[:,11] = smsnet_df1['RT'] \n",
    "        DF[:,12] = '-' \n",
    "        DF[:,13] = 1\n",
    "        DF[:,14] = Mass_peptide \n",
    "        DF[:,15] = ppm\n",
    "        DF[:,17] = smsnet_df1['aaScore'] \n",
    "        DF[:,18] = smsnet_df1['Peptide']\n",
    "        DF[:,19] = 'HCD' \n",
    "        name = ['Fraction','Source File','Feature','Peptide','Scan','Tag Length','Denovo Score','ALC (%)','length','m/z','z','RT','Predict RT','Area','Mass','ppm','PTM','local confidence (%)','tag (>=0%)','mode']\n",
    "        df = pd.DataFrame(DF, columns=name)\n",
    "        df.to_csv('../../../Stitch_Assembly/SMSNet/test.csv',index=False)\n",
    "        df['Denovo Score'] = df['Denovo Score'].astype(int)\n",
    "        df = df[df['Denovo Score'] >= SocreAll[tool]]\n",
    "        df.to_csv(smsnet_out,index=False)\n",
    "\n",
    "        with open(batchfile_dir,'r') as fr:\n",
    "            lines = fr.readlines()\n",
    "        newfile = resultdir + batchfile_dir.split('/')[-1]\n",
    "        with open(newfile,'w') as fw:\n",
    "            for line in lines:\n",
    "                if 'Path     : ' in line:\n",
    "                    fw.write(line.replace('../datasets/200305_HER_test_04_DENOVO.csv', smsnet_out))\n",
    "                elif 'CutoffALC: ' in line:\n",
    "                    fw.write(line.replace('95', str(quality_cutoff_Stitch_local)))\n",
    "                else:\n",
    "                    fw.write(line)\n",
    "\n",
    "    subprocess.run(\n",
    "    ['../../../stitch-v1.5.0-linux/stitch.bin', 'run', newfile], stdout=subprocess.DEVNULL)\n",
    "\n",
    "Names = {\n",
    " '50ugmAb1':'mAb1',\n",
    " '100ugmAb2':'mAb2',\n",
    " '200ugmAb3':'mAb3',\n",
    " '20230210-mAb4':'mAb4',\n",
    " '20230707-mAb5':'mAb5',\n",
    " '20231203-mAb6':'mAb6',\n",
    " '20231221-mAb7':'mAb7',    \n",
    " '20250415-mAb8':'mAb8'}\n",
    "\n",
    "path = '../../../denovo/SMSNet/'\n",
    "out_path = '../../../Stitch_Assembly/SMSNet/'\n",
    "mgf_path = '../../../monoclonal_antibody/'\n",
    "mabs = os.listdir(path)\n",
    "mabs = [mab for mab in mabs if '131ab_EThcd' not in mab and '.yaml' not in mab and 'Herceptin' not in mab and 'anti-FLAG-M2' not in mab]\n",
    "for mab in mabs:\n",
    "    print(mab)\n",
    "    smsnet_path = path + mab +'/'+'spectrum_'+Names[mab]+'_HCD_forSMSNet'\n",
    "    resultdir = out_path + mab +'/'\n",
    "    mgf_file = mgf_path + mab +'/process1/'+'spectrum_'+ Names[mab] +'_HCD.mgf'\n",
    "    batchfile_dir = '../../../stitch-v1.5.0-linux/batchfiles/' + mab + '.txt'\n",
    "    process_smsnet_Stitch(smsnet_path,mgf_file,batchfile_dir,resultdir)       "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}